<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.6"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: bio::alphabet::dna16sam Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign">
   <div id="projectname">BioC++<span id="projectnumber">&#160;core-0.7.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
    <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <span id="MSearchSelect"                onmouseover="return searchBox.OnSearchSelectShow()"                onmouseout="return searchBox.OnSearchSelectHide()">&#160;</span>
          <input type="text" id="MSearchField" value="" placeholder="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.6 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('classbio_1_1alphabet_1_1dna16sam.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="classbio_1_1alphabet_1_1dna16sam-members.html">List of all members</a>  </div>
  <div class="headertitle"><div class="title">bio::alphabet::dna16sam Class Reference<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a> &raquo; <a class="el" href="group__nucleotide.html">Nucleotide</a></div></div></div>
</div><!--header-->
<div class="contents">

<p>A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=')..  
 <a href="classbio_1_1alphabet_1_1dna16sam.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="dna16sam_8hpp_source.html">bio/alphabet/nucleotide/dna16sam.hpp</a>&gt;</code></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Inheritance diagram for bio::alphabet::dna16sam:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><div class="zoom"><iframe scrolling="no" frameborder="0" src="classbio_1_1alphabet_1_1dna16sam__inherit__graph.svg" width="100%" height="553"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-methods" name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Constructors, destructor and assignment</div></td></tr>
<tr class="memitem:af84edd8d26ca106a7a69a4690ae53533"><td class="memItemLeft" align="right" valign="top"><a id="af84edd8d26ca106a7a69a4690ae53533" name="af84edd8d26ca106a7a69a4690ae53533"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><b>dna16sam</b> () noexcept=default</td></tr>
<tr class="memdesc:af84edd8d26ca106a7a69a4690ae53533"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:af84edd8d26ca106a7a69a4690ae53533"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5c7ccc29de00b3856037260c64cd79d3"><td class="memItemLeft" align="right" valign="top"><a id="a5c7ccc29de00b3856037260c64cd79d3" name="a5c7ccc29de00b3856037260c64cd79d3"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><b>dna16sam</b> (<a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a5c7ccc29de00b3856037260c64cd79d3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a5c7ccc29de00b3856037260c64cd79d3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4d8ff512a5d2e0e7cae6d36a3db8a566"><td class="memItemLeft" align="right" valign="top"><a id="a4d8ff512a5d2e0e7cae6d36a3db8a566" name="a4d8ff512a5d2e0e7cae6d36a3db8a566"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><b>dna16sam</b> (<a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:a4d8ff512a5d2e0e7cae6d36a3db8a566"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a4d8ff512a5d2e0e7cae6d36a3db8a566"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a244356932bea00a4e555a256233806f0"><td class="memItemLeft" align="right" valign="top"><a id="a244356932bea00a4e555a256233806f0" name="a244356932bea00a4e555a256233806f0"></a>
constexpr <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator=</b> (<a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a244356932bea00a4e555a256233806f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a244356932bea00a4e555a256233806f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a34df33b73d12437fc33abdaea226f3ba"><td class="memItemLeft" align="right" valign="top"><a id="a34df33b73d12437fc33abdaea226f3ba" name="a34df33b73d12437fc33abdaea226f3ba"></a>
constexpr <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator=</b> (<a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:a34df33b73d12437fc33abdaea226f3ba"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a34df33b73d12437fc33abdaea226f3ba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8bb2192e1198286c24bbf79dff24f60a"><td class="memItemLeft" align="right" valign="top"><a id="a8bb2192e1198286c24bbf79dff24f60a" name="a8bb2192e1198286c24bbf79dff24f60a"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>~dna16sam</b> () noexcept=default</td></tr>
<tr class="memdesc:a8bb2192e1198286c24bbf79dff24f60a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a8bb2192e1198286c24bbf79dff24f60a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:ab6125825046be181c7ea74f0dc431400"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html#ab6125825046be181c7ea74f0dc431400">complement</a> () const noexcept</td></tr>
<tr class="memdesc:ab6125825046be181c7ea74f0dc431400"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the complement of the letter.  <br /></td></tr>
<tr class="separator:ab6125825046be181c7ea74f0dc431400"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a83aefc6fee4de78ac746137fc9d73073"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet_1_1base.html#aea55b90bf80cbf099153009056ea4fa3">char_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1base.html#a83aefc6fee4de78ac746137fc9d73073">to_char</a> () const noexcept</td></tr>
<tr class="memdesc:a83aefc6fee4de78ac746137fc9d73073"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter as a character of char_type.  <br /></td></tr>
<tr class="separator:a83aefc6fee4de78ac746137fc9d73073"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af88a11dccd87dafe33b180fc24343b5e"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet_1_1base.html#a89ed469e334a469b1da901ba3b0785a4">rank_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1base.html#af88a11dccd87dafe33b180fc24343b5e">to_rank</a> () const noexcept</td></tr>
<tr class="memdesc:af88a11dccd87dafe33b180fc24343b5e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter's numeric value (rank in the alphabet).  <br /></td></tr>
<tr class="separator:af88a11dccd87dafe33b180fc24343b5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Write functions</div></td></tr>
<tr class="memitem:a9dfef1259c4c0112e490255343de0d2f"><td class="memItemLeft" align="right" valign="top">constexpr derived_type &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1base.html#a9dfef1259c4c0112e490255343de0d2f">assign_char</a> (<a class="el" href="classbio_1_1alphabet_1_1base.html#aea55b90bf80cbf099153009056ea4fa3">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a9dfef1259c4c0112e490255343de0d2f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a character, implicitly converts invalid characters.  <br /></td></tr>
<tr class="separator:a9dfef1259c4c0112e490255343de0d2f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42e4ec1907956d855b96c3ea22383ce3"><td class="memItemLeft" align="right" valign="top">constexpr derived_type &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1base.html#a42e4ec1907956d855b96c3ea22383ce3">assign_rank</a> (<a class="el" href="classbio_1_1alphabet_1_1base.html#a89ed469e334a469b1da901ba3b0785a4">rank_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a42e4ec1907956d855b96c3ea22383ce3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a numeric value.  <br /></td></tr>
<tr class="separator:a42e4ec1907956d855b96c3ea22383ce3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-static-methods" name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a93e8abe9d38986faac6431136bba069b"><td class="memItemLeft" align="right" valign="top">static constexpr bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html#a93e8abe9d38986faac6431136bba069b">char_is_valid</a> (<a class="el" href="classbio_1_1alphabet_1_1base.html#aea55b90bf80cbf099153009056ea4fa3">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a93e8abe9d38986faac6431136bba069b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Validate whether a character value has a one-to-one mapping to an alphabet value.  <br /></td></tr>
<tr class="separator:a93e8abe9d38986faac6431136bba069b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pub-static-attribs" name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a834f007f409b2761dc774fc45e457b11"><td class="memItemLeft" align="right" valign="top"><a id="a834f007f409b2761dc774fc45e457b11" name="a834f007f409b2761dc774fc45e457b11"></a>
static constexpr size_t&#160;</td><td class="memItemRight" valign="bottom"><b>alphabet_size</b> = <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a></td></tr>
<tr class="memdesc:a834f007f409b2761dc774fc45e457b11"><td class="mdescLeft">&#160;</td><td class="mdescRight">The size of the alphabet, i.e. the number of different values it can take. <br /></td></tr>
<tr class="separator:a834f007f409b2761dc774fc45e457b11"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="pro-types" name="pro-types"></a>
Protected Types</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Member types</div></td></tr>
<tr class="memitem:aea55b90bf80cbf099153009056ea4fa3"><td class="memItemLeft" align="right" valign="top"><a id="aea55b90bf80cbf099153009056ea4fa3" name="aea55b90bf80cbf099153009056ea4fa3"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>char_type</b> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">std::conditional_t</a>&lt; std::same_as&lt; <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a>, void &gt;, char, <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a> &gt;</td></tr>
<tr class="memdesc:aea55b90bf80cbf099153009056ea4fa3"><td class="mdescLeft">&#160;</td><td class="mdescRight">The char representation; conditional needed to make semi alphabet definitions legal. <br /></td></tr>
<tr class="separator:aea55b90bf80cbf099153009056ea4fa3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a89ed469e334a469b1da901ba3b0785a4"><td class="memItemLeft" align="right" valign="top"><a id="a89ed469e334a469b1da901ba3b0785a4" name="a89ed469e334a469b1da901ba3b0785a4"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>rank_type</b> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">meta::detail::min_viable_uint_t</a>&lt; <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a> - 1 &gt;</td></tr>
<tr class="memdesc:a89ed469e334a469b1da901ba3b0785a4"><td class="mdescLeft">&#160;</td><td class="mdescRight">The type of the alphabet when represented as a number (e.g. via <a class="el" href="group__alphabet.html#ga2a48a5c1883184e362f09e32ac33241e" title="Return the rank representation of a (semi-)alphabet object.">to_rank()</a>). <br /></td></tr>
<tr class="separator:a89ed469e334a469b1da901ba3b0785a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="related" name="related"></a>
Related Functions</h2></td></tr>
<tr><td class="ititle" colspan="2"><p>(Note that these are not member functions.) </p>
</td></tr>
<tr><td colspan="2"><div class="groupHeader">Literals</div></td></tr>
<tr class="memitem:a4e42355b37b437dd0172aec0bc326051"><td class="memItemLeft" align="right" valign="top">consteval <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html#a4e42355b37b437dd0172aec0bc326051">operator&quot;&quot;_dna16sam</a> (char const c)</td></tr>
<tr class="memdesc:a4e42355b37b437dd0172aec0bc326051"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> char literal.  <br /></td></tr>
<tr class="separator:a4e42355b37b437dd0172aec0bc326051"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7705ecb2474ae2e0b1027e1a724042a3"><td class="memTemplParams" colspan="2">template&lt;meta::detail::literal_buffer_string str&gt; </td></tr>
<tr class="memitem:a7705ecb2474ae2e0b1027e1a724042a3"><td class="memTemplItemLeft" align="right" valign="top">constexpr <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> &gt;&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html#a7705ecb2474ae2e0b1027e1a724042a3">operator&quot;&quot;_dna16sam</a> ()</td></tr>
<tr class="memdesc:a7705ecb2474ae2e0b1027e1a724042a3"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> string literal.  <br /></td></tr>
<tr class="separator:a7705ecb2474ae2e0b1027e1a724042a3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=').</p>
<p>. </p>
<p>The <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> alphabet is the nucleotide alphabet used inside the SAM, BAM and CRAM formats. It has all the letters of the <a class="el" href="classbio_1_1alphabet_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::dna15</a> alphabet and the extra alphabet character '=' which denotes a nucleotide character identical to the reference. Without the context of this reference sequence, no assumptions can be made about the actual value of '=' letter.</p>
<p>Note that you can assign 'U' as a character to <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">dna16sam</a> and it will silently be converted to 'T'. Lower case letters are accepted when assigning from char (just like <a class="el" href="classbio_1_1alphabet_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::dna15</a>) and unknown characters are silently converted to 'N'.</p>
<p>The complement is the same as for <a class="el" href="classbio_1_1alphabet_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::dna15</a>, with the addition that the complement of '=' is unknown and therefore set to 'N'.</p>
<p>The binary representation of an object of this type corresponds to the binary representation in BAM/CRAM.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna16sam_8hpp.html">bio/alphabet/nucleotide/dna16sam.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using namespace </span><a class="code hl_namespace" href="namespacebio_1_1alphabet_1_1literals.html">bio::alphabet::literals</a>;</div>
<div class="line"> </div>
<div class="line">    <a class="code hl_class" href="classbio_1_1alphabet_1_1dna16sam.html">bio::alphabet::dna16sam</a> my_letter{<span class="charliteral">&#39;A&#39;</span>_dna16sam};</div>
<div class="line"> </div>
<div class="line">    my_letter.<a class="code hl_function" href="classbio_1_1alphabet_1_1base.html#a9dfef1259c4c0112e490255343de0d2f">assign_char</a>(<span class="charliteral">&#39;=&#39;</span>);</div>
<div class="line"> </div>
<div class="line">    my_letter.assign_char(<span class="charliteral">&#39;F&#39;</span>); <span class="comment">// unknown characters are implicitly converted to N.</span></div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{}\n&quot;</span>, my_letter); <span class="comment">// &quot;N&quot;;</span></div>
<div class="line">}</div>
<div class="ttc" id="aclassbio_1_1alphabet_1_1base_html_a9dfef1259c4c0112e490255343de0d2f"><div class="ttname"><a href="classbio_1_1alphabet_1_1base.html#a9dfef1259c4c0112e490255343de0d2f">bio::alphabet::base::assign_char</a></div><div class="ttdeci">constexpr derived_type &amp; assign_char(char_type const c) noexcept</div><div class="ttdoc">Assign from a character, implicitly converts invalid characters.</div><div class="ttdef"><b>Definition:</b> base.hpp:145</div></div>
<div class="ttc" id="aclassbio_1_1alphabet_1_1dna16sam_html"><div class="ttname"><a href="classbio_1_1alphabet_1_1dna16sam.html">bio::alphabet::dna16sam</a></div><div class="ttdoc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=')....</div><div class="ttdef"><b>Definition:</b> dna16sam.hpp:49</div></div>
<div class="ttc" id="adna16sam_8hpp_html"><div class="ttname"><a href="dna16sam_8hpp.html">dna16sam.hpp</a></div><div class="ttdoc">Provides bio::alphabet::dna16sam.</div></div>
<div class="ttc" id="afmt_8hpp_html"><div class="ttname"><a href="fmt_8hpp.html">fmt.hpp</a></div><div class="ttdoc">Core alphabet concept and free function/type trait wrappers.</div></div>
<div class="ttc" id="anamespacebio_1_1alphabet_1_1literals_html"><div class="ttname"><a href="namespacebio_1_1alphabet_1_1literals.html">bio::alphabet::literals</a></div><div class="ttdoc">An inline namespace for alphabet literals. It exists to safely allow using namespace.</div><div class="ttdef"><b>Definition:</b> aa10li.hpp:183</div></div>
</div><!-- fragment --> </div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a9dfef1259c4c0112e490255343de0d2f" name="a9dfef1259c4c0112e490255343de0d2f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a9dfef1259c4c0112e490255343de0d2f">&#9670;&#160;</a></span>assign_char()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a>  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr derived_type &amp; <a class="el" href="classbio_1_1alphabet_1_1base.html">bio::alphabet::base</a>&lt; derived_type, <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a>, <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a> &gt;::assign_char </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet_1_1base.html#aea55b90bf80cbf099153009056ea4fa3">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a character, implicitly converts invalid characters. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The character to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga0c2eb25dabf59ccf9992b2661083ea5b" title="Assign a char to an alphabet object.">bio::alphabet::assign_char_to</a>, required to model <a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a42e4ec1907956d855b96c3ea22383ce3" name="a42e4ec1907956d855b96c3ea22383ce3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a42e4ec1907956d855b96c3ea22383ce3">&#9670;&#160;</a></span>assign_rank()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a>  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr derived_type &amp; <a class="el" href="classbio_1_1alphabet_1_1base.html">bio::alphabet::base</a>&lt; derived_type, <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a>, <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a> &gt;::assign_rank </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet_1_1base.html#a89ed469e334a469b1da901ba3b0785a4">rank_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a numeric value. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The rank to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#gab718eab4e0f0e1b44f78b7600dee5e4d" title="Assign a rank to an alphabet object.">bio::alphabet::assign_rank_to</a>, required to model <a class="el" href="conceptbio_1_1alphabet_1_1semialphabet.html" title="The basis for bio::alphabet::alphabet, but requires only rank interface (not char).">bio::alphabet::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a93e8abe9d38986faac6431136bba069b" name="a93e8abe9d38986faac6431136bba069b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a93e8abe9d38986faac6431136bba069b">&#9670;&#160;</a></span>char_is_valid()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static constexpr bool <a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html">bio::alphabet::nucleotide_base</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> , <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a> &gt;::char_is_valid </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet_1_1base.html#aea55b90bf80cbf099153009056ea4fa3">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Validate whether a character value has a one-to-one mapping to an alphabet value. </p>
<p>Satisfies the bio::alphabet::semialphabet::char_is_valid_for() requirement via the <a class="el" href="group__alphabet.html#ga64c00e15570d3e7394f425c810e617ce" title="Returns whether a character is in the valid set of a bio::alphabet::alphabet (usually implies a bijec...">bio::alphabet::char_is_valid_for()</a> wrapper.</p>
<p>Behaviour specific to nucleotides: True also for lower case letters that silently convert to their upper case <b>and</b> true also for U/T respectively, e.g. 'U' is a valid character for <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a>, because its informational content is identical to 'T'.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="ab6125825046be181c7ea74f0dc431400" name="ab6125825046be181c7ea74f0dc431400"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab6125825046be181c7ea74f0dc431400">&#9670;&#160;</a></span>complement()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a>  <a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html">bio::alphabet::nucleotide_base</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> , <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a> &gt;::complement </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the complement of the letter. </p>
<p>See <a class="el" href="group__nucleotide.html">Nucleotide</a> for the actual values.</p>
<p>Provides an implementation for <a class="el" href="group__nucleotide.html#ga174261f8f888bbce628ffe01097a9b48" title="Return the complement of a nucleotide object.">bio::alphabet::complement</a>, required to model <a class="el" href="conceptbio_1_1alphabet_1_1nucleotide.html" title="A concept that indicates whether an alphabet represents nucleotides.">bio::alphabet::nucleotide</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a83aefc6fee4de78ac746137fc9d73073" name="a83aefc6fee4de78ac746137fc9d73073"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a83aefc6fee4de78ac746137fc9d73073">&#9670;&#160;</a></span>to_char()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a>  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet_1_1base.html#aea55b90bf80cbf099153009056ea4fa3">char_type</a> <a class="el" href="classbio_1_1alphabet_1_1base.html">bio::alphabet::base</a>&lt; derived_type, <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a>, <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a> &gt;::to_char </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter as a character of char_type. </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga064e386f9360cca0fd3702b4957a27a0" title="Return the char representation of an alphabet object.">bio::alphabet::to_char</a>, required to model <a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="af88a11dccd87dafe33b180fc24343b5e" name="af88a11dccd87dafe33b180fc24343b5e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af88a11dccd87dafe33b180fc24343b5e">&#9670;&#160;</a></span>to_rank()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a>  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet_1_1base.html#a89ed469e334a469b1da901ba3b0785a4">rank_type</a> <a class="el" href="classbio_1_1alphabet_1_1base.html">bio::alphabet::base</a>&lt; derived_type, <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a>, <a class="el" href="group__alphabet.html#gaa16ddb3f7d394f68af7f530aeb24e5cf">char_t</a> &gt;::to_rank </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter's numeric value (rank in the alphabet). </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga2a48a5c1883184e362f09e32ac33241e" title="Return the rank representation of a (semi-)alphabet object.">bio::alphabet::to_rank</a>, required to model <a class="el" href="conceptbio_1_1alphabet_1_1semialphabet.html" title="The basis for bio::alphabet::alphabet, but requires only rank interface (not char).">bio::alphabet::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<h2 class="groupheader">Friends And Related Function Documentation</h2>
<a id="a7705ecb2474ae2e0b1027e1a724042a3" name="a7705ecb2474ae2e0b1027e1a724042a3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7705ecb2474ae2e0b1027e1a724042a3">&#9670;&#160;</a></span>operator&quot;&quot;_dna16sam() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;meta::detail::literal_buffer_string str&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> &gt; operator&quot;&quot;_dna16sam </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">related</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> string literal. </p>
<dl class="section return"><dt>Returns</dt><dd>std::vector&lt;bio::alphabet::dna16sam&gt;</dd></dl>
<p>You can use this string literal to easily create a std::vector&lt;bio::alphabet::dna16sam&gt;:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna16sam_8hpp.html">bio/alphabet/nucleotide/dna16sam.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using namespace </span><a class="code hl_namespace" href="namespacebio_1_1alphabet_1_1literals.html">bio::alphabet::literals</a>;</div>
<div class="line"> </div>
<div class="line">    <a class="code hl_classRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector&lt;bio::alphabet::dna16sam&gt;</a> foo = <span class="stringliteral">&quot;ACG==A&quot;</span>_dna16sam;</div>
<div class="line">    <span class="keyword">auto</span> bar = <span class="stringliteral">&quot;A=GTTC&quot;</span>_dna16sam;</div>
<div class="line"> </div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{}\n{}\n&quot;</span>, foo, bar);</div>
<div class="line">}</div>
<div class="ttc" id="avector_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a></div></div>
</div><!-- fragment --> 
</div>
</div>
<a id="a4e42355b37b437dd0172aec0bc326051" name="a4e42355b37b437dd0172aec0bc326051"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a4e42355b37b437dd0172aec0bc326051">&#9670;&#160;</a></span>operator&quot;&quot;_dna16sam() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">consteval <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a> operator&quot;&quot;_dna16sam </td>
          <td>(</td>
          <td class="paramtype">char const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">related</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> char literal. </p>
<dl class="section return"><dt>Returns</dt><dd><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>bio/alphabet/nucleotide/<a class="el" href="dna16sam_8hpp_source.html">dna16sam.hpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="navelem"><a class="el" href="namespacebio.html">bio</a></li><li class="navelem"><a class="el" href="namespacebio_1_1alphabet.html">alphabet</a></li><li class="navelem"><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a></li>
    <li class="footer">Generated on Thu Mar 30 2023 15:25:52 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.6 </li>
  </ul>
</div>
</body>
</html>
