<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: bio::dna3bs Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">0.1.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('classbio_1_1dna3bs.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#related">Related Functions</a> &#124;
<a href="classbio_1_1dna3bs-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">bio::dna3bs Class Reference<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a> &raquo; <a class="el" href="group__nucleotide.html">Nucleotide</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).  
 <a href="classbio_1_1dna3bs.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="dna3bs_8hpp_source.html">bio/alphabet/nucleotide/dna3bs.hpp</a>&gt;</code></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Inheritance diagram for bio::dna3bs:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><div class="zoom"><iframe scrolling="no" frameborder="0" src="classbio_1_1dna3bs__inherit__graph.svg" width="100%" height="572"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div></div>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Constructors, destructor and assignment</div></td></tr>
<tr class="memitem:a7101da2e42e6aaf20dc7f48aab74dcb4"><td class="memItemLeft" align="right" valign="top"><a id="a7101da2e42e6aaf20dc7f48aab74dcb4"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a7101da2e42e6aaf20dc7f48aab74dcb4">dna3bs</a> () noexcept=default</td></tr>
<tr class="memdesc:a7101da2e42e6aaf20dc7f48aab74dcb4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a7101da2e42e6aaf20dc7f48aab74dcb4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a983742ed357c4b373d1480088046152d"><td class="memItemLeft" align="right" valign="top"><a id="a983742ed357c4b373d1480088046152d"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a983742ed357c4b373d1480088046152d">dna3bs</a> (<a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a983742ed357c4b373d1480088046152d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a983742ed357c4b373d1480088046152d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5fac09054d48557c2224f669173e4820"><td class="memItemLeft" align="right" valign="top"><a id="a5fac09054d48557c2224f669173e4820"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a5fac09054d48557c2224f669173e4820">dna3bs</a> (<a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:a5fac09054d48557c2224f669173e4820"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a5fac09054d48557c2224f669173e4820"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4ac1fcf7e456ab1b96517811c3c36e29"><td class="memItemLeft" align="right" valign="top"><a id="a4ac1fcf7e456ab1b96517811c3c36e29"></a>
constexpr <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a4ac1fcf7e456ab1b96517811c3c36e29">operator=</a> (<a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a4ac1fcf7e456ab1b96517811c3c36e29"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a4ac1fcf7e456ab1b96517811c3c36e29"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76dbdc9d03d60f794126f7474f939972"><td class="memItemLeft" align="right" valign="top"><a id="a76dbdc9d03d60f794126f7474f939972"></a>
constexpr <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a76dbdc9d03d60f794126f7474f939972">operator=</a> (<a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:a76dbdc9d03d60f794126f7474f939972"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a76dbdc9d03d60f794126f7474f939972"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8688993570dfe071cdc291ee04a5be7c"><td class="memItemLeft" align="right" valign="top"><a id="a8688993570dfe071cdc291ee04a5be7c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a8688993570dfe071cdc291ee04a5be7c">~dna3bs</a> () noexcept=default</td></tr>
<tr class="memdesc:a8688993570dfe071cdc291ee04a5be7c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a8688993570dfe071cdc291ee04a5be7c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a0e9f68ab92a2638509288eeb840aa2ec"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1nucleotide__base.html#a0e9f68ab92a2638509288eeb840aa2ec">complement</a> () const noexcept</td></tr>
<tr class="memdesc:a0e9f68ab92a2638509288eeb840aa2ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the complement of the letter.  <a href="classbio_1_1nucleotide__base.html#a0e9f68ab92a2638509288eeb840aa2ec">More...</a><br /></td></tr>
<tr class="separator:a0e9f68ab92a2638509288eeb840aa2ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a6616ba2cb25069f5192589ad53651a88"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">to_char</a> () const noexcept</td></tr>
<tr class="memdesc:a6616ba2cb25069f5192589ad53651a88"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter as a character of char_type.  <a href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">More...</a><br /></td></tr>
<tr class="separator:a6616ba2cb25069f5192589ad53651a88"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a443c2f7716e384ea2acec1d2bced7c8b"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">to_rank</a> () const noexcept</td></tr>
<tr class="memdesc:a443c2f7716e384ea2acec1d2bced7c8b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter's numeric value (rank in the alphabet).  <a href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">More...</a><br /></td></tr>
<tr class="separator:a443c2f7716e384ea2acec1d2bced7c8b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Write functions</div></td></tr>
<tr class="memitem:a822d399ab808842870a4a4feb5490849"><td class="memItemLeft" align="right" valign="top">constexpr derived_type &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">assign_char</a> (<a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a822d399ab808842870a4a4feb5490849"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a character, implicitly converts invalid characters.  <a href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">More...</a><br /></td></tr>
<tr class="separator:a822d399ab808842870a4a4feb5490849"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acdd042ae2f04d986f65a7b60edd6e83e"><td class="memItemLeft" align="right" valign="top">constexpr derived_type &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">assign_rank</a> (<a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> const c) noexcept</td></tr>
<tr class="memdesc:acdd042ae2f04d986f65a7b60edd6e83e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a numeric value.  <a href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">More...</a><br /></td></tr>
<tr class="separator:acdd042ae2f04d986f65a7b60edd6e83e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:ad6758a49085273e9f1b4811ebd2d1c8a"><td class="memItemLeft" align="right" valign="top">static constexpr bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1nucleotide__base.html#ad6758a49085273e9f1b4811ebd2d1c8a">char_is_valid</a> (<a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:ad6758a49085273e9f1b4811ebd2d1c8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Validate whether a character value has a one-to-one mapping to an alphabet value.  <a href="classbio_1_1nucleotide__base.html#ad6758a49085273e9f1b4811ebd2d1c8a">More...</a><br /></td></tr>
<tr class="separator:ad6758a49085273e9f1b4811ebd2d1c8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a989cf18d33d15b0d858c2dd540480468"><td class="memItemLeft" align="right" valign="top"><a id="a989cf18d33d15b0d858c2dd540480468"></a>
static constexpr <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">detail::min_viable_uint_t</a>&lt; size &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a989cf18d33d15b0d858c2dd540480468">alphabet_size</a> = size</td></tr>
<tr class="memdesc:a989cf18d33d15b0d858c2dd540480468"><td class="mdescLeft">&#160;</td><td class="mdescRight">The size of the alphabet, i.e. the number of different values it can take. <br /></td></tr>
<tr class="separator:a989cf18d33d15b0d858c2dd540480468"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-types"></a>
Protected Types</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Member types</div></td></tr>
<tr class="memitem:a9934c83725128fc1cde969618f2b4f66"><td class="memItemLeft" align="right" valign="top"><a id="a9934c83725128fc1cde969618f2b4f66"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">std::conditional_t</a>&lt; std::same_as&lt; char_t, void &gt;, char, char_t &gt;</td></tr>
<tr class="memdesc:a9934c83725128fc1cde969618f2b4f66"><td class="mdescLeft">&#160;</td><td class="mdescRight">The char representation; conditional needed to make semi alphabet definitions legal. <br /></td></tr>
<tr class="separator:a9934c83725128fc1cde969618f2b4f66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a040aa76e8e3fc0e5735c4b135f302c13"><td class="memItemLeft" align="right" valign="top"><a id="a040aa76e8e3fc0e5735c4b135f302c13"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">detail::min_viable_uint_t</a>&lt; size - 1 &gt;</td></tr>
<tr class="memdesc:a040aa76e8e3fc0e5735c4b135f302c13"><td class="mdescLeft">&#160;</td><td class="mdescRight">The type of the alphabet when represented as a number (e.g. via <a class="el" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b" title="Return the letter&#39;s numeric value (rank in the alphabet).">to_rank()</a>). <br /></td></tr>
<tr class="separator:a040aa76e8e3fc0e5735c4b135f302c13"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="related"></a>
Related Functions</h2></td></tr>
<tr><td class="ititle" colspan="2"><p>(Note that these are not member functions.) </p>
</td></tr>
<tr class="memitem:a1c51e23a1a125a3a554a92eed266d911"><td class="memItemLeft" align="right" valign="top"><a id="a1c51e23a1a125a3a554a92eed266d911"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a1c51e23a1a125a3a554a92eed266d911">dna3bs_vector</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a>&lt; <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> &gt;</td></tr>
<tr class="memdesc:a1c51e23a1a125a3a554a92eed266d911"><td class="mdescLeft">&#160;</td><td class="mdescRight">Alias for an <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a> of <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">bio::dna3bs</a>. <br /></td></tr>
<tr class="separator:a1c51e23a1a125a3a554a92eed266d911"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Literals</div></td></tr>
<tr class="memitem:a1a688956c109d7472ce6aeeb7ac8b277"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a1a688956c109d7472ce6aeeb7ac8b277">operator&quot;&quot;_dna3bs</a> (char const c) noexcept</td></tr>
<tr class="memdesc:a1a688956c109d7472ce6aeeb7ac8b277"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">bio::dna3bs</a> char literal.  <a href="classbio_1_1dna3bs.html#a1a688956c109d7472ce6aeeb7ac8b277">More...</a><br /></td></tr>
<tr class="separator:a1a688956c109d7472ce6aeeb7ac8b277"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6d873fe73b62acf41564b4dd6b6f7708"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classbio_1_1dna3bs.html#a1c51e23a1a125a3a554a92eed266d911">dna3bs_vector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1dna3bs.html#a6d873fe73b62acf41564b4dd6b6f7708">operator&quot;&quot;_dna3bs</a> (char const *s, <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/size_t.html">std::size_t</a> n)</td></tr>
<tr class="memdesc:a6d873fe73b62acf41564b4dd6b6f7708"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">bio::dna3bs</a> string literal.  <a href="classbio_1_1dna3bs.html#a6d873fe73b62acf41564b4dd6b6f7708">More...</a><br /></td></tr>
<tr class="separator:a6d873fe73b62acf41564b4dd6b6f7708"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)). </p>
<p>This alphabet represents a reduced version that can be used when dealing with bisulfite-converted data. All 'C's are converted to a 'T' in order to allow comparison of normal sequences with bisulfite-converted sequences. For completeness, this nucleotide alphabet has a complement table, however, it is not recommended to use it when dealing with bisulfite data because the complement of T is ambiguous in reads from bisulfite sequencing. A 'T' can represent a true thymidine or an unmethylated 'C' that was converted into a 'T'. Therefore, complementing a dna4bs sequence will further reduce the alphabet to only 'T' and 'A', thereby loosing all information about 'G'. When working with bisulfite data, we recommend to create the reverse complement of the dna4/5/15 range first and convert to <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">dna3bs</a> later. This avoids simplifying the data by automatically setting 'A' as the complement of 'C'. As an example: The sequence 'ACGTGC' in <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">dna4</a> would be 'ATGTGT' in <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">dna3bs</a>. The complement of this <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">dna3bs</a> sequence would be 'TATATA', however when complementing the <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">dna4</a> sequence first and afterwards transforming it into <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">dna3bs</a>, it would be 'TGTATG' which preserves more information from the original sequence.</p>
<p>Like most alphabets, this alphabet cannot be initialised directly from its character representation. Instead initialise/assign from the character literal or use the function <a class="el" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849" title="Assign from a character, implicitly converts invalid characters.">bio::dna3bs::assign_char()</a>.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna3bs_8hpp.html">bio/alphabet/nucleotide/dna3bs.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna3bs;</div>
<div class="line"> </div>
<div class="line">    <a class="code" href="classbio_1_1dna3bs.html">bio::dna3bs</a> my_letter{<span class="charliteral">&#39;A&#39;</span>_dna3bs};</div>
<div class="line"> </div>
<div class="line">    my_letter.<a class="code" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">assign_char</a>(<span class="charliteral">&#39;C&#39;</span>); <span class="comment">// all C will be converted to T.</span></div>
<div class="line">    <span class="keywordflow">if</span> (my_letter.to_char() == <span class="charliteral">&#39;T&#39;</span>)</div>
<div class="line">        fmt::print(<span class="stringliteral">&quot;{}&quot;</span>, <span class="stringliteral">&quot;yeah\n&quot;</span>); <span class="comment">// &quot;yeah&quot;;</span></div>
<div class="line"> </div>
<div class="line">    my_letter.assign_char(<span class="charliteral">&#39;F&#39;</span>); <span class="comment">// unknown characters are implicitly converted to A.</span></div>
<div class="line">    <span class="keywordflow">if</span> (my_letter.to_char() == <span class="charliteral">&#39;A&#39;</span>)</div>
<div class="line">        fmt::print(<span class="stringliteral">&quot;{}&quot;</span>, <span class="stringliteral">&quot;yeah\n&quot;</span>); <span class="comment">// &quot;yeah&quot;;</span></div>
<div class="line"> </div>
<div class="line">    <span class="keywordflow">return</span> 0;</div>
<div class="line">}</div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_a822d399ab808842870a4a4feb5490849"><div class="ttname"><a href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">bio::alphabet_base::assign_char</a></div><div class="ttdeci">constexpr derived_type &amp; assign_char(char_type const c) noexcept</div><div class="ttdoc">Assign from a character, implicitly converts invalid characters.</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:143</div></div>
<div class="ttc" id="aclassbio_1_1dna3bs_html"><div class="ttname"><a href="classbio_1_1dna3bs.html">bio::dna3bs</a></div><div class="ttdoc">The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).</div><div class="ttdef"><b>Definition:</b> dna3bs.hpp:57</div></div>
<div class="ttc" id="adna3bs_8hpp_html"><div class="ttname"><a href="dna3bs_8hpp.html">dna3bs.hpp</a></div><div class="ttdoc">Provides bio::dna3bs, container aliases and string literals.</div></div>
<div class="ttc" id="afmt_8hpp_html"><div class="ttname"><a href="fmt_8hpp.html">fmt.hpp</a></div><div class="ttdoc">Core alphabet concept and free function/type trait wrappers.</div></div>
</div><!-- fragment --> </div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a822d399ab808842870a4a4feb5490849"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a822d399ab808842870a4a4feb5490849">&#9670;&nbsp;</a></span>assign_char()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr derived_type&amp; <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::assign_char </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a character, implicitly converts invalid characters. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The character to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga05da42eb0b9a48f4d1a554c7d3a9044d" title="Assign a character to an alphabet object.">bio::assign_char_to</a>, required to model <a class="el" href="interfacebio_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="acdd042ae2f04d986f65a7b60edd6e83e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acdd042ae2f04d986f65a7b60edd6e83e">&#9670;&nbsp;</a></span>assign_rank()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr derived_type&amp; <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::assign_rank </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a numeric value. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The rank to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#gae34fa2b37a8b40f2f2fa64fed7bbb410" title="Assign a rank to an alphabet object.">bio::assign_rank_to</a>, required to model <a class="el" href="interfacebio_1_1semialphabet.html" title="The basis for bio::alphabet, but requires only rank interface (not char).">bio::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="ad6758a49085273e9f1b4811ebd2d1c8a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad6758a49085273e9f1b4811ebd2d1c8a">&#9670;&nbsp;</a></span>char_is_valid()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static constexpr bool <a class="el" href="classbio_1_1nucleotide__base.html">bio::nucleotide_base</a>&lt; <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> , size &gt;::char_is_valid </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Validate whether a character value has a one-to-one mapping to an alphabet value. </p>
<p>Satisfies the <a class="el" href="group__alphabet.html#ga510b0657b8a3ae7d304999afd8fdcd52" title="Returns whether a character is in the valid set of a bio::alphabet (usually implies a bijective mappi...">bio::semialphabet::char_is_valid_for()</a> requirement via the <a class="el" href="group__alphabet.html#ga510b0657b8a3ae7d304999afd8fdcd52" title="Returns whether a character is in the valid set of a bio::alphabet (usually implies a bijective mappi...">bio::char_is_valid_for()</a> wrapper.</p>
<p>Behaviour specific to nucleotides: True also for lower case letters that silently convert to their upper case <b>and</b> true also for U/T respectively, e.g. 'U' is a valid character for <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">bio::dna4</a>, because its informational content is identical to 'T'.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a0e9f68ab92a2638509288eeb840aa2ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0e9f68ab92a2638509288eeb840aa2ec">&#9670;&nbsp;</a></span>complement()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a>  <a class="el" href="classbio_1_1nucleotide__base.html">bio::nucleotide_base</a>&lt; <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> , size &gt;::complement </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the complement of the letter. </p>
<p>See <a class="el" href="group__nucleotide.html">Nucleotide</a> for the actual values.</p>
<p>Provides an implementation for <a class="el" href="group__nucleotide.html#ga4b00de5b7d39cf416e62e14edef83f03" title="Return the complement of a nucleotide object.">bio::complement</a>, required to model <a class="el" href="interfacebio_1_1nucleotide__alphabet.html" title="A concept that indicates whether an alphabet represents nucleotides.">bio::nucleotide_alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a6616ba2cb25069f5192589ad53651a88"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6616ba2cb25069f5192589ad53651a88">&#9670;&nbsp;</a></span>to_char()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::to_char </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter as a character of char_type. </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8" title="Return the char representation of an alphabet object.">bio::to_char</a>, required to model <a class="el" href="interfacebio_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a443c2f7716e384ea2acec1d2bced7c8b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a443c2f7716e384ea2acec1d2bced7c8b">&#9670;&nbsp;</a></span>to_rank()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::to_rank </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter's numeric value (rank in the alphabet). </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0" title="Return the rank representation of a (semi-)alphabet object.">bio::to_rank</a>, required to model <a class="el" href="interfacebio_1_1semialphabet.html" title="The basis for bio::alphabet, but requires only rank interface (not char).">bio::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<h2 class="groupheader">Friends And Related Function Documentation</h2>
<a id="a6d873fe73b62acf41564b4dd6b6f7708"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6d873fe73b62acf41564b4dd6b6f7708">&#9670;&nbsp;</a></span>operator&quot;&quot;_dna3bs() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classbio_1_1dna3bs.html#a1c51e23a1a125a3a554a92eed266d911">dna3bs_vector</a> operator&quot;&quot;_dna3bs </td>
          <td>(</td>
          <td class="paramtype">char const *&#160;</td>
          <td class="paramname"><em>s</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/size_t.html">std::size_t</a>&#160;</td>
          <td class="paramname"><em>n</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">related</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">bio::dna3bs</a> string literal. </p>
<dl class="section return"><dt>Returns</dt><dd>bio::dna3bs_vector</dd></dl>
<p>You can use this string literal to easily assign to dna3bs_vector:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna3bs_8hpp.html">bio/alphabet/nucleotide/dna3bs.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna3bs;</div>
<div class="line"> </div>
<div class="line">    bio::dna3bs_vector foo{<span class="stringliteral">&quot;ACGTTA&quot;</span>_dna3bs}; <span class="comment">// ATGTTA</span></div>
<div class="line">    bio::dna3bs_vector bar = <span class="stringliteral">&quot;ATGTTA&quot;</span>_dna3bs;</div>
<div class="line"> </div>
<div class="line">    <span class="keywordflow">if</span> (foo == bar)</div>
<div class="line">        fmt::print(<span class="stringliteral">&quot;{}&quot;</span>, <span class="stringliteral">&quot;yeah\n&quot;</span>); <span class="comment">// &quot;yeah&quot;;</span></div>
<div class="line"> </div>
<div class="line">    <span class="keyword">auto</span> bax = <span class="stringliteral">&quot;ACGTTA&quot;</span>_dna3bs;</div>
<div class="line"> </div>
<div class="line">    <span class="keywordflow">return</span> 0;</div>
<div class="line">}</div>
</div><!-- fragment --> 
</div>
</div>
<a id="a1a688956c109d7472ce6aeeb7ac8b277"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1a688956c109d7472ce6aeeb7ac8b277">&#9670;&nbsp;</a></span>operator&quot;&quot;_dna3bs() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1dna3bs.html">dna3bs</a> operator&quot;&quot;_dna3bs </td>
          <td>(</td>
          <td class="paramtype">char const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">related</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The <a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">bio::dna3bs</a> char literal. </p>
<dl class="section return"><dt>Returns</dt><dd><a class="el" href="classbio_1_1dna3bs.html" title="The three letter reduced DNA alphabet for bisulfite sequencing mode (A,G,T(=C)).">bio::dna3bs</a> </dd></dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>bio/alphabet/nucleotide/<a class="el" href="dna3bs_8hpp_source.html">dna3bs.hpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="navelem"><a class="el" href="namespacebio.html">bio</a></li><li class="navelem"><a class="el" href="classbio_1_1dna3bs.html">dna3bs</a></li>
    <li class="footer">Generated on Sun Jun 12 2022 17:23:28 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
