<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: bio::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt; Class Template Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">0.1.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('classbio_1_1qualified.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-types">Public Types</a> &#124;
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#related">Related Functions</a> &#124;
<a href="classbio_1_1qualified-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">bio::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt; Class Template Reference<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a> &raquo; <a class="el" href="group__quality.html">Quality</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Joins an arbitrary alphabet with a quality alphabet.  
 <a href="classbio_1_1qualified.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="qualified_8hpp_source.html">bio/alphabet/quality/qualified.hpp</a>&gt;</code></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Inheritance diagram for bio::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><div class="zoom"><iframe scrolling="no" frameborder="0" src="classbio_1_1qualified__inherit__graph.svg" width="100%" height="600"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div></div>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-types"></a>
Public Types</h2></td></tr>
<tr class="memitem:a8a72615f8efe9eefd94b8fdbbf89a1f4"><td class="memItemLeft" align="right" valign="top"><a id="a8a72615f8efe9eefd94b8fdbbf89a1f4"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a8a72615f8efe9eefd94b8fdbbf89a1f4">char_type</a> = <a class="el" href="group__alphabet.html#ga474c9ea1cc2c81b65e22ee09c746fc17">alphabet_char_t</a>&lt; <a class="el" href="classbio_1_1qualified.html#aee5b2f1b2a140900eec8da2f080b01f0">sequence_alphabet_type</a> &gt;</td></tr>
<tr class="memdesc:a8a72615f8efe9eefd94b8fdbbf89a1f4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Equals the char_type of sequence_alphabet_type. <br /></td></tr>
<tr class="separator:a8a72615f8efe9eefd94b8fdbbf89a1f4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abad06ebdbf064d8a687c04a5904ff7ae"><td class="memItemLeft" align="right" valign="top"><a id="abad06ebdbf064d8a687c04a5904ff7ae"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#abad06ebdbf064d8a687c04a5904ff7ae">phred_type</a> = <a class="el" href="group__quality.html#gaec6fd89f46ef56795d66568184bdba0a">alphabet_phred_t</a>&lt; <a class="el" href="classbio_1_1qualified.html#afc3cbd30ec272acf49346c22ff29d0fe">quality_alphabet_type</a> &gt;</td></tr>
<tr class="memdesc:abad06ebdbf064d8a687c04a5904ff7ae"><td class="mdescLeft">&#160;</td><td class="mdescRight">Equals the phred_type of the quality_alphabet_type. <br /></td></tr>
<tr class="separator:abad06ebdbf064d8a687c04a5904ff7ae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc3cbd30ec272acf49346c22ff29d0fe"><td class="memItemLeft" align="right" valign="top"><a id="afc3cbd30ec272acf49346c22ff29d0fe"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#afc3cbd30ec272acf49346c22ff29d0fe">quality_alphabet_type</a> = quality_alphabet_t</td></tr>
<tr class="memdesc:afc3cbd30ec272acf49346c22ff29d0fe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Second template parameter as member type. <br /></td></tr>
<tr class="separator:afc3cbd30ec272acf49346c22ff29d0fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aee5b2f1b2a140900eec8da2f080b01f0"><td class="memItemLeft" align="right" valign="top"><a id="aee5b2f1b2a140900eec8da2f080b01f0"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#aee5b2f1b2a140900eec8da2f080b01f0">sequence_alphabet_type</a> = sequence_alphabet_t</td></tr>
<tr class="memdesc:aee5b2f1b2a140900eec8da2f080b01f0"><td class="mdescLeft">&#160;</td><td class="mdescRight">First template parameter as member type. <br /></td></tr>
<tr class="separator:aee5b2f1b2a140900eec8da2f080b01f0"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Constructors, destructor and assignment</div></td></tr>
<tr class="memitem:afa2f08cdcee04786c9d0ce2c9bd7ac53"><td class="memItemLeft" align="right" valign="top"><a id="afa2f08cdcee04786c9d0ce2c9bd7ac53"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#afa2f08cdcee04786c9d0ce2c9bd7ac53">qualified</a> () noexcept=default</td></tr>
<tr class="memdesc:afa2f08cdcee04786c9d0ce2c9bd7ac53"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:afa2f08cdcee04786c9d0ce2c9bd7ac53"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a63823a76f653799668572341f6ffe584"><td class="memItemLeft" align="right" valign="top"><a id="a63823a76f653799668572341f6ffe584"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a63823a76f653799668572341f6ffe584">qualified</a> (<a class="el" href="classbio_1_1qualified.html">qualified</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a63823a76f653799668572341f6ffe584"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a63823a76f653799668572341f6ffe584"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa119859d627c6c8101e3b52f07237da8"><td class="memItemLeft" align="right" valign="top"><a id="aa119859d627c6c8101e3b52f07237da8"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#aa119859d627c6c8101e3b52f07237da8">qualified</a> (<a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:aa119859d627c6c8101e3b52f07237da8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:aa119859d627c6c8101e3b52f07237da8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7eab90c38ab064ed565387cd4fb5f584"><td class="memItemLeft" align="right" valign="top"><a id="a7eab90c38ab064ed565387cd4fb5f584"></a>
constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a7eab90c38ab064ed565387cd4fb5f584">operator=</a> (<a class="el" href="classbio_1_1qualified.html">qualified</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a7eab90c38ab064ed565387cd4fb5f584"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a7eab90c38ab064ed565387cd4fb5f584"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aba16bc4f0a9e3741f06ac50f3a9338d9"><td class="memItemLeft" align="right" valign="top"><a id="aba16bc4f0a9e3741f06ac50f3a9338d9"></a>
constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#aba16bc4f0a9e3741f06ac50f3a9338d9">operator=</a> (<a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:aba16bc4f0a9e3741f06ac50f3a9338d9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:aba16bc4f0a9e3741f06ac50f3a9338d9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9606f1e64d27eb8c14b6a476ec8ef471"><td class="memItemLeft" align="right" valign="top"><a id="a9606f1e64d27eb8c14b6a476ec8ef471"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a9606f1e64d27eb8c14b6a476ec8ef471">~qualified</a> () noexcept=default</td></tr>
<tr class="memdesc:a9606f1e64d27eb8c14b6a476ec8ef471"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a9606f1e64d27eb8c14b6a476ec8ef471"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a102f4110dcd9618f8522a8b5eff16ffb"><td class="memItemLeft" align="right" valign="top">constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a102f4110dcd9618f8522a8b5eff16ffb">qualified</a> (component_type const alph) noexcept</td></tr>
<tr class="memdesc:a102f4110dcd9618f8522a8b5eff16ffb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Construction via a value of one of the components.  <a href="classbio_1_1qualified.html#a102f4110dcd9618f8522a8b5eff16ffb">More...</a><br /></td></tr>
<tr class="separator:a102f4110dcd9618f8522a8b5eff16ffb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a70ee86132b4f9183f5fa31e1c65d3d47"><td class="memItemLeft" align="right" valign="top">constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a70ee86132b4f9183f5fa31e1c65d3d47">qualified</a> (indirect_component_type const alph) noexcept</td></tr>
<tr class="memdesc:a70ee86132b4f9183f5fa31e1c65d3d47"><td class="mdescLeft">&#160;</td><td class="mdescRight">Construction via a value of a subtype that is assignable to one of the components.  <a href="classbio_1_1qualified.html#a70ee86132b4f9183f5fa31e1c65d3d47">More...</a><br /></td></tr>
<tr class="separator:a70ee86132b4f9183f5fa31e1c65d3d47"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1c3aeab9107551898ef8669a49275223"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a1c3aeab9107551898ef8669a49275223">operator=</a> (component_type const alph) noexcept</td></tr>
<tr class="memdesc:a1c3aeab9107551898ef8669a49275223"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assignment via a value of one of the components.  <a href="classbio_1_1qualified.html#a1c3aeab9107551898ef8669a49275223">More...</a><br /></td></tr>
<tr class="separator:a1c3aeab9107551898ef8669a49275223"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab7c5301e4b65a8db3539b19e641bd9a1"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#ab7c5301e4b65a8db3539b19e641bd9a1">operator=</a> (indirect_component_type const alph) noexcept</td></tr>
<tr class="memdesc:ab7c5301e4b65a8db3539b19e641bd9a1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assignment via a value of a subtype that is assignable to one of the components.  <a href="classbio_1_1qualified.html#ab7c5301e4b65a8db3539b19e641bd9a1">More...</a><br /></td></tr>
<tr class="separator:ab7c5301e4b65a8db3539b19e641bd9a1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Write functions</div></td></tr>
<tr class="memitem:a3a92eaf9e8907d3371dd0a8a26f4c944"><td class="memItemLeft" align="right" valign="top"><a id="a3a92eaf9e8907d3371dd0a8a26f4c944"></a>
constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a3a92eaf9e8907d3371dd0a8a26f4c944">assign_char</a> (<a class="el" href="classbio_1_1qualified.html#a8a72615f8efe9eefd94b8fdbbf89a1f4">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a3a92eaf9e8907d3371dd0a8a26f4c944"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a character. This modifies the internal sequence letter. <br /></td></tr>
<tr class="separator:a3a92eaf9e8907d3371dd0a8a26f4c944"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac45db81a51a37b8b0c1b79cb128542a1"><td class="memItemLeft" align="right" valign="top"><a id="ac45db81a51a37b8b0c1b79cb128542a1"></a>
constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#ac45db81a51a37b8b0c1b79cb128542a1">assign_phred</a> (<a class="el" href="classbio_1_1qualified.html#abad06ebdbf064d8a687c04a5904ff7ae">phred_type</a> const c) noexcept</td></tr>
<tr class="memdesc:ac45db81a51a37b8b0c1b79cb128542a1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a phred value. This modifies the internal quality letter. <br /></td></tr>
<tr class="separator:ac45db81a51a37b8b0c1b79cb128542a1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a0d830a0a03683269b0d71692ae079236"><td class="memItemLeft" align="right" valign="top"><a id="a0d830a0a03683269b0d71692ae079236"></a>
constexpr <a class="el" href="classbio_1_1qualified.html#abad06ebdbf064d8a687c04a5904ff7ae">phred_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a0d830a0a03683269b0d71692ae079236">to_phred</a> () const noexcept</td></tr>
<tr class="memdesc:a0d830a0a03683269b0d71692ae079236"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the phred value. This reads the internal quality letter. <br /></td></tr>
<tr class="separator:a0d830a0a03683269b0d71692ae079236"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1b2364e5921c2a637540743462b3b5d8"><td class="memItemLeft" align="right" valign="top"><a id="a1b2364e5921c2a637540743462b3b5d8"></a>
constexpr <a class="el" href="classbio_1_1qualified.html#a8a72615f8efe9eefd94b8fdbbf89a1f4">char_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a1b2364e5921c2a637540743462b3b5d8">to_char</a> () const noexcept</td></tr>
<tr class="memdesc:a1b2364e5921c2a637540743462b3b5d8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a character. This reads the internal sequence letter. <br /></td></tr>
<tr class="separator:a1b2364e5921c2a637540743462b3b5d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af210bb1f34ba0adff5c5b798bcfc270d"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#af210bb1f34ba0adff5c5b798bcfc270d">complement</a> () const noexcept requires <a class="el" href="interfacebio_1_1nucleotide__alphabet.html">nucleotide_alphabet</a>&lt; sequence_alphabet_t &gt;</td></tr>
<tr class="memdesc:af210bb1f34ba0adff5c5b798bcfc270d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return a qualified where the quality is preserved, but the sequence letter is complemented.  <a href="classbio_1_1qualified.html#af210bb1f34ba0adff5c5b798bcfc270d">More...</a><br /></td></tr>
<tr class="separator:af210bb1f34ba0adff5c5b798bcfc270d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr><td colspan="2"><div class="groupText"><p>All read operations are constant complexity. </p>
</div></td></tr>
<tr class="memitem:ac09202f0f79dfb23ca37cd097e505727"><td class="memItemLeft" align="right" valign="top"><a id="ac09202f0f79dfb23ca37cd097e505727"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__tuple__base.html#ac09202f0f79dfb23ca37cd097e505727">operator type</a> () const noexcept</td></tr>
<tr class="memdesc:ac09202f0f79dfb23ca37cd097e505727"><td class="mdescLeft">&#160;</td><td class="mdescRight">Implicit cast to a single letter. Works only if the type is unique in the type list. <br /></td></tr>
<tr class="separator:ac09202f0f79dfb23ca37cd097e505727"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a443c2f7716e384ea2acec1d2bced7c8b"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a77c05b494c8fecc9b1fce43f5298a329">rank_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">to_rank</a> () const noexcept</td></tr>
<tr class="memdesc:a443c2f7716e384ea2acec1d2bced7c8b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter's numeric value (rank in the alphabet).  <a href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">More...</a><br /></td></tr>
<tr class="separator:a443c2f7716e384ea2acec1d2bced7c8b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Write functions</div></td></tr>
<tr class="memitem:a822d399ab808842870a4a4feb5490849"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">assign_char</a> (<a class="el" href="classbio_1_1qualified.html#a8a72615f8efe9eefd94b8fdbbf89a1f4">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a822d399ab808842870a4a4feb5490849"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a character, implicitly converts invalid characters.  <a href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">More...</a><br /></td></tr>
<tr class="separator:a822d399ab808842870a4a4feb5490849"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acdd042ae2f04d986f65a7b60edd6e83e"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">assign_rank</a> (<a class="el" href="classbio_1_1alphabet__base.html#a77c05b494c8fecc9b1fce43f5298a329">rank_type</a> const c) noexcept</td></tr>
<tr class="memdesc:acdd042ae2f04d986f65a7b60edd6e83e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a numeric value.  <a href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">More...</a><br /></td></tr>
<tr class="separator:acdd042ae2f04d986f65a7b60edd6e83e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:a760d1ffca88905c051931d6e2de9810e"><td class="memItemLeft" align="right" valign="top"><a id="a760d1ffca88905c051931d6e2de9810e"></a>
static constexpr bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a760d1ffca88905c051931d6e2de9810e">char_is_valid</a> (<a class="el" href="classbio_1_1qualified.html#a8a72615f8efe9eefd94b8fdbbf89a1f4">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a760d1ffca88905c051931d6e2de9810e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Validate whether a character is valid in the sequence alphabet. <br /></td></tr>
<tr class="separator:a760d1ffca88905c051931d6e2de9810e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:af5c69d209add96d96b2c5a5dce3d508c"><td class="memItemLeft" align="right" valign="top"><a id="af5c69d209add96d96b2c5a5dce3d508c"></a>
static constexpr <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">meta::detail::min_viable_uint_t</a>&lt; size &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#af5c69d209add96d96b2c5a5dce3d508c">alphabet_size</a></td></tr>
<tr class="memdesc:af5c69d209add96d96b2c5a5dce3d508c"><td class="mdescLeft">&#160;</td><td class="mdescRight">The size of the alphabet, i.e. the number of different values it can take. <br /></td></tr>
<tr class="separator:af5c69d209add96d96b2c5a5dce3d508c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-types"></a>
Protected Types</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Member types</div></td></tr>
<tr class="memitem:a77c05b494c8fecc9b1fce43f5298a329"><td class="memItemLeft" align="right" valign="top"><a id="a77c05b494c8fecc9b1fce43f5298a329"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a77c05b494c8fecc9b1fce43f5298a329">rank_type</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">meta::detail::min_viable_uint_t</a>&lt; size - 1 &gt;</td></tr>
<tr class="memdesc:a77c05b494c8fecc9b1fce43f5298a329"><td class="mdescLeft">&#160;</td><td class="mdescRight">The type of the alphabet when represented as a number (e.g. via <a class="el" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b" title="Return the letter&#39;s numeric value (rank in the alphabet).">to_rank()</a>). <br /></td></tr>
<tr class="separator:a77c05b494c8fecc9b1fce43f5298a329"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="related"></a>
Related Functions</h2></td></tr>
<tr><td class="ititle" colspan="2"><p>(Note that these are not member functions.) </p>
</td></tr>
<tr class="memitem:a766054fe09be32310d2929f2460f857c"><td class="memTemplParams" colspan="2"><a id="a766054fe09be32310d2929f2460f857c"></a>
template&lt;typename sequence_alphabet_type , typename quality_alphabet_type &gt; </td></tr>
<tr class="memitem:a766054fe09be32310d2929f2460f857c"><td class="memTemplItemLeft" align="right" valign="top">&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html#a766054fe09be32310d2929f2460f857c">qualified</a> (<a class="el" href="classbio_1_1qualified.html#aee5b2f1b2a140900eec8da2f080b01f0">sequence_alphabet_type</a> &amp;&amp;, <a class="el" href="classbio_1_1qualified.html#afc3cbd30ec272acf49346c22ff29d0fe">quality_alphabet_type</a> &amp;&amp;) -&gt; <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/decay.html">std::decay_t</a>&lt; <a class="el" href="classbio_1_1qualified.html#aee5b2f1b2a140900eec8da2f080b01f0">sequence_alphabet_type</a> &gt;, <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/decay.html">std::decay_t</a>&lt; <a class="el" href="classbio_1_1qualified.html#afc3cbd30ec272acf49346c22ff29d0fe">quality_alphabet_type</a> &gt;&gt;</td></tr>
<tr class="memdesc:a766054fe09be32310d2929f2460f857c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Type deduction guide enables usage of qualified without specifying template args. <br /></td></tr>
<tr class="separator:a766054fe09be32310d2929f2460f857c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><h3>template&lt;writable_alphabet sequence_alphabet_t, writable_quality_alphabet quality_alphabet_t&gt;<br />
class bio::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</h3>

<p>Joins an arbitrary alphabet with a quality alphabet. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">sequence_alphabet_t</td><td>Type of the alphabet; must satisfy <a class="el" href="interfacebio_1_1writable__alphabet.html" title="Refines bio::alphabet and adds assignability.">bio::writable_alphabet</a>. </td></tr>
    <tr><td class="paramname">quality_alphabet_t</td><td>Type of the quality; must satisfy <a class="el" href="interfacebio_1_1writable__quality__alphabet.html" title="A concept that indicates whether a writable alphabet represents quality scores.">bio::writable_quality_alphabet</a>.</td></tr>
  </table>
  </dd>
</dl>
<p>This composite pairs an arbitrary alphabet with a quality alphabet, where each alphabet character is stored together with its quality score in a single value. That way, you can can conveniently access the character and score information at each position of the qualified-sequence. The use case that this was designed for is a nucleotide sequence with corresponding quality scores, e.g. obtained when reading in a FASTQ file of Illumina reads. The composite also allows to store quality scores for different or extended alphabets like a <code>qualified&lt;char, <a class="el" href="classbio_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range).">phred42</a>&gt;</code> or a <code>qualified&lt;gapped&lt;<a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">dna4</a>&gt;, <a class="el" href="classbio_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range).">phred42</a>&gt;</code> sequence. The rank values correspond to numeric values in the size of the composite, while the character values are taken from the sequence alphabet and the phred values are taken from the quality alphabet.</p>
<p>As with all <code><a class="el" href="classbio_1_1alphabet__tuple__base.html" title="The CRTP base for a combined alphabet that contains multiple values of different alphabets at the sam...">bio::alphabet_tuple_base</a></code> s you may access the individual alphabet letters in regular c++ tuple notation, i.e. <code>get&lt;0&gt;(t)</code> and objects can be brace-initialised with the individual members.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="qualified_8hpp.html">bio/alphabet/quality/qualified.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="phred42_8hpp.html">bio/alphabet/quality/phred42.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna4_8hpp.html">bio/alphabet/nucleotide/dna4.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna4;</div>
<div class="line"> </div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter{<span class="charliteral">&#39;A&#39;</span>_dna4, <a class="code" href="classbio_1_1phred42.html">bio::phred42</a>{7}};</div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{} {} {}\n&quot;</span>,</div>
<div class="line">               <a class="code" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0">bio::to_rank</a>(letter),</div>
<div class="line">               <a class="code" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0">bio::to_rank</a>(get&lt;0&gt;(letter)),</div>
<div class="line">               <a class="code" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0">bio::to_rank</a>(get&lt;1&gt;(letter)));</div>
<div class="line">    <span class="comment">// 28 0 7</span></div>
<div class="line"> </div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{} {} {}\n&quot;</span>,</div>
<div class="line">               <a class="code" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8">bio::to_char</a>(letter),</div>
<div class="line">               <a class="code" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8">bio::to_char</a>(get&lt;0&gt;(letter)),</div>
<div class="line">               <a class="code" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8">bio::to_char</a>(get&lt;1&gt;(letter)));</div>
<div class="line">    <span class="comment">// A A (</span></div>
<div class="line"> </div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{} {}\n&quot;</span>, <a class="code" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445">bio::to_phred</a>(letter), <a class="code" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445">bio::to_phred</a>(get&lt;1&gt;(letter)));</div>
<div class="line">    <span class="comment">// 7 7</span></div>
<div class="line"> </div>
<div class="line">    <span class="comment">// modify via structured bindings and references:</span></div>
<div class="line">    <span class="keyword">auto</span> &amp; [ seq_l, qual_l ] = letter;</div>
<div class="line">    seq_l = <span class="charliteral">&#39;G&#39;</span>_dna4;</div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{}\n&quot;</span>, <a class="code" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8">bio::to_char</a>(letter));</div>
<div class="line">    <span class="comment">// G</span></div>
<div class="line">}</div>
<div class="ttc" id="aclassbio_1_1phred42_html"><div class="ttname"><a href="classbio_1_1phred42.html">bio::phred42</a></div><div class="ttdoc">Quality type for traditional Sanger and modern Illumina Phred scores (typical range).</div><div class="ttdef"><b>Definition:</b> phred42.hpp:45</div></div>
<div class="ttc" id="aclassbio_1_1qualified_html"><div class="ttname"><a href="classbio_1_1qualified.html">bio::qualified</a></div><div class="ttdoc">Joins an arbitrary alphabet with a quality alphabet.</div><div class="ttdef"><b>Definition:</b> qualified.hpp:61</div></div>
<div class="ttc" id="adna4_8hpp_html"><div class="ttname"><a href="dna4_8hpp.html">dna4.hpp</a></div><div class="ttdoc">Provides bio::dna4, container aliases and string literals.</div></div>
<div class="ttc" id="afmt_8hpp_html"><div class="ttname"><a href="fmt_8hpp.html">fmt.hpp</a></div><div class="ttdoc">Core alphabet concept and free function/type trait wrappers.</div></div>
<div class="ttc" id="agroup__alphabet_html_ga9189b35f800d7e5abad51e021ea811a0"><div class="ttname"><a href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0">bio::to_rank</a></div><div class="ttdeci">constexpr auto to_rank</div><div class="ttdoc">Return the rank representation of a (semi-)alphabet object.</div><div class="ttdef"><b>Definition:</b> concept.hpp:140</div></div>
<div class="ttc" id="agroup__alphabet_html_gab295f04aea840000aafc640b5e6528c8"><div class="ttname"><a href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8">bio::to_char</a></div><div class="ttdeci">constexpr auto to_char</div><div class="ttdoc">Return the char representation of an alphabet object.</div><div class="ttdef"><b>Definition:</b> concept.hpp:316</div></div>
<div class="ttc" id="agroup__quality_html_ga0a0a566ce2f93ff9cb4f556b0a3a2445"><div class="ttname"><a href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445">bio::to_phred</a></div><div class="ttdeci">constexpr auto to_phred</div><div class="ttdoc">The public getter function for the phred representation of a quality score.</div><div class="ttdef"><b>Definition:</b> concept.hpp:90</div></div>
<div class="ttc" id="aphred42_8hpp_html"><div class="ttname"><a href="phred42_8hpp.html">phred42.hpp</a></div><div class="ttdoc">Provides bio::phred42 quality scores.</div></div>
<div class="ttc" id="aqualified_8hpp_html"><div class="ttname"><a href="qualified_8hpp.html">qualified.hpp</a></div><div class="ttdoc">Provides quality alphabet composites.</div></div>
</div><!-- fragment --><p>This <a class="el" href="classbio_1_1alphabet__tuple__base.html" title="The CRTP base for a combined alphabet that contains multiple values of different alphabets at the sam...">bio::alphabet_tuple_base</a> itself fulfils both <a class="el" href="interfacebio_1_1writable__alphabet.html" title="Refines bio::alphabet and adds assignability.">bio::writable_alphabet</a> and <a class="el" href="interfacebio_1_1writable__quality__alphabet.html" title="A concept that indicates whether a writable alphabet represents quality scores.">bio::writable_quality_alphabet</a>. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a id="a102f4110dcd9618f8522a8b5eff16ffb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a102f4110dcd9618f8522a8b5eff16ffb">&#9670;&nbsp;</a></span>qualified() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;writable_alphabet sequence_alphabet_t, writable_quality_alphabet quality_alphabet_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">bio::qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt;::<a class="el" href="classbio_1_1qualified.html">qualified</a> </td>
          <td>(</td>
          <td class="paramtype">component_type const&#160;</td>
          <td class="paramname"><em>alph</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Construction via a value of one of the components. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">component_type</td><td>Must be one uniquely contained in the type list of the composite. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alph</td><td>The value of a component that should be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Note: Since the <a class="el" href="classbio_1_1alphabet__tuple__base.html" title="The CRTP base for a combined alphabet that contains multiple values of different alphabets at the sam...">alphabet_tuple_base</a> is a CRTP base class, we show the working examples with one of its derived classes (<a class="el" href="classbio_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::qualified</a>). </p><div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna4_8hpp.html">bio/alphabet/nucleotide/dna4.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="phred42_8hpp.html">bio/alphabet/quality/phred42.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="qualified_8hpp.html">bio/alphabet/quality/qualified.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna4;</div>
<div class="line"> </div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter1{<span class="charliteral">&#39;C&#39;</span>_dna4};</div>
<div class="line">    <span class="comment">// creates {&#39;C&#39;_dna4, bio::phred42{0}}</span></div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter2{<a class="code" href="classbio_1_1phred42.html">bio::phred42</a>{1}};</div>
<div class="line">    <span class="comment">// creates {&#39;A&#39;_dna4, bio::phred42{1}}</span></div>
<div class="line"> </div>
<div class="line">    <span class="keywordflow">if</span> (letter1 == letter2)</div>
<div class="line">        fmt::print(<span class="stringliteral">&quot;{}&quot;</span>, <span class="stringliteral">&quot;yeah\n&quot;</span>); <span class="comment">// yeah</span></div>
<div class="line">}</div>
</div><!-- fragment --> 
</div>
</div>
<a id="a70ee86132b4f9183f5fa31e1c65d3d47"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a70ee86132b4f9183f5fa31e1c65d3d47">&#9670;&nbsp;</a></span>qualified() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;writable_alphabet sequence_alphabet_t, writable_quality_alphabet quality_alphabet_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">bio::qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt;::<a class="el" href="classbio_1_1qualified.html">qualified</a> </td>
          <td>(</td>
          <td class="paramtype">indirect_component_type const&#160;</td>
          <td class="paramname"><em>alph</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Construction via a value of a subtype that is assignable to one of the components. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">indirect_component_type</td><td>Type that models <a class="el" href="interfacebio_1_1weakly__assignable__from.html" title="Resolves to std::is_assignable_v&lt;t&gt;.">bio::weakly_assignable_from</a> for one of the component types. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alph</td><td>The value that should be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Note that the value will be assigned to the <b>FIRST</b> type T that fulfils <code>assignable_from&lt;T, indirect_component_type&gt;</code>, regardless if other types are also fit for assignment.</p>
<p>Note: Since the <a class="el" href="classbio_1_1alphabet__tuple__base.html" title="The CRTP base for a combined alphabet that contains multiple values of different alphabets at the sam...">alphabet_tuple_base</a> is a CRTP base class, we show the working examples with one of its derived classes (<a class="el" href="classbio_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::qualified</a>). </p><div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="alphabet_2nucleotide_2all_8hpp.html">bio/alphabet/nucleotide/all.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="phred42_8hpp.html">bio/alphabet/quality/phred42.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="qualified_8hpp.html">bio/alphabet/quality/qualified.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna4;</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_rna4;</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// The following creates {&#39;C&#39;_dna4, phred42{0}}</span></div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter1{<span class="charliteral">&#39;C&#39;</span>_dna4};</div>
<div class="line">    <span class="comment">// The following also creates {&#39;C&#39;_dna4, bio::phred42{0}}, since rna4 assignable to dna4</span></div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter2{<span class="charliteral">&#39;C&#39;</span>_rna4};</div>
<div class="line"> </div>
<div class="line">    <span class="keywordflow">if</span> (letter1 == letter2)</div>
<div class="line">        fmt::print(<span class="stringliteral">&quot;{}&quot;</span>, <span class="stringliteral">&quot;yeah\n&quot;</span>); <span class="comment">// yeah</span></div>
<div class="line">}</div>
<div class="ttc" id="aalphabet_2nucleotide_2all_8hpp_html"><div class="ttname"><a href="alphabet_2nucleotide_2all_8hpp.html">all.hpp</a></div><div class="ttdoc">Meta-header for the nucleotide submodule; includes all headers from alphabet/nucleotide/.</div></div>
</div><!-- fragment --> 
</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a id="a822d399ab808842870a4a4feb5490849"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a822d399ab808842870a4a4feb5490849">&#9670;&nbsp;</a></span>assign_char()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; &amp; <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; , size, void  &gt;::assign_char </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a character, implicitly converts invalid characters. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The character to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga05da42eb0b9a48f4d1a554c7d3a9044d" title="Assign a character to an alphabet object.">bio::assign_char_to</a>, required to model <a class="el" href="interfacebio_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="acdd042ae2f04d986f65a7b60edd6e83e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acdd042ae2f04d986f65a7b60edd6e83e">&#9670;&nbsp;</a></span>assign_rank()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; &amp; <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; , size, void  &gt;::assign_rank </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a77c05b494c8fecc9b1fce43f5298a329">rank_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a numeric value. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The rank to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#gae34fa2b37a8b40f2f2fa64fed7bbb410" title="Assign a rank to an alphabet object.">bio::assign_rank_to</a>, required to model <a class="el" href="interfacebio_1_1semialphabet.html" title="The basis for bio::alphabet, but requires only rank interface (not char).">bio::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="af210bb1f34ba0adff5c5b798bcfc270d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af210bb1f34ba0adff5c5b798bcfc270d">&#9670;&nbsp;</a></span>complement()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;writable_alphabet sequence_alphabet_t, writable_quality_alphabet quality_alphabet_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a> <a class="el" href="classbio_1_1qualified.html">bio::qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt;::complement </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return a qualified where the quality is preserved, but the sequence letter is complemented. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__nucleotide.html#ga4b00de5b7d39cf416e62e14edef83f03" title="Return the complement of a nucleotide object.">bio::complement</a> </dd>
<dd>
bio::nucleotide_alphabet::complement </dd></dl>

</div>
</div>
<a id="a1c3aeab9107551898ef8669a49275223"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a1c3aeab9107551898ef8669a49275223">&#9670;&nbsp;</a></span>operator=() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;writable_alphabet sequence_alphabet_t, writable_quality_alphabet quality_alphabet_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&amp; <a class="el" href="classbio_1_1qualified.html">bio::qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt;::operator= </td>
          <td>(</td>
          <td class="paramtype">component_type const&#160;</td>
          <td class="paramname"><em>alph</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assignment via a value of one of the components. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">component_type</td><td>One of the component types. Must be uniquely contained in the type list of the composite. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alph</td><td>The value of a component that should be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Note: Since the <a class="el" href="classbio_1_1alphabet__tuple__base.html" title="The CRTP base for a combined alphabet that contains multiple values of different alphabets at the sam...">alphabet_tuple_base</a> is a CRTP base class, we show the working examples with one of its derived classes (<a class="el" href="classbio_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::qualified</a>). </p><div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="dna4_8hpp.html">bio/alphabet/nucleotide/dna4.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="phred42_8hpp.html">bio/alphabet/quality/phred42.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="qualified_8hpp.html">bio/alphabet/quality/qualified.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna4;</div>
<div class="line"> </div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter1{<span class="charliteral">&#39;T&#39;</span>_dna4, <a class="code" href="classbio_1_1phred42.html">bio::phred42</a>{1}};</div>
<div class="line"> </div>
<div class="line">    letter1 = <span class="charliteral">&#39;C&#39;</span>_dna4;           <span class="comment">// yields {&#39;C&#39;_dna4, bio::phred42{1}}</span></div>
<div class="line">    letter1 = <a class="code" href="classbio_1_1phred42.html">bio::phred42</a>{2}; <span class="comment">// yields {&#39;C&#39;_dna4, bio::phred42{2}}</span></div>
<div class="line">}</div>
</div><!-- fragment --> 
</div>
</div>
<a id="ab7c5301e4b65a8db3539b19e641bd9a1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ab7c5301e4b65a8db3539b19e641bd9a1">&#9670;&nbsp;</a></span>operator=() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;writable_alphabet sequence_alphabet_t, writable_quality_alphabet quality_alphabet_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1qualified.html">qualified</a>&amp; <a class="el" href="classbio_1_1qualified.html">bio::qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt;::operator= </td>
          <td>(</td>
          <td class="paramtype">indirect_component_type const&#160;</td>
          <td class="paramname"><em>alph</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assignment via a value of a subtype that is assignable to one of the components. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">indirect_component_type</td><td>Type that models <a class="el" href="interfacebio_1_1weakly__assignable__from.html" title="Resolves to std::is_assignable_v&lt;t&gt;.">bio::weakly_assignable_from</a> for one of the component types. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alph</td><td>The value of a component that should be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Note: Since the <a class="el" href="classbio_1_1alphabet__tuple__base.html" title="The CRTP base for a combined alphabet that contains multiple values of different alphabets at the sam...">alphabet_tuple_base</a> is a CRTP base class, we show the working examples with one of its derived classes (<a class="el" href="classbio_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::qualified</a>). </p><div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="alphabet_2nucleotide_2all_8hpp.html">bio/alphabet/nucleotide/all.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="phred42_8hpp.html">bio/alphabet/quality/phred42.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="qualified_8hpp.html">bio/alphabet/quality/qualified.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna4;</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_rna4;</div>
<div class="line"> </div>
<div class="line">    <a class="code" href="classbio_1_1qualified.html">bio::qualified&lt;bio::dna4, bio::phred42&gt;</a> letter1{<span class="charliteral">&#39;T&#39;</span>_dna4, <a class="code" href="classbio_1_1phred42.html">bio::phred42</a>{1}};</div>
<div class="line"> </div>
<div class="line">    letter1 = <span class="charliteral">&#39;C&#39;</span>_rna4; <span class="comment">// yields {&#39;C&#39;_dna4, bio::phred42{1}}</span></div>
<div class="line">}</div>
</div><!-- fragment --> 
</div>
</div>
<a id="a443c2f7716e384ea2acec1d2bced7c8b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a443c2f7716e384ea2acec1d2bced7c8b">&#9670;&nbsp;</a></span>to_rank()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a77c05b494c8fecc9b1fce43f5298a329">rank_type</a> <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; <a class="el" href="classbio_1_1qualified.html">qualified</a>&lt; sequence_alphabet_t, quality_alphabet_t &gt; , size, void  &gt;::to_rank </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter's numeric value (rank in the alphabet). </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0" title="Return the rank representation of a (semi-)alphabet object.">bio::to_rank</a>, required to model <a class="el" href="interfacebio_1_1semialphabet.html" title="The basis for bio::alphabet, but requires only rank interface (not char).">bio::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>bio/alphabet/quality/<a class="el" href="qualified_8hpp_source.html">qualified.hpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="navelem"><a class="el" href="namespacebio.html">bio</a></li><li class="navelem"><a class="el" href="classbio_1_1qualified.html">qualified</a></li>
    <li class="footer">Generated on Mon Jul 4 2022 17:20:01 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
