<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: bio::sam_dna16 Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">0.1.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('classbio_1_1sam__dna16.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-static-methods">Static Public Member Functions</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#related">Related Functions</a> &#124;
<a href="classbio_1_1sam__dna16-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">bio::sam_dna16 Class Reference<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a> &raquo; <a class="el" href="group__nucleotide.html">Nucleotide</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=').  
 <a href="classbio_1_1sam__dna16.html#details">More...</a></p>

<p><code>#include &lt;<a class="el" href="sam__dna16_8hpp_source.html">bio/alphabet/nucleotide/sam_dna16.hpp</a>&gt;</code></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Inheritance diagram for bio::sam_dna16:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><div class="zoom"><iframe scrolling="no" frameborder="0" src="classbio_1_1sam__dna16__inherit__graph.svg" width="100%" height="517"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div></div>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Constructors, destructor and assignment</div></td></tr>
<tr class="memitem:ad90a57de2388bd1dad8c3a43b21cf08f"><td class="memItemLeft" align="right" valign="top"><a id="ad90a57de2388bd1dad8c3a43b21cf08f"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#ad90a57de2388bd1dad8c3a43b21cf08f">sam_dna16</a> () noexcept=default</td></tr>
<tr class="memdesc:ad90a57de2388bd1dad8c3a43b21cf08f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:ad90a57de2388bd1dad8c3a43b21cf08f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76d30ed01c6e416317b756d1d024d5b7"><td class="memItemLeft" align="right" valign="top"><a id="a76d30ed01c6e416317b756d1d024d5b7"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#a76d30ed01c6e416317b756d1d024d5b7">sam_dna16</a> (<a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a76d30ed01c6e416317b756d1d024d5b7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a76d30ed01c6e416317b756d1d024d5b7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9dd7accb146c3cbd665a4331efb272d2"><td class="memItemLeft" align="right" valign="top"><a id="a9dd7accb146c3cbd665a4331efb272d2"></a>
constexpr&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#a9dd7accb146c3cbd665a4331efb272d2">sam_dna16</a> (<a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:a9dd7accb146c3cbd665a4331efb272d2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a9dd7accb146c3cbd665a4331efb272d2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a535794e258d71f2dcbb8b2e845d253e5"><td class="memItemLeft" align="right" valign="top"><a id="a535794e258d71f2dcbb8b2e845d253e5"></a>
constexpr <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#a535794e258d71f2dcbb8b2e845d253e5">operator=</a> (<a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> const &amp;) noexcept=default</td></tr>
<tr class="memdesc:a535794e258d71f2dcbb8b2e845d253e5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:a535794e258d71f2dcbb8b2e845d253e5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab848d8ac8840918072f92d8f4f344154"><td class="memItemLeft" align="right" valign="top"><a id="ab848d8ac8840918072f92d8f4f344154"></a>
constexpr <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#ab848d8ac8840918072f92d8f4f344154">operator=</a> (<a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> &amp;&amp;) noexcept=default</td></tr>
<tr class="memdesc:ab848d8ac8840918072f92d8f4f344154"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:ab848d8ac8840918072f92d8f4f344154"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aea395d447a3c900e0d9d979904eba81c"><td class="memItemLeft" align="right" valign="top"><a id="aea395d447a3c900e0d9d979904eba81c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#aea395d447a3c900e0d9d979904eba81c">~sam_dna16</a> () noexcept=default</td></tr>
<tr class="memdesc:aea395d447a3c900e0d9d979904eba81c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Defaulted. <br /></td></tr>
<tr class="separator:aea395d447a3c900e0d9d979904eba81c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a0e9f68ab92a2638509288eeb840aa2ec"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1nucleotide__base.html#a0e9f68ab92a2638509288eeb840aa2ec">complement</a> () const noexcept</td></tr>
<tr class="memdesc:a0e9f68ab92a2638509288eeb840aa2ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the complement of the letter.  <a href="classbio_1_1nucleotide__base.html#a0e9f68ab92a2638509288eeb840aa2ec">More...</a><br /></td></tr>
<tr class="separator:a0e9f68ab92a2638509288eeb840aa2ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Read functions</div></td></tr>
<tr class="memitem:a6616ba2cb25069f5192589ad53651a88"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">to_char</a> () const noexcept</td></tr>
<tr class="memdesc:a6616ba2cb25069f5192589ad53651a88"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter as a character of char_type.  <a href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">More...</a><br /></td></tr>
<tr class="separator:a6616ba2cb25069f5192589ad53651a88"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a443c2f7716e384ea2acec1d2bced7c8b"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">to_rank</a> () const noexcept</td></tr>
<tr class="memdesc:a443c2f7716e384ea2acec1d2bced7c8b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the letter's numeric value (rank in the alphabet).  <a href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">More...</a><br /></td></tr>
<tr class="separator:a443c2f7716e384ea2acec1d2bced7c8b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Write functions</div></td></tr>
<tr class="memitem:a822d399ab808842870a4a4feb5490849"><td class="memItemLeft" align="right" valign="top">constexpr derived_type &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">assign_char</a> (<a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:a822d399ab808842870a4a4feb5490849"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a character, implicitly converts invalid characters.  <a href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">More...</a><br /></td></tr>
<tr class="separator:a822d399ab808842870a4a4feb5490849"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acdd042ae2f04d986f65a7b60edd6e83e"><td class="memItemLeft" align="right" valign="top">constexpr derived_type &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">assign_rank</a> (<a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> const c) noexcept</td></tr>
<tr class="memdesc:acdd042ae2f04d986f65a7b60edd6e83e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign from a numeric value.  <a href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">More...</a><br /></td></tr>
<tr class="separator:acdd042ae2f04d986f65a7b60edd6e83e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-methods"></a>
Static Public Member Functions</h2></td></tr>
<tr class="memitem:ad6758a49085273e9f1b4811ebd2d1c8a"><td class="memItemLeft" align="right" valign="top">static constexpr bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1nucleotide__base.html#ad6758a49085273e9f1b4811ebd2d1c8a">char_is_valid</a> (<a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const c) noexcept</td></tr>
<tr class="memdesc:ad6758a49085273e9f1b4811ebd2d1c8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Validate whether a character value has a one-to-one mapping to an alphabet value.  <a href="classbio_1_1nucleotide__base.html#ad6758a49085273e9f1b4811ebd2d1c8a">More...</a><br /></td></tr>
<tr class="separator:ad6758a49085273e9f1b4811ebd2d1c8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:a989cf18d33d15b0d858c2dd540480468"><td class="memItemLeft" align="right" valign="top"><a id="a989cf18d33d15b0d858c2dd540480468"></a>
static constexpr <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">detail::min_viable_uint_t</a>&lt; size &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a989cf18d33d15b0d858c2dd540480468">alphabet_size</a> = size</td></tr>
<tr class="memdesc:a989cf18d33d15b0d858c2dd540480468"><td class="mdescLeft">&#160;</td><td class="mdescRight">The size of the alphabet, i.e. the number of different values it can take. <br /></td></tr>
<tr class="separator:a989cf18d33d15b0d858c2dd540480468"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pro-types"></a>
Protected Types</h2></td></tr>
<tr><td colspan="2"><div class="groupHeader">Member types</div></td></tr>
<tr class="memitem:a9934c83725128fc1cde969618f2b4f66"><td class="memItemLeft" align="right" valign="top"><a id="a9934c83725128fc1cde969618f2b4f66"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">std::conditional_t</a>&lt; std::same_as&lt; char_t, void &gt;, char, char_t &gt;</td></tr>
<tr class="memdesc:a9934c83725128fc1cde969618f2b4f66"><td class="mdescLeft">&#160;</td><td class="mdescRight">The char representation; conditional needed to make semi alphabet definitions legal. <br /></td></tr>
<tr class="separator:a9934c83725128fc1cde969618f2b4f66"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a040aa76e8e3fc0e5735c4b135f302c13"><td class="memItemLeft" align="right" valign="top"><a id="a040aa76e8e3fc0e5735c4b135f302c13"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/conditional.html">detail::min_viable_uint_t</a>&lt; size - 1 &gt;</td></tr>
<tr class="memdesc:a040aa76e8e3fc0e5735c4b135f302c13"><td class="mdescLeft">&#160;</td><td class="mdescRight">The type of the alphabet when represented as a number (e.g. via <a class="el" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b" title="Return the letter&#39;s numeric value (rank in the alphabet).">to_rank()</a>). <br /></td></tr>
<tr class="separator:a040aa76e8e3fc0e5735c4b135f302c13"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="related"></a>
Related Functions</h2></td></tr>
<tr><td class="ititle" colspan="2"><p>(Note that these are not member functions.) </p>
</td></tr>
<tr class="memitem:a134746621718b69dcac674faab459138"><td class="memItemLeft" align="right" valign="top"><a id="a134746621718b69dcac674faab459138"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#a134746621718b69dcac674faab459138">sam_dna16_vector</a> = <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a>&lt; <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> &gt;</td></tr>
<tr class="memdesc:a134746621718b69dcac674faab459138"><td class="mdescLeft">&#160;</td><td class="mdescRight">Alias for an <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a> of <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a>. <br /></td></tr>
<tr class="separator:a134746621718b69dcac674faab459138"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Literals</div></td></tr>
<tr class="memitem:afbcd88c55d87c90e0b390235514446a8"><td class="memItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#afbcd88c55d87c90e0b390235514446a8">operator&quot;&quot;_sam_dna16</a> (char const c) noexcept</td></tr>
<tr class="memdesc:afbcd88c55d87c90e0b390235514446a8"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a> char literal.  <a href="classbio_1_1sam__dna16.html#afbcd88c55d87c90e0b390235514446a8">More...</a><br /></td></tr>
<tr class="separator:afbcd88c55d87c90e0b390235514446a8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3fffe37101808410674b94bdfa5060d8"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classbio_1_1sam__dna16.html#a134746621718b69dcac674faab459138">sam_dna16_vector</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1sam__dna16.html#a3fffe37101808410674b94bdfa5060d8">operator&quot;&quot;_sam_dna16</a> (char const *s, size_t n)</td></tr>
<tr class="memdesc:a3fffe37101808410674b94bdfa5060d8"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a> string literal.  <a href="classbio_1_1sam__dna16.html#a3fffe37101808410674b94bdfa5060d8">More...</a><br /></td></tr>
<tr class="separator:a3fffe37101808410674b94bdfa5060d8"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('='). </p>
<p>The <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a> alphabet is the nucleotide alphabet used inside the SAM, BAM and CRAM formats. It has all the letters of the <a class="el" href="classbio_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap.">bio::dna15</a> alphabet and the extra alphabet character '=' which denotes a nucleotide character identical to the reference. Without the context of this reference sequence, no assumptions can be made about the actual value of '=' letter.</p>
<p>Note that you can assign 'U' as a character to <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">sam_dna16</a> and it will silently be converted to 'T'. Lower case letters are accepted when assigning from char (just like <a class="el" href="classbio_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap.">bio::dna15</a>) and unknown characters are silently converted to 'N'.</p>
<p>The complement is the same as for <a class="el" href="classbio_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap.">bio::dna15</a>, with the addition that the complement of '=' is unknown and therefore set to 'N'.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="sam__dna16_8hpp.html">bio/alphabet/nucleotide/sam_dna16.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_sam_dna16;</div>
<div class="line"> </div>
<div class="line">    <a class="code" href="classbio_1_1sam__dna16.html">bio::sam_dna16</a> my_letter{<span class="charliteral">&#39;A&#39;</span>_sam_dna16};</div>
<div class="line"> </div>
<div class="line">    my_letter.<a class="code" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">assign_char</a>(<span class="charliteral">&#39;=&#39;</span>);</div>
<div class="line"> </div>
<div class="line">    my_letter.assign_char(<span class="charliteral">&#39;F&#39;</span>); <span class="comment">// unknown characters are implicitly converted to N.</span></div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{}\n&quot;</span>, my_letter); <span class="comment">// &quot;N&quot;;</span></div>
<div class="line">}</div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_a822d399ab808842870a4a4feb5490849"><div class="ttname"><a href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">bio::alphabet_base::assign_char</a></div><div class="ttdeci">constexpr derived_type &amp; assign_char(char_type const c) noexcept</div><div class="ttdoc">Assign from a character, implicitly converts invalid characters.</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:143</div></div>
<div class="ttc" id="aclassbio_1_1sam__dna16_html"><div class="ttname"><a href="classbio_1_1sam__dna16.html">bio::sam_dna16</a></div><div class="ttdoc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=').</div><div class="ttdef"><b>Definition:</b> sam_dna16.hpp:47</div></div>
<div class="ttc" id="afmt_8hpp_html"><div class="ttname"><a href="fmt_8hpp.html">fmt.hpp</a></div><div class="ttdoc">Core alphabet concept and free function/type trait wrappers.</div></div>
<div class="ttc" id="asam__dna16_8hpp_html"><div class="ttname"><a href="sam__dna16_8hpp.html">sam_dna16.hpp</a></div><div class="ttdoc">Provides bio::sam_dna16.</div></div>
</div><!-- fragment --> </div><h2 class="groupheader">Member Function Documentation</h2>
<a id="a822d399ab808842870a4a4feb5490849"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a822d399ab808842870a4a4feb5490849">&#9670;&nbsp;</a></span>assign_char()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr derived_type&amp; <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::assign_char </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a character, implicitly converts invalid characters. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The character to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga05da42eb0b9a48f4d1a554c7d3a9044d" title="Assign a character to an alphabet object.">bio::assign_char_to</a>, required to model <a class="el" href="interfacebio_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="acdd042ae2f04d986f65a7b60edd6e83e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#acdd042ae2f04d986f65a7b60edd6e83e">&#9670;&nbsp;</a></span>assign_rank()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr derived_type&amp; <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::assign_rank </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign from a numeric value. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">c</td><td>The rank to be assigned.</td></tr>
  </table>
  </dd>
</dl>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#gae34fa2b37a8b40f2f2fa64fed7bbb410" title="Assign a rank to an alphabet object.">bio::assign_rank_to</a>, required to model <a class="el" href="interfacebio_1_1semialphabet.html" title="The basis for bio::alphabet, but requires only rank interface (not char).">bio::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="ad6758a49085273e9f1b4811ebd2d1c8a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ad6758a49085273e9f1b4811ebd2d1c8a">&#9670;&nbsp;</a></span>char_is_valid()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">static constexpr bool <a class="el" href="classbio_1_1nucleotide__base.html">bio::nucleotide_base</a>&lt; <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> , size &gt;::char_is_valid </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">static</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Validate whether a character value has a one-to-one mapping to an alphabet value. </p>
<p>Satisfies the <a class="el" href="group__alphabet.html#ga510b0657b8a3ae7d304999afd8fdcd52" title="Returns whether a character is in the valid set of a bio::alphabet (usually implies a bijective mappi...">bio::semialphabet::char_is_valid_for()</a> requirement via the <a class="el" href="group__alphabet.html#ga510b0657b8a3ae7d304999afd8fdcd52" title="Returns whether a character is in the valid set of a bio::alphabet (usually implies a bijective mappi...">bio::char_is_valid_for()</a> wrapper.</p>
<p>Behaviour specific to nucleotides: True also for lower case letters that silently convert to their upper case <b>and</b> true also for U/T respectively, e.g. 'U' is a valid character for <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">bio::dna4</a>, because its informational content is identical to 'T'.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a0e9f68ab92a2638509288eeb840aa2ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a0e9f68ab92a2638509288eeb840aa2ec">&#9670;&nbsp;</a></span>complement()</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a>  <a class="el" href="classbio_1_1nucleotide__base.html">bio::nucleotide_base</a>&lt; <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> , size &gt;::complement </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the complement of the letter. </p>
<p>See <a class="el" href="group__nucleotide.html">Nucleotide</a> for the actual values.</p>
<p>Provides an implementation for <a class="el" href="group__nucleotide.html#ga4b00de5b7d39cf416e62e14edef83f03" title="Return the complement of a nucleotide object.">bio::complement</a>, required to model <a class="el" href="interfacebio_1_1nucleotide__alphabet.html" title="A concept that indicates whether an alphabet represents nucleotides.">bio::nucleotide_alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a6616ba2cb25069f5192589ad53651a88"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a6616ba2cb25069f5192589ad53651a88">&#9670;&nbsp;</a></span>to_char()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a9934c83725128fc1cde969618f2b4f66">char_type</a> <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::to_char </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter as a character of char_type. </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8" title="Return the char representation of an alphabet object.">bio::to_char</a>, required to model <a class="el" href="interfacebio_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<a id="a443c2f7716e384ea2acec1d2bced7c8b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a443c2f7716e384ea2acec1d2bced7c8b">&#9670;&nbsp;</a></span>to_rank()</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename derived_type , size_t size, typename char_t  = char&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1alphabet__base.html#a040aa76e8e3fc0e5735c4b135f302c13">rank_type</a> <a class="el" href="classbio_1_1alphabet__base.html">bio::alphabet_base</a>&lt; derived_type, size, char_t &gt;::to_rank </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span><span class="mlabel">inherited</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the letter's numeric value (rank in the alphabet). </p>
<p>Provides an implementation for <a class="el" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0" title="Return the rank representation of a (semi-)alphabet object.">bio::to_rank</a>, required to model <a class="el" href="interfacebio_1_1semialphabet.html" title="The basis for bio::alphabet, but requires only rank interface (not char).">bio::semialphabet</a>.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Guaranteed not to throw. </p>

</div>
</div>
<h2 class="groupheader">Friends And Related Function Documentation</h2>
<a id="a3fffe37101808410674b94bdfa5060d8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a3fffe37101808410674b94bdfa5060d8">&#9670;&nbsp;</a></span>operator&quot;&quot;_sam_dna16() <span class="overload">[1/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classbio_1_1sam__dna16.html#a134746621718b69dcac674faab459138">sam_dna16_vector</a> operator&quot;&quot;_sam_dna16 </td>
          <td>(</td>
          <td class="paramtype">char const *&#160;</td>
          <td class="paramname"><em>s</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t&#160;</td>
          <td class="paramname"><em>n</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">related</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a> string literal. </p>
<dl class="section return"><dt>Returns</dt><dd>bio::sam_dna16_vector </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">s</td><td>The string literal to assign from. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">n</td><td>The length of the string literal s.</td></tr>
  </table>
  </dd>
</dl>
<p>You can use this string literal to easily assign to bio::sam_dna16_vector:</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="sam__dna16_8hpp.html">bio/alphabet/nucleotide/sam_dna16.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_sam_dna16;</div>
<div class="line"> </div>
<div class="line">    bio::sam_dna16_vector foo{<span class="stringliteral">&quot;ACgtTA&quot;</span>_sam_dna16};</div>
<div class="line">    bio::sam_dna16_vector bar = <span class="stringliteral">&quot;ACG==A&quot;</span>_sam_dna16;</div>
<div class="line">    <span class="keyword">auto</span> bax = <span class="stringliteral">&quot;A=GTT!&quot;</span>_sam_dna16;</div>
<div class="line"> </div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;{}\n{}\n{}\n&quot;</span>, foo, bar, bax);</div>
<div class="line">}</div>
</div><!-- fragment --> 
</div>
</div>
<a id="afbcd88c55d87c90e0b390235514446a8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#afbcd88c55d87c90e0b390235514446a8">&#9670;&nbsp;</a></span>operator&quot;&quot;_sam_dna16() <span class="overload">[2/2]</span></h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a> operator&quot;&quot;_sam_dna16 </td>
          <td>(</td>
          <td class="paramtype">char const&#160;</td>
          <td class="paramname"><em>c</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">related</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The <a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a> char literal. </p>
<dl class="section return"><dt>Returns</dt><dd><a class="el" href="classbio_1_1sam__dna16.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;).">bio::sam_dna16</a> </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">c</td><td>The character to assign from. </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>bio/alphabet/nucleotide/<a class="el" href="sam__dna16_8hpp_source.html">sam_dna16.hpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="navelem"><a class="el" href="namespacebio.html">bio</a></li><li class="navelem"><a class="el" href="classbio_1_1sam__dna16.html">sam_dna16</a></li>
    <li class="footer">Generated on Sun Jun 12 2022 17:23:28 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
