<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.6"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: Nucleotide</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign">
   <div id="projectname">BioC++<span id="projectnumber">&#160;core-0.7.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
    <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <span id="MSearchSelect"                onmouseover="return searchBox.OnSearchSelectShow()"                onmouseout="return searchBox.OnSearchSelectHide()">&#160;</span>
          <input type="text" id="MSearchField" value="" placeholder="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.6 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('group__nucleotide.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#concepts">Concepts</a>  </div>
  <div class="headertitle"><div class="title">Nucleotide<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a></div></div></div>
</div><!--header-->
<div class="contents">

<p>Provides the different DNA and RNA alphabet types.  
<a href="#details">More...</a></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Nucleotide:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__nucleotide.svg" width="216" height="36"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe>
</div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna15.html">bio::alphabet::dna15</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..  <a href="classbio_1_1alphabet_1_1dna15.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">bio::alphabet::dna16sam</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=')..  <a href="classbio_1_1alphabet_1_1dna16sam.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna4.html">bio::alphabet::dna4</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The four letter DNA alphabet of A,C,G,T..  <a href="classbio_1_1alphabet_1_1dna4.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna5.html">bio::alphabet::dna5</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The five letter DNA alphabet of A,C,G,T and the unknown character N..  <a href="classbio_1_1alphabet_1_1dna5.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html">bio::alphabet::nucleotide_base&lt; derived_type, size &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::base</a> and is used by the nucleotides.  <a href="classbio_1_1alphabet_1_1nucleotide__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1rna15.html">bio::alphabet::rna15</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap..  <a href="classbio_1_1alphabet_1_1rna15.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1rna4.html">bio::alphabet::rna4</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The four letter RNA alphabet of A,C,G,U..  <a href="classbio_1_1alphabet_1_1rna4.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1rna5.html">bio::alphabet::rna5</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The five letter RNA alphabet of A,C,G,U and the unknown character N..  <a href="classbio_1_1alphabet_1_1rna5.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="concepts" name="concepts"></a>
Concepts</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1nucleotide.html">bio::alphabet::nucleotide</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether an alphabet represents nucleotides. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Function objects (Nucleotide)</h2></td></tr>
<tr class="memitem:ga174261f8f888bbce628ffe01097a9b48"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__nucleotide.html#ga174261f8f888bbce628ffe01097a9b48">bio::alphabet::complement</a></td></tr>
<tr class="memdesc:ga174261f8f888bbce628ffe01097a9b48"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the complement of a nucleotide object.  <br /></td></tr>
<tr class="separator:ga174261f8f888bbce628ffe01097a9b48"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Provides the different DNA and RNA alphabet types. </p>
<h3>Introduction</h3>
<p>Nucleotide sequences are at the core of most bioinformatic data processing and while it is possible to represent them in a regular <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/string/basic_string.html">std::string</a>, it makes sense to have specialised data structures in most cases. This sub-module offers multiple nucleotide alphabets that can be used with regular containers and ranges.</p>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadCenter">Letter   </th><th class="markdownTableHeadNone">Description   </th><th class="markdownTableHeadCenter"><a class="el" href="classbio_1_1alphabet_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::dna15</a>   </th><th class="markdownTableHeadCenter"><a class="el" href="classbio_1_1alphabet_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N..">bio::alphabet::dna5</a>   </th><th class="markdownTableHeadCenter"><a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a>   </th><th class="markdownTableHeadCenter"><a class="el" href="classbio_1_1alphabet_1_1rna15.html" title="The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::rna15</a>   </th><th class="markdownTableHeadCenter"><a class="el" href="classbio_1_1alphabet_1_1rna5.html" title="The five letter RNA alphabet of A,C,G,U and the unknown character N..">bio::alphabet::rna5</a>   </th><th class="markdownTableHeadCenter"><a class="el" href="classbio_1_1alphabet_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U..">bio::alphabet::rna4</a>    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyNone">Adenine   </td><td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyCenter">A    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyNone">Cytosine   </td><td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyCenter">C    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyNone">Guanine   </td><td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyCenter">G    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">T   </td><td class="markdownTableBodyNone">Thymine (DNA)   </td><td class="markdownTableBodyCenter">T   </td><td class="markdownTableBodyCenter">T   </td><td class="markdownTableBodyCenter">T   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">U</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">U</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">U</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">U   </td><td class="markdownTableBodyNone">Uracil (RNA)   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">T</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">T</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">T</span>   </td><td class="markdownTableBodyCenter">U   </td><td class="markdownTableBodyCenter">U   </td><td class="markdownTableBodyCenter">U    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">M   </td><td class="markdownTableBodyNone">A <em>or</em> C   </td><td class="markdownTableBodyCenter">M   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">M   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">R   </td><td class="markdownTableBodyNone">A <em>or</em> G   </td><td class="markdownTableBodyCenter">R   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">R   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">W   </td><td class="markdownTableBodyNone">A <em>or</em> T   </td><td class="markdownTableBodyCenter">W   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">W   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">Y   </td><td class="markdownTableBodyNone">C <em>or</em> T   </td><td class="markdownTableBodyCenter">Y   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">C</span>   </td><td class="markdownTableBodyCenter">Y   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">C</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">S   </td><td class="markdownTableBodyNone">C <em>or</em> G   </td><td class="markdownTableBodyCenter">S   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">C</span>   </td><td class="markdownTableBodyCenter">S   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">C</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">K   </td><td class="markdownTableBodyNone">G <em>or</em> T   </td><td class="markdownTableBodyCenter">K   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">G</span>   </td><td class="markdownTableBodyCenter">K   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">G</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">V   </td><td class="markdownTableBodyNone">A <em>or</em> C <em>or</em> G   </td><td class="markdownTableBodyCenter">V   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">V   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">H   </td><td class="markdownTableBodyNone">A <em>or</em> C <em>or</em> T   </td><td class="markdownTableBodyCenter">H   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">H   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">D   </td><td class="markdownTableBodyNone">A <em>or</em> G <em>or</em> T   </td><td class="markdownTableBodyCenter">D   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">D   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">B   </td><td class="markdownTableBodyNone">C <em>or</em> G <em>or</em> T   </td><td class="markdownTableBodyCenter">B   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">C</span>   </td><td class="markdownTableBodyCenter">B   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">N</span>   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">C</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">N   </td><td class="markdownTableBodyNone">A <em>or</em> C <em>or</em> G <em>or</em> T   </td><td class="markdownTableBodyCenter">N   </td><td class="markdownTableBodyCenter">N   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>   </td><td class="markdownTableBodyCenter">N   </td><td class="markdownTableBodyCenter">N   </td><td class="markdownTableBodyCenter"><span style="color:LightGrey">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter"><b>Size</b>   </td><td class="markdownTableBodyNone"></td><td class="markdownTableBodyCenter">15   </td><td class="markdownTableBodyCenter">5   </td><td class="markdownTableBodyCenter">4   </td><td class="markdownTableBodyCenter">15   </td><td class="markdownTableBodyCenter">5   </td><td class="markdownTableBodyCenter">4   </td></tr>
</table>
<p>Keep in mind, that while we think of "the nucleotide alphabet" as consisting of four bases, there are indeed more characters defined with different levels of ambiguity. Depending on your application it will make sense to preserve this ambiguity or to discard it to save space and/or optimise computations. BioC++ offers six distinct nucleotide alphabet types to accommodate for this. There is a seventh alphabet <a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" title="A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign (&#39;=&#39;)....">bio::alphabet::dna16sam</a> which implements the alphabet used in SAM/BAM/CRAM.</p>
<p>The specialised RNA alphabets are provided for convenience, however the DNA alphabets can handle being assigned a &lsquo;'U&rsquo;` character, as well. See below for the details.</p>
<p>Which alphabet to chose?</p><ol type="1">
<li>in most cases, take <a class="el" href="classbio_1_1alphabet_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::dna15</a> (includes all IUPAC characters)</li>
<li>if you are memory constrained and sequence data is actually the main memory consumer, use <a class="el" href="classbio_1_1alphabet_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N..">bio::alphabet::dna5</a></li>
<li>if you use specialised algorithms that profit from a 2-bit representation, use <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a></li>
<li>if you are doing only RNA input/output, use the respective bio::alphabet::rna* type</li>
<li>to actually save space from using smaller alphabets, you need a compressed container (e.g. <a class="el" href="classbio_1_1ranges_1_1bitcompressed__vector.html" title="A space-optimised version of std::vector that compresses multiple letters into a single byte.">bio::ranges::bitcompressed_vector</a>)</li>
</ol>
<h3>Printing and conversion to char</h3>
<p>As with all alphabets in BioC++, none of the nucleotide alphabets can be directly converted to char or printed with iostreams. You need to explicitly call <a class="el" href="group__alphabet.html#ga064e386f9360cca0fd3702b4957a27a0" title="Return the char representation of an alphabet object.">bio::alphabet::to_char</a> to convert to char or use the {fmt}-library which automatically converts.</p>
<p><code>T</code> and <code>U</code> are represented by the same rank and you cannot differentiate between them. The only difference between e.g. <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a> and <a class="el" href="classbio_1_1alphabet_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U..">bio::alphabet::rna4</a> is the output when calling to_char().</p>
<h3>Assignment and conversions between nucleotide types</h3>
<ul>
<li>Nucleotide types defined here are <b>implicitly</b> convertible to each other if they have the same size (e.g. <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a> ↔ <a class="el" href="classbio_1_1alphabet_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U..">bio::alphabet::rna4</a>).</li>
<li>Other nucleotide types are <b>explicitly</b> convertible to each other through their character representation.</li>
<li>All ranges of nucleotide alphabets are convertible to each other via bio::views::convert.</li>
<li>None of the nucleotide alphabets can be directly converted or assigned from <code>char</code>. You need to explicitly call <code>assign_char</code> or use a literal (see below).</li>
</ul>
<p>When assigning from <code>char</code> or converting from a larger nucleotide alphabet to a smaller one, <em>loss of information</em> can occur since obviously some bases are not available. When converting to <a class="el" href="classbio_1_1alphabet_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N..">bio::alphabet::dna5</a> or <a class="el" href="classbio_1_1alphabet_1_1rna5.html" title="The five letter RNA alphabet of A,C,G,U and the unknown character N..">bio::alphabet::rna5</a>, non-canonical bases (letters other than A, C, G, T, U) are converted to &lsquo;'N&rsquo;` to preserve ambiguity at that position, while for <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a> and <a class="el" href="classbio_1_1alphabet_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U..">bio::alphabet::rna4</a> they are converted to the first of the possibilities they represent (because there is no letter &lsquo;'N&rsquo;` to represent ambiguity). See the greyed out values in the table at the top for an overview of which conversions take place.</p>
<p><code>char</code> values that are none of the IUPAC symbols, e.g. 'P', are always converted to the equivalent of assigning 'N', i.e. they result in 'A' for <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a> and <a class="el" href="classbio_1_1alphabet_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U..">bio::alphabet::rna4</a>, and in 'N' for the other alphabets.</p>
<h3>Literals</h3>
<p>To avoid writing <code>dna4{}.assign_char('C')</code> every time, you may instead use the literal <code>'C'_dna4</code>. All nucleotide types defined here have character literals and also string literals which return a vector of the respective type.</p>
<h3>Concept</h3>
<p>The nucleotide submodule defines <a class="el" href="conceptbio_1_1alphabet_1_1nucleotide.html" title="A concept that indicates whether an alphabet represents nucleotides.">bio::alphabet::nucleotide</a> which encompasses all the alphabets defined in the submodule and refines <a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a>. The only additional requirement is that their values can be complemented, see below.</p>
<h3>Complement</h3>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadCenter">Letter   </th><th class="markdownTableHeadNone">Description   </th><th class="markdownTableHeadCenter">Complement    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">A   </td><td class="markdownTableBodyNone">Adenine   </td><td class="markdownTableBodyCenter">T    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">C   </td><td class="markdownTableBodyNone">Cytosine   </td><td class="markdownTableBodyCenter">G    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">G   </td><td class="markdownTableBodyNone">Guanine   </td><td class="markdownTableBodyCenter">C    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">T   </td><td class="markdownTableBodyNone">Thymine (DNA)   </td><td class="markdownTableBodyCenter">A    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">U   </td><td class="markdownTableBodyNone">Uracil (RNA)   </td><td class="markdownTableBodyCenter">A    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">M   </td><td class="markdownTableBodyNone">A <em>or</em> C   </td><td class="markdownTableBodyCenter">K    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">R   </td><td class="markdownTableBodyNone">A <em>or</em> G   </td><td class="markdownTableBodyCenter">Y    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">W   </td><td class="markdownTableBodyNone">A <em>or</em> T   </td><td class="markdownTableBodyCenter">W    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">Y   </td><td class="markdownTableBodyNone">C <em>or</em> T   </td><td class="markdownTableBodyCenter">R    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">S   </td><td class="markdownTableBodyNone">C <em>or</em> G   </td><td class="markdownTableBodyCenter">S    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">K   </td><td class="markdownTableBodyNone">G <em>or</em> T   </td><td class="markdownTableBodyCenter">M    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">V   </td><td class="markdownTableBodyNone">A <em>or</em> C <em>or</em> G   </td><td class="markdownTableBodyCenter">B    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">H   </td><td class="markdownTableBodyNone">A <em>or</em> C <em>or</em> T   </td><td class="markdownTableBodyCenter">D    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">D   </td><td class="markdownTableBodyNone">A <em>or</em> G <em>or</em> T   </td><td class="markdownTableBodyCenter">H    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">B   </td><td class="markdownTableBodyNone">C <em>or</em> G <em>or</em> T   </td><td class="markdownTableBodyCenter">V    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">N   </td><td class="markdownTableBodyNone">A <em>or</em> C <em>or</em> G <em>or</em> T   </td><td class="markdownTableBodyCenter">N   </td></tr>
</table>
<p>In the typical structure of DNA molecules (or double-stranded RNA), each nucleotide has a complement that it pairs with. To generate the complement value of a nucleotide letter, you can call an implementation of bio::alphabet::nucleotide::complement() on it.</p>
<p>For the ambiguous letters, the complement is the (possibly also ambiguous) letter representing the variant of the individual complements. </p>
<h2 class="groupheader">Variable Documentation</h2>
<a id="ga174261f8f888bbce628ffe01097a9b48" name="ga174261f8f888bbce628ffe01097a9b48"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga174261f8f888bbce628ffe01097a9b48">&#9670;&#160;</a></span>complement</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr auto bio::alphabet::complement</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Return the complement of a nucleotide object. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">alph_type</td><td>Type of the argument. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alph</td><td>The nucleotide object for which you want to receive the complement. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The complement character of <code>alph</code>, e.g. 'C' for 'G'.</dd></dl>
<p>This is a function object. Invoke it with the parameter(s) specified above.</p>
<p>It is defined for all nucleotide alphabets in BioC++.</p>
<h3>Example</h3>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="alphabet_2nucleotide_2concept_8hpp.html">bio/alphabet/nucleotide/concept.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="rna5_8hpp.html">bio/alphabet/nucleotide/rna5.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">using namespace </span><a class="code hl_namespace" href="namespacebio_1_1alphabet_1_1literals.html">bio::alphabet::literals</a>;</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main()</div>
<div class="line">{</div>
<div class="line">    <span class="keyword">auto</span> r1 = <span class="charliteral">&#39;A&#39;</span>_rna5.complement();        <span class="comment">// calls member function rna5::complement(); r1 == &#39;U&#39;_rna5</span></div>
<div class="line">    <span class="keyword">auto</span> r2 = <a class="code hl_variable" href="group__nucleotide.html#ga174261f8f888bbce628ffe01097a9b48">bio::alphabet::complement</a>(<span class="charliteral">&#39;A&#39;</span>_rna5); <span class="comment">// calls global complement() function on the rna5 object; r2 == &#39;U&#39;_rna5</span></div>
<div class="line">}</div>
<div class="ttc" id="aalphabet_2nucleotide_2concept_8hpp_html"><div class="ttname"><a href="alphabet_2nucleotide_2concept_8hpp.html">concept.hpp</a></div><div class="ttdoc">Provides bio::alphabet::nucleotide.</div></div>
<div class="ttc" id="agroup__nucleotide_html_ga174261f8f888bbce628ffe01097a9b48"><div class="ttname"><a href="group__nucleotide.html#ga174261f8f888bbce628ffe01097a9b48">bio::alphabet::complement</a></div><div class="ttdeci">constexpr auto complement</div><div class="ttdoc">Return the complement of a nucleotide object.</div><div class="ttdef"><b>Definition:</b> concept.hpp:68</div></div>
<div class="ttc" id="anamespacebio_1_1alphabet_1_1literals_html"><div class="ttname"><a href="namespacebio_1_1alphabet_1_1literals.html">bio::alphabet::literals</a></div><div class="ttdoc">An inline namespace for alphabet literals. It exists to safely allow using namespace.</div><div class="ttdef"><b>Definition:</b> aa10li.hpp:183</div></div>
<div class="ttc" id="arna5_8hpp_html"><div class="ttname"><a href="rna5_8hpp.html">rna5.hpp</a></div><div class="ttdoc">Provides bio::alphabet::rna5, container aliases and string literals.</div></div>
</div><!-- fragment --><h3>Customisation point</h3>
<p>This is a customisation point (see <a class="el" href="biocpp_customisation.html">Customisation</a>). <b>If you don't want to create your own alphabet, everything below is irrelevant to you!</b></p>
<p>This object acts as a wrapper and looks for an implementation with the following signature:</p>
<div class="fragment"><div class="line"><span class="keyword">constexpr</span> alph_type tag_invoke(<a class="code hl_struct" href="structbio_1_1alphabet_1_1custom_1_1complement.html">bio::alphabet::custom::complement</a>, alph_type <span class="keyword">const</span> alph) <span class="keyword">noexcept</span></div>
<div class="line">{}</div>
<div class="ttc" id="astructbio_1_1alphabet_1_1custom_1_1complement_html"><div class="ttname"><a href="structbio_1_1alphabet_1_1custom_1_1complement.html">bio::alphabet::custom::complement</a></div><div class="ttdoc">Customisation tag for bio::alphabet::complement.</div><div class="ttdef"><b>Definition:</b> tag.hpp:49</div></div>
</div><!-- fragment --><p>Functions are found via ADL and considered only if they are marked <code>noexcept</code> (<code>constexpr</code> is not required, but recommended). The returned type must be (implicitly) convertible to <code>alph_type</code> (usually it should be the same type).</p>
<p>To specify the behaviour for your own alphabet type, simply provide the above function as a <code>friend</code> or free function. </p>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="footer">Generated on Thu Mar 30 2023 15:25:52 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.6 </li>
  </ul>
</div>
</body>
</html>
