<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: Aminoacid</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">0.1.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__aminoacid.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<div class="title">Aminoacid<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Provides the amino acid alphabets and functionality for translation from nucleotide.  
<a href="#details">More...</a></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Aminoacid:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__aminoacid.svg" width="230" height="36"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1aa10li.html">bio::aa10li</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The reduced Li amino acid alphabet.  <a href="classbio_1_1aa10li.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1aa10murphy.html">bio::aa10murphy</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The reduced Murphy amino acid alphabet.  <a href="classbio_1_1aa10murphy.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1aa20.html">bio::aa20</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The canonical amino acid alphabet.  <a href="classbio_1_1aa20.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1aa27.html">bio::aa27</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The twenty-seven letter amino acid alphabet.  <a href="classbio_1_1aa27.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">interface &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="interfacebio_1_1aminoacid__alphabet.html">aminoacid_alphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether an alphabet represents amino acids.  <a href="interfacebio_1_1aminoacid__alphabet.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1aminoacid__base.html">bio::aminoacid_base&lt; derived_type, size &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet_base</a> and is used by the amino acids.  <a href="classbio_1_1aminoacid__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structbio_1_1aminoacid__empty__base.html">bio::aminoacid_empty_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">This is an empty base class that can be inherited by types that shall model <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a>.  <a href="structbio_1_1aminoacid__empty__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:gaca1435b57fb28ffa65862d911e23a960"><td class="memTemplParams" colspan="2">template&lt;genetic_code gc = genetic_code::CANONICAL, nucleotide_alphabet nucl_type&gt; </td></tr>
<tr class="memitem:gaca1435b57fb28ffa65862d911e23a960"><td class="memTemplItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#gaca1435b57fb28ffa65862d911e23a960">bio::translate_triplet</a> (nucl_type const &amp;n1, nucl_type const &amp;n2, nucl_type const &amp;n3) noexcept</td></tr>
<tr class="memdesc:gaca1435b57fb28ffa65862d911e23a960"><td class="mdescLeft">&#160;</td><td class="mdescRight">Translate one nucleotide triplet into single amino acid (single nucleotide interface).  <a href="group__aminoacid.html#gaca1435b57fb28ffa65862d911e23a960">More...</a><br /></td></tr>
<tr class="separator:gaca1435b57fb28ffa65862d911e23a960"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga056c30dd6a4534f5ad8395b146a7d2b3"><td class="memTemplParams" colspan="2">template&lt;genetic_code gc = genetic_code::CANONICAL, std::ranges::input_range range_type&gt; </td></tr>
<tr class="memitem:ga056c30dd6a4534f5ad8395b146a7d2b3"><td class="memTemplItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#ga056c30dd6a4534f5ad8395b146a7d2b3">bio::translate_triplet</a> (range_type &amp;&amp;input_range)</td></tr>
<tr class="memdesc:ga056c30dd6a4534f5ad8395b146a7d2b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Translate one nucleotide triplet into single amino acid (range interface).  <a href="group__aminoacid.html#ga056c30dd6a4534f5ad8395b146a7d2b3">More...</a><br /></td></tr>
<tr class="separator:ga056c30dd6a4534f5ad8395b146a7d2b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga594a06e9c7719f77a3828239c149492a"><td class="memTemplParams" colspan="2">template&lt;genetic_code gc = genetic_code::CANONICAL, std::ranges::random_access_range rng_t&gt; </td></tr>
<tr class="memitem:ga594a06e9c7719f77a3828239c149492a"><td class="memTemplItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#ga594a06e9c7719f77a3828239c149492a">bio::translate_triplet</a> (rng_t &amp;&amp;input_range)</td></tr>
<tr class="memdesc:ga594a06e9c7719f77a3828239c149492a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Translate one nucleotide triplet into single amino acid (range interface, input range allows random access).  <a href="group__aminoacid.html#ga594a06e9c7719f77a3828239c149492a">More...</a><br /></td></tr>
<tr class="separator:ga594a06e9c7719f77a3828239c149492a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaac5a13cc5cb827eb7df84b6216268a73"><td class="memTemplParams" colspan="2">template&lt;genetic_code gc = genetic_code::CANONICAL, typename tuple_type &gt; </td></tr>
<tr class="memitem:gaac5a13cc5cb827eb7df84b6216268a73"><td class="memTemplItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#gaac5a13cc5cb827eb7df84b6216268a73">bio::translate_triplet</a> (tuple_type const &amp;input_tuple) noexcept</td></tr>
<tr class="memdesc:gaac5a13cc5cb827eb7df84b6216268a73"><td class="mdescLeft">&#160;</td><td class="mdescRight">Translate one nucleotide triplet into single amino acid (tuple interface).  <a href="group__aminoacid.html#gaac5a13cc5cb827eb7df84b6216268a73">More...</a><br /></td></tr>
<tr class="separator:gaac5a13cc5cb827eb7df84b6216268a73"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:ga32d1a56e5ebf85b772d5af0f471a4974"><td class="memTemplParams" colspan="2">template&lt;typename t &gt; </td></tr>
<tr class="memitem:ga32d1a56e5ebf85b772d5af0f471a4974"><td class="memTemplItemLeft" align="right" valign="top">constexpr bool&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#ga32d1a56e5ebf85b772d5af0f471a4974">bio::enable_aminoacid</a> = detail::adl_only::enable_aminoacid_dispatcher::dispatch&lt;<a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/remove_cvref.html">std::remove_cvref_t</a>&lt;t&gt;&gt;()</td></tr>
<tr class="memdesc:ga32d1a56e5ebf85b772d5af0f471a4974"><td class="mdescLeft">&#160;</td><td class="mdescRight">A trait that indicates whether a type shall model <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a>.  <a href="group__aminoacid.html#ga32d1a56e5ebf85b772d5af0f471a4974">More...</a><br /></td></tr>
<tr class="separator:ga32d1a56e5ebf85b772d5af0f471a4974"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Provides the amino acid alphabets and functionality for translation from nucleotide. </p>
<h3>Introduction</h3>
<p>Amino acid sequences are an important part of bioinformatic data processing and used by many applications and while it is possible to represent them in a regular <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/string/basic_string.html">std::string</a>, it makes sense to have specialised data structures in most cases. This sub-module offers the 27 letter aminoacid alphabet as well as three reduced versions that can be used with regular container and ranges. The 27 letter amino acid alphabet contains the 20 canonical amino acids, 2 additional proteinogenic amino acids (Pyrrolysine and Selenocysteine) and a termination letter (*). Additionally 4 wildcard letters are offered which allow a more generic usage for example in case of ambiguous amino acids (e.g. J which means either Isoleucine or Leucine). See also <a href="https://en.wikipedia.org/wiki/Amino_acid">https://en.wikipedia.org/wiki/Amino_acid</a> for more information about the amino acid alphabet.</p>
<h3>Conversions</h3>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadNone">Amino acid name   </th><th class="markdownTableHeadNone">Three letter code   </th><th class="markdownTableHeadNone">One letter code   </th><th class="markdownTableHeadNone">Remapped in<br  />
 <a class="el" href="classbio_1_1aa20.html" title="The canonical amino acid alphabet.">bio::aa20</a>   </th><th class="markdownTableHeadNone">Remapped in<br  />
 <a class="el" href="classbio_1_1aa10murphy.html" title="The reduced Murphy amino acid alphabet.">bio::aa10murphy</a>   </th><th class="markdownTableHeadNone">Remapped in<br  />
 <a class="el" href="classbio_1_1aa10li.html" title="The reduced Li amino acid alphabet.">bio::aa10li</a>    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Alanine   </td><td class="markdownTableBodyNone">Ala   </td><td class="markdownTableBodyNone">A   </td><td class="markdownTableBodyNone">A   </td><td class="markdownTableBodyNone">A   </td><td class="markdownTableBodyNone">A    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Arginine   </td><td class="markdownTableBodyNone">Arg   </td><td class="markdownTableBodyNone">R   </td><td class="markdownTableBodyNone">R   </td><td class="markdownTableBodyNone"><span style="color:red">K</span>   </td><td class="markdownTableBodyNone"><span style="color:red">K</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Asparagine   </td><td class="markdownTableBodyNone">Asn   </td><td class="markdownTableBodyNone">N   </td><td class="markdownTableBodyNone">N   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>   </td><td class="markdownTableBodyNone"><span style="color:red">H</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Aspartic acid   </td><td class="markdownTableBodyNone">Asp   </td><td class="markdownTableBodyNone">D   </td><td class="markdownTableBodyNone">D   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Cysteine   </td><td class="markdownTableBodyNone">Cys   </td><td class="markdownTableBodyNone">C   </td><td class="markdownTableBodyNone">C   </td><td class="markdownTableBodyNone">C   </td><td class="markdownTableBodyNone">C    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Tyrosine   </td><td class="markdownTableBodyNone">Tyr   </td><td class="markdownTableBodyNone">Y   </td><td class="markdownTableBodyNone">Y   </td><td class="markdownTableBodyNone"><span style="color:red">F</span>   </td><td class="markdownTableBodyNone"><span style="color:red">F</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Glutamic acid   </td><td class="markdownTableBodyNone">Glu   </td><td class="markdownTableBodyNone">E   </td><td class="markdownTableBodyNone">E   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Glutamine   </td><td class="markdownTableBodyNone">Gln   </td><td class="markdownTableBodyNone">Q   </td><td class="markdownTableBodyNone">Q   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Glycine   </td><td class="markdownTableBodyNone">Gly   </td><td class="markdownTableBodyNone">G   </td><td class="markdownTableBodyNone">G   </td><td class="markdownTableBodyNone">G   </td><td class="markdownTableBodyNone">G    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Histidine   </td><td class="markdownTableBodyNone">His   </td><td class="markdownTableBodyNone">H   </td><td class="markdownTableBodyNone">H   </td><td class="markdownTableBodyNone">H   </td><td class="markdownTableBodyNone">H    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Isoleucine   </td><td class="markdownTableBodyNone">Ile   </td><td class="markdownTableBodyNone">I   </td><td class="markdownTableBodyNone">I   </td><td class="markdownTableBodyNone">I   </td><td class="markdownTableBodyNone">I    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Leucine   </td><td class="markdownTableBodyNone">leu   </td><td class="markdownTableBodyNone">L   </td><td class="markdownTableBodyNone">L   </td><td class="markdownTableBodyNone"><span style="color:red">I</span>   </td><td class="markdownTableBodyNone"><span style="color:red">J</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Lysine   </td><td class="markdownTableBodyNone">Lys   </td><td class="markdownTableBodyNone">K   </td><td class="markdownTableBodyNone">K   </td><td class="markdownTableBodyNone">K   </td><td class="markdownTableBodyNone">K    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Methionine   </td><td class="markdownTableBodyNone">Met   </td><td class="markdownTableBodyNone">M   </td><td class="markdownTableBodyNone">M   </td><td class="markdownTableBodyNone"><span style="color:red">I</span>   </td><td class="markdownTableBodyNone"><span style="color:red">J</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Phenylalanine   </td><td class="markdownTableBodyNone">Phe   </td><td class="markdownTableBodyNone">F   </td><td class="markdownTableBodyNone">F   </td><td class="markdownTableBodyNone">F   </td><td class="markdownTableBodyNone">F    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Proline   </td><td class="markdownTableBodyNone">Pro   </td><td class="markdownTableBodyNone">P   </td><td class="markdownTableBodyNone">P   </td><td class="markdownTableBodyNone">P   </td><td class="markdownTableBodyNone">P    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Serine   </td><td class="markdownTableBodyNone">Ser   </td><td class="markdownTableBodyNone">S   </td><td class="markdownTableBodyNone">S   </td><td class="markdownTableBodyNone">S   </td><td class="markdownTableBodyNone"><span style="color:red">A</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Threonine   </td><td class="markdownTableBodyNone">Thr   </td><td class="markdownTableBodyNone">T   </td><td class="markdownTableBodyNone">T   </td><td class="markdownTableBodyNone"><span style="color:red">S</span>   </td><td class="markdownTableBodyNone"><span style="color:red">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Tryptophan   </td><td class="markdownTableBodyNone">Trp   </td><td class="markdownTableBodyNone">W   </td><td class="markdownTableBodyNone">W   </td><td class="markdownTableBodyNone"><span style="color:red">F</span>   </td><td class="markdownTableBodyNone"><span style="color:red">F</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Valine   </td><td class="markdownTableBodyNone">Val   </td><td class="markdownTableBodyNone">V   </td><td class="markdownTableBodyNone">V   </td><td class="markdownTableBodyNone"><span style="color:red">I</span>   </td><td class="markdownTableBodyNone"><span style="color:red">I</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Selenocysteine   </td><td class="markdownTableBodyNone">Sec   </td><td class="markdownTableBodyNone">U   </td><td class="markdownTableBodyNone"><span style="color:red">C</span>   </td><td class="markdownTableBodyNone"><span style="color:red">C</span>   </td><td class="markdownTableBodyNone"><span style="color:red">C</span>    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Pyrrolysine   </td><td class="markdownTableBodyNone">Pyl   </td><td class="markdownTableBodyNone">O   </td><td class="markdownTableBodyNone"><span style="color:red">K</span>   </td><td class="markdownTableBodyNone"><span style="color:red">K</span>   </td><td class="markdownTableBodyNone"><span style="color:red">K</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Asparagine or aspartic acid   </td><td class="markdownTableBodyNone">Asx   </td><td class="markdownTableBodyNone">B   </td><td class="markdownTableBodyNone"><span style="color:red">D</span>   </td><td class="markdownTableBodyNone">B   </td><td class="markdownTableBodyNone">B    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Glutamine or glutamic acid   </td><td class="markdownTableBodyNone">Glx   </td><td class="markdownTableBodyNone">Z   </td><td class="markdownTableBodyNone"><span style="color:red">E</span>   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>   </td><td class="markdownTableBodyNone"><span style="color:red">B</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Leucine or Isoleucine   </td><td class="markdownTableBodyNone">Xle   </td><td class="markdownTableBodyNone">J   </td><td class="markdownTableBodyNone"><span style="color:red">L</span>   </td><td class="markdownTableBodyNone"><span style="color:red">I</span>   </td><td class="markdownTableBodyNone">J    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyNone">Unknown   </td><td class="markdownTableBodyNone">Xaa   </td><td class="markdownTableBodyNone">X   </td><td class="markdownTableBodyNone"><span style="color:red">S</span>   </td><td class="markdownTableBodyNone"><span style="color:red">S</span>   </td><td class="markdownTableBodyNone"><span style="color:red">A</span>    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyNone">Stop Codon   </td><td class="markdownTableBodyNone">N/A   </td><td class="markdownTableBodyNone">*   </td><td class="markdownTableBodyNone"><span style="color:red">W</span>   </td><td class="markdownTableBodyNone"><span style="color:red">F</span>   </td><td class="markdownTableBodyNone"><span style="color:red">F</span>   </td></tr>
</table>
<p>All amino acid alphabets provide static value members (like an enum) for all amino acids in the form of the one-letter representation. As shown above, alphabets smaller than 27 internally represent multiple amino acids as one.<br  />
For most cases it is highly recommended to use <a class="el" href="classbio_1_1aa27.html" title="The twenty-seven letter amino acid alphabet.">bio::aa27</a> as <a class="el" href="classbio_1_1aa20.html" title="The canonical amino acid alphabet.">bio::aa20</a> provides no benefits in regard to space consumption (both need 5bits). Use it only when you know you need to interface with other software of formats that only support the canonical set. </p>
<h2 class="groupheader">Function Documentation</h2>
<a id="gaca1435b57fb28ffa65862d911e23a960"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaca1435b57fb28ffa65862d911e23a960">&#9670;&nbsp;</a></span>translate_triplet() <span class="overload">[1/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;genetic_code gc = genetic_code::CANONICAL, nucleotide_alphabet nucl_type&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a> bio::translate_triplet </td>
          <td>(</td>
          <td class="paramtype">nucl_type const &amp;&#160;</td>
          <td class="paramname"><em>n1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">nucl_type const &amp;&#160;</td>
          <td class="paramname"><em>n2</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">nucl_type const &amp;&#160;</td>
          <td class="paramname"><em>n3</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Translate one nucleotide triplet into single amino acid (single nucleotide interface). </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">nucl_type</td><td>The type of input nucleotides. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">n1</td><td>First nucleotide in triplet. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">n2</td><td>Second nucleotide in triplet. </td></tr>
    <tr><td class="paramdir">[in]</td><td class="paramname">n3</td><td>Third nucleotide in triplet.</td></tr>
  </table>
  </dd>
</dl>
<p>Translates single nucleotides into amino acid according to given genetic code.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>No-throw guarantee. </p>

</div>
</div>
<a id="ga056c30dd6a4534f5ad8395b146a7d2b3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga056c30dd6a4534f5ad8395b146a7d2b3">&#9670;&nbsp;</a></span>translate_triplet() <span class="overload">[2/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;genetic_code gc = genetic_code::CANONICAL, std::ranges::input_range range_type&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a> bio::translate_triplet </td>
          <td>(</td>
          <td class="paramtype">range_type &amp;&amp;&#160;</td>
          <td class="paramname"><em>input_range</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">constexpr</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Translate one nucleotide triplet into single amino acid (range interface). </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">range_type</td><td>Type of input_range; must satisfy std::ranges::input_range. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">input_range</td><td>Range of three nucleotides that should be converted to amino acid.</td></tr>
  </table>
  </dd>
</dl>
<p>Translates range with 3 nucleotides into amino acid according to given genetic code.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Strong exception guarantee (never modifies data).</p>
<dl class="deprecated"><dt><b><a class="el" href="deprecated.html#_deprecated000002">Deprecated:</a></b></dt><dd>Use <a class="el" href="group__aminoacid.html#gaca1435b57fb28ffa65862d911e23a960" title="Translate one nucleotide triplet into single amino acid (single nucleotide interface).">bio::translate_triplet(nucl_type const &amp; n1, nucl_type const &amp; n2, nucl_type const &amp; n3)</a> instead. </dd></dl>

</div>
</div>
<a id="ga594a06e9c7719f77a3828239c149492a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga594a06e9c7719f77a3828239c149492a">&#9670;&nbsp;</a></span>translate_triplet() <span class="overload">[3/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;genetic_code gc = genetic_code::CANONICAL, std::ranges::random_access_range rng_t&gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a> bio::translate_triplet </td>
          <td>(</td>
          <td class="paramtype">rng_t &amp;&amp;&#160;</td>
          <td class="paramname"><em>input_range</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">constexpr</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Translate one nucleotide triplet into single amino acid (range interface, input range allows random access). </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">rng_t</td><td>Type of input_range; must satisfy std::ranges::random_access_range. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">input_range</td><td>Range of three nucleotides that should be converted to amino acid.</td></tr>
  </table>
  </dd>
</dl>
<p>Translates range with 3 nucleotides into amino acid according to given genetic code.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>Strong exception guarantee (never modifies data).</p>
<dl class="deprecated"><dt><b><a class="el" href="deprecated.html#_deprecated000003">Deprecated:</a></b></dt><dd>Use <a class="el" href="group__aminoacid.html#gaca1435b57fb28ffa65862d911e23a960" title="Translate one nucleotide triplet into single amino acid (single nucleotide interface).">bio::translate_triplet(nucl_type const &amp; n1, nucl_type const &amp; n2, nucl_type const &amp; n3)</a> instead. </dd></dl>

</div>
</div>
<a id="gaac5a13cc5cb827eb7df84b6216268a73"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaac5a13cc5cb827eb7df84b6216268a73">&#9670;&nbsp;</a></span>translate_triplet() <span class="overload">[4/4]</span></h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;genetic_code gc = genetic_code::CANONICAL, typename tuple_type &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr <a class="el" href="classbio_1_1aa27.html">aa27</a> bio::translate_triplet </td>
          <td>(</td>
          <td class="paramtype">tuple_type const &amp;&#160;</td>
          <td class="paramname"><em>input_tuple</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">constexpr</span><span class="mlabel">noexcept</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Translate one nucleotide triplet into single amino acid (tuple interface). </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">tuple_type</td><td>Type of <code>input_tuple</code>. Usually <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/tuple.html">std::tuple</a>, but similar types like <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/array.html">std::array</a> and <a class="el" href="structbio_1_1pod__tuple.html" title="cond">bio::pod_tuple</a> are also supported. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[in]</td><td class="paramname">input_tuple</td><td>Triplet of nucleotides that should be converted to amino acid.</td></tr>
  </table>
  </dd>
</dl>
<p>Translates <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/tuple.html">std::tuple</a> or <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/array.html">std::array</a> with 3 nucleotides into amino acid according to given genetic code.</p>
<h3>Complexity</h3>
<p>Constant.</p>
<h3>Exceptions</h3>
<p>No-throw guarantee.</p>
<dl class="deprecated"><dt><b><a class="el" href="deprecated.html#_deprecated000001">Deprecated:</a></b></dt><dd>Use <a class="el" href="group__aminoacid.html#gaca1435b57fb28ffa65862d911e23a960" title="Translate one nucleotide triplet into single amino acid (single nucleotide interface).">bio::translate_triplet(nucl_type const &amp; n1, nucl_type const &amp; n2, nucl_type const &amp; n3)</a> instead. </dd></dl>

</div>
</div>
<h2 class="groupheader">Variable Documentation</h2>
<a id="ga32d1a56e5ebf85b772d5af0f471a4974"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga32d1a56e5ebf85b772d5af0f471a4974">&#9670;&nbsp;</a></span>enable_aminoacid</h2>

<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename t &gt; </div>
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr bool bio::enable_aminoacid = detail::adl_only::enable_aminoacid_dispatcher::dispatch&lt;<a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/remove_cvref.html">std::remove_cvref_t</a>&lt;t&gt;&gt;()</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>A trait that indicates whether a type shall model <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a>. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">t</td><td>Type of the argument.</td></tr>
  </table>
  </dd>
</dl>
<p>This is an auxiliary trait that is checked by <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a> to verify that a type is an amino acid. This trait should never be read from, instead use <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a>. However, user-defined alphabets that want to model <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a> need to make sure that it evaluates to <code>true</code> for their type.</p>
<h3>Specialisation</h3>
<p>Do not specialise this trait directly. It acts as a wrapper and looks for two possible implementations (in this order):</p>
<ol type="1">
<li>A <code>static</code> member variable <code>enable_aminoacid</code> of the class <code><a class="el" href="structbio_1_1custom_1_1alphabet.html" title="A type that can be specialised to provide customisation point implementations so that third party typ...">bio::custom::alphabet</a>&lt;t&gt;</code>.</li>
<li>A free function <code>constexpr bool enable_aminoacid(t) noexcept</code> in the namespace of your type (or as <code>friend</code>).</li>
</ol>
<p>If none of these is found, the default value is defined as:</p>
<ul>
<li><code>true</code> if the type inherits from <a class="el" href="structbio_1_1aminoacid__empty__base.html" title="This is an empty base class that can be inherited by types that shall model bio::aminoacid_alphabet.">bio::aminoacid_empty_base</a> (or <a class="el" href="classbio_1_1aminoacid__base.html" title="A CRTP-base that refines bio::alphabet_base and is used by the amino acids.">bio::aminoacid_base</a>),</li>
<li><code>false</code> otherwise.</li>
</ul>
<p>Implementations of 1. and 2. are required to be marked <code>constexpr</code> and the value / return value must be convertible to <code>bool</code>. Implementations of 2. are required to be marked <code>noexcept</code>. The value passed to functions implementing 2. shall be ignored, it is only used for finding the function via argument-dependent lookup. In case that your type is not bio::is_constexpr_default_constructible_v and you wish to provide an implementation for 2., instead overload for <code><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/type_identity.html">std::type_identity</a>&lt;t&gt;</code>.</p>
<p>To make a type model <a class="el" href="interfacebio_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::aminoacid_alphabet</a>, it is recommended that you derive from <a class="el" href="classbio_1_1aminoacid__base.html" title="A CRTP-base that refines bio::alphabet_base and is used by the amino acids.">bio::aminoacid_base</a>. If that is not possible, choose option 2., and only implement option 1. as a last resort.</p>
<h3>Example</h3>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="alphabet_2aminoacid_2concept_8hpp.html">bio/alphabet/aminoacid/concept.hpp</a>&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">namespace </span>your_namespace</div>
<div class="line">{</div>
<div class="line"> </div>
<div class="line"><span class="comment">// your own aminoacid definition</span></div>
<div class="line"><span class="keyword">struct </span>your_aa : <a class="code" href="structbio_1_1aminoacid__empty__base.html">bio::aminoacid_empty_base</a></div>
<div class="line">{</div>
<div class="line">    <span class="comment">//...</span></div>
<div class="line">};</div>
<div class="line"> </div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line">static_assert(bio::enable_aminoacid&lt;your_namespace::your_aa&gt; == <span class="keyword">true</span>);</div>
<div class="line"> </div>
<div class="line"><span class="comment">/***** OR *****/</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">namespace </span>your_namespace2</div>
<div class="line">{</div>
<div class="line"> </div>
<div class="line"><span class="comment">// your own aminoacid definition</span></div>
<div class="line"><span class="keyword">struct </span>your_aa</div>
<div class="line">{</div>
<div class="line">    <span class="comment">//...</span></div>
<div class="line">};</div>
<div class="line"> </div>
<div class="line">constexpr <span class="keywordtype">bool</span> <a class="code" href="group__aminoacid.html#ga32d1a56e5ebf85b772d5af0f471a4974">enable_aminoacid</a>(your_aa) noexcept</div>
<div class="line">{</div>
<div class="line">    <span class="keywordflow">return</span> <span class="keyword">true</span>;</div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line">}</div>
<div class="line"> </div>
<div class="line">static_assert(bio::enable_aminoacid&lt;your_namespace2::your_aa&gt; == <span class="keyword">true</span>);</div>
<div class="ttc" id="aalphabet_2aminoacid_2concept_8hpp_html"><div class="ttname"><a href="alphabet_2aminoacid_2concept_8hpp.html">concept.hpp</a></div><div class="ttdoc">Provides bio::aminoacid_alphabet.</div></div>
<div class="ttc" id="agroup__aminoacid_html_ga32d1a56e5ebf85b772d5af0f471a4974"><div class="ttname"><a href="group__aminoacid.html#ga32d1a56e5ebf85b772d5af0f471a4974">bio::enable_aminoacid</a></div><div class="ttdeci">constexpr bool enable_aminoacid</div><div class="ttdoc">A trait that indicates whether a type shall model bio::aminoacid_alphabet.</div><div class="ttdef"><b>Definition:</b> concept.hpp:124</div></div>
<div class="ttc" id="astructbio_1_1aminoacid__empty__base_html"><div class="ttname"><a href="structbio_1_1aminoacid__empty__base.html">bio::aminoacid_empty_base</a></div><div class="ttdoc">This is an empty base class that can be inherited by types that shall model bio::aminoacid_alphabet.</div><div class="ttdef"><b>Definition:</b> concept.hpp:34</div></div>
</div><!-- fragment --><h3>Customisation point</h3>
<p>This is a customisation point (see <a class="el" href="biocpp_customisation.html">Customisation</a>). To change the default behaviour for your own alphabet, follow the above instructions. </p>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="footer">Generated on Sun Jun 12 2022 17:23:28 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
