<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: Quality</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">0.1.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__quality.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#typedef-members">Typedefs</a>  </div>
  <div class="headertitle">
<div class="title">Quality<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Provides the various quality score types.  
<a href="#details">More...</a></p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Quality:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__quality.svg" width="208" height="36"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1phred42.html">bio::phred42</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quality type for traditional Sanger and modern Illumina Phred scores (typical range).  <a href="classbio_1_1phred42.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1phred63.html">bio::phred63</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quality type for traditional Sanger and modern Illumina Phred scores (full range).  <a href="classbio_1_1phred63.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1phred68legacy.html">bio::phred68legacy</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quality type for Solexa and deprecated Illumina formats.  <a href="classbio_1_1phred68legacy.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1qualified.html">bio::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Joins an arbitrary alphabet with a quality alphabet.  <a href="classbio_1_1qualified.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">interface &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="interfacebio_1_1quality__alphabet.html">quality_alphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether an alphabet represents quality scores.  <a href="interfacebio_1_1quality__alphabet.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1quality__base.html">bio::quality_base&lt; derived_type, size &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet_base</a> and is used by the quality alphabets.  <a href="classbio_1_1quality__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">interface &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="interfacebio_1_1writable__quality__alphabet.html">writable_quality_alphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether a writable alphabet represents quality scores.  <a href="interfacebio_1_1writable__quality__alphabet.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:gaec6fd89f46ef56795d66568184bdba0a"><td class="memTemplParams" colspan="2"><a id="gaec6fd89f46ef56795d66568184bdba0a"></a>
template&lt;typename alphabet_type &gt; </td></tr>
<tr class="memitem:gaec6fd89f46ef56795d66568184bdba0a"><td class="memTemplItemLeft" align="right" valign="top">using&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__quality.html#gaec6fd89f46ef56795d66568184bdba0a">bio::alphabet_phred_t</a> = decltype(<a class="el" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445">bio::to_phred</a>(std::declval&lt; alphabet_type &gt;()))</td></tr>
<tr class="memdesc:gaec6fd89f46ef56795d66568184bdba0a"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <code>phred_type</code> of the alphabet; defined as the return type of <a class="el" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445" title="The public getter function for the phred representation of a quality score.">bio::to_phred</a>. <br /></td></tr>
<tr class="separator:gaec6fd89f46ef56795d66568184bdba0a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Function objects (Quality)</h2></td></tr>
<tr class="memitem:ga0a0a566ce2f93ff9cb4f556b0a3a2445"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445">bio::to_phred</a> = detail::adl_only::to_phred_fn{}</td></tr>
<tr class="memdesc:ga0a0a566ce2f93ff9cb4f556b0a3a2445"><td class="mdescLeft">&#160;</td><td class="mdescRight">The public getter function for the phred representation of a quality score.  <a href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445">More...</a><br /></td></tr>
<tr class="separator:ga0a0a566ce2f93ff9cb4f556b0a3a2445"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab0eb4bc93da56673fb1e3e1f8994c0b3"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__quality.html#gab0eb4bc93da56673fb1e3e1f8994c0b3">bio::assign_phred_to</a> = detail::adl_only::assign_phred_to_fn{}</td></tr>
<tr class="memdesc:gab0eb4bc93da56673fb1e3e1f8994c0b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign a phred score to a quality alphabet object.  <a href="group__quality.html#gab0eb4bc93da56673fb1e3e1f8994c0b3">More...</a><br /></td></tr>
<tr class="separator:gab0eb4bc93da56673fb1e3e1f8994c0b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Provides the various quality score types. </p>
<h3>Introduction</h3>
<p>Quality score sequences are usually output together with the DNA (or RNA) sequence by sequencing machines like the Illumina Genome Analyzer. The quality score of a nucleotide is also known as Phred score and is an integer score being inversely proportional to the propability <img class="formulaInl" alt="$p$" src="form_0.png" width="11" height="12"/> that a base call is <b>incorrect</b>. Which roughly means that the higher a Phred score is, the higher is the probabality that the corresponding nucleotide is correct for that position. There exists two common variants of its computation:</p><ul>
<li>Salinger format with <img class="formulaInl" alt="$Q = -10log_{10}(p)$" src="form_1.png" width="126" height="19"/></li>
<li>Solexa format with <img class="formulaInl" alt="$Q = -10log_{10}(p/(1-p))$" src="form_2.png" width="187" height="19"/> Thus, despite implicit conversion between different quality types is supported, for very low quality levels the scores vary significantly and need to be corrected by an offset before being compared. The Phred score range does not fit into one digit, and is therefore mapped to ASCII characters. Depending on the format and the analyzer machine generation, the mappings can differ. Output files storing DNA sequences and their quality scores are usually stored in the <b>FASTQ</b> format indicated by the file endings <em>fastq</em> or <em>fq</em>. This sub-module provides multiple quality alphabets that can be used in combination with regular containers and ranges.</li>
</ul>
<h3>Encoding Schemes</h3>
<table class="markdownTable">
<tr class="markdownTableHead">
<th class="markdownTableHeadCenter">Format   </th><th class="markdownTableHeadLeft">Quality Type   </th><th class="markdownTableHeadCenter">Phred Score Range   </th><th class="markdownTableHeadCenter">Rank Range   </th><th class="markdownTableHeadCenter">ASCII Range   </th><th class="markdownTableHeadCenter">Assert    </th></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">Sanger, Illumina 1.8+ short   </td><td class="markdownTableBodyLeft"><a class="el" href="classbio_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range).">bio::phred42</a>   </td><td class="markdownTableBodyCenter">[0 .. 41]   </td><td class="markdownTableBodyCenter">[0 .. 41]   </td><td class="markdownTableBodyCenter">['!' .. 'J']   </td><td class="markdownTableBodyCenter">Phred score in [0 .. 61]    </td></tr>
<tr class="markdownTableRowEven">
<td class="markdownTableBodyCenter">Sanger, Illumina 1.8+ long   </td><td class="markdownTableBodyLeft"><a class="el" href="classbio_1_1phred63.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (full range).">bio::phred63</a>   </td><td class="markdownTableBodyCenter">[0 .. 62]   </td><td class="markdownTableBodyCenter">[0 .. 62]   </td><td class="markdownTableBodyCenter">['!' .. '_']   </td><td class="markdownTableBodyCenter">Phred score in [0 .. 62]    </td></tr>
<tr class="markdownTableRowOdd">
<td class="markdownTableBodyCenter">Solexa, Illumina [1.0; 1.8[   </td><td class="markdownTableBodyLeft"><a class="el" href="classbio_1_1phred68legacy.html" title="Quality type for Solexa and deprecated Illumina formats.">bio::phred68legacy</a>   </td><td class="markdownTableBodyCenter">[-5 .. 62]   </td><td class="markdownTableBodyCenter">[0 .. 67]   </td><td class="markdownTableBodyCenter">[';' .. '~']   </td><td class="markdownTableBodyCenter">Phred score in [-5 .. 62]   </td></tr>
</table>
<p>The most distributed format is the <em>Sanger</em> or <em>Illumina 1.8+</em> format. Despite typical Phred scores for Illumina machines range from 0 to maximal 41, it is possible that processed reads reach higher scores. If you don't intend to handle Phred scores larger than 41, we recommend to use <a class="el" href="classbio_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range).">bio::phred42</a> due to its more space efficient implementation. For other formats, like Solexa and Illumina 1.0 to 1.7 the type <a class="el" href="classbio_1_1phred68legacy.html" title="Quality type for Solexa and deprecated Illumina formats.">bio::phred68legacy</a> is provided. To cover also the Solexa format, the Phred score is stored as a <b>signed</b> integer starting at -5. An overview of all the score formats and their encodings can be found here: <a href="https://en.wikipedia.org/wiki/FASTQ_format#Encoding">https://en.wikipedia.org/wiki/FASTQ_format#Encoding</a>.</p>
<h3>Concept</h3>
<p>The quality submodule defines the <a class="el" href="interfacebio_1_1writable__quality__alphabet.html" title="A concept that indicates whether a writable alphabet represents quality scores.">bio::writable_quality_alphabet</a> which encompasses all the alphabets, defined in the submodule, and refines the <a class="el" href="interfacebio_1_1writable__alphabet.html" title="Refines bio::alphabet and adds assignability.">bio::writable_alphabet</a> by providing Phred score assignment and conversion operations. Additionally, this submodule defines the <a class="el" href="interfacebio_1_1quality__alphabet.html" title="A concept that indicates whether an alphabet represents quality scores.">bio::quality_alphabet</a>, which only requires readablity and not assignability.</p>
<h3>Assignment and Conversion</h3>
<p>Quality alphabets can be converted to their char and rank representation via <code><a class="el" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8" title="Return the char representation of an alphabet object.">bio::to_char</a></code> and <code><a class="el" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0" title="Return the rank representation of a (semi-)alphabet object.">bio::to_rank</a></code> respectively (like all other alphabets). Additionally they can be converted to their Phred representation via <code><a class="el" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445" title="The public getter function for the phred representation of a quality score.">bio::to_phred</a></code>.</p>
<p>Likewise, assignment happens via <code><a class="el" href="group__alphabet.html#ga05da42eb0b9a48f4d1a554c7d3a9044d" title="Assign a character to an alphabet object.">bio::assign_char_to</a></code>, <code><a class="el" href="group__alphabet.html#gae34fa2b37a8b40f2f2fa64fed7bbb410" title="Assign a rank to an alphabet object.">bio::assign_rank_to</a></code> and <code><a class="el" href="group__quality.html#gab0eb4bc93da56673fb1e3e1f8994c0b3" title="Assign a phred score to a quality alphabet object.">bio::assign_phred_to</a></code>. Phred values outside the representable range, but inside the legal range, are converted to the closest Phred score, e.g. assigning 60 to a <code><a class="el" href="classbio_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range).">bio::phred42</a></code> will result in a Phred score of 41. Assigning Phred values outside the legal range results in undefined behaviour.</p>
<p>All quality alphabets are explicitly convertible to each other via their Phred representation. Values not present in one alphabet are mapped to the closest value in the target alphabet (e.g. a <code><a class="el" href="classbio_1_1phred63.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (full range).">bio::phred63</a></code> letter with value 60 will convert to a <code><a class="el" href="classbio_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range).">bio::phred42</a></code> letter of score 41). </p>
<h2 class="groupheader">Variable Documentation</h2>
<a id="gab0eb4bc93da56673fb1e3e1f8994c0b3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab0eb4bc93da56673fb1e3e1f8994c0b3">&#9670;&nbsp;</a></span>assign_phred_to</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr auto bio::assign_phred_to = detail::adl_only::assign_phred_to_fn{}</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Assign a phred score to a quality alphabet object. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">your_type</td><td>The type of the target object. Must model the <a class="el" href="interfacebio_1_1quality__alphabet.html" title="A concept that indicates whether an alphabet represents quality scores.">bio::quality_alphabet</a>. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chr</td><td>The phred score being assigned; must be of the <a class="el" href="group__quality.html#gaec6fd89f46ef56795d66568184bdba0a" title="The phred_type of the alphabet; defined as the return type of bio::to_phred.">bio::alphabet_phred_t</a> of the target object. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Reference to <code>alph</code> if <code>alph</code> was given as lvalue, otherwise a copy.</dd></dl>
<p>This is a function object. Invoke it with the parameter(s) specified above.</p>
<p>It acts as a wrapper and looks for three possible implementations (in this order):</p>
<ol type="1">
<li>A static member function <code>assign_phred_to(phred_type const chr, your_type &amp; a)</code> of the class <code><a class="el" href="structbio_1_1custom_1_1alphabet.html" title="A type that can be specialised to provide customisation point implementations so that third party typ...">bio::custom::alphabet</a>&lt;your_type&gt;</code>.</li>
<li>A free function <code>assign_phred_to(phred_type const chr, your_type &amp; a)</code> in the namespace of your type (or as <code>friend</code>).</li>
<li>A member function called <code>assign_phred(phred_type const chr)</code> (not <code>assign_phred_to</code>).</li>
</ol>
<p>Functions are only considered for one of the above cases if they are marked <code>noexcept</code> (<code>constexpr</code> is not required, but recommended) and if the returned type is <code>your_type &amp;</code>.</p>
<p>Every writable quality alphabet type must provide one of the above. <em>Note</em> that temporaries of <code>your_type</code> are handled by this function object and <b>do not</b> require an additional overload.</p>
<h3>Customisation point</h3>
<p>This is a customisation point (see <a class="el" href="biocpp_customisation.html">Customisation</a>). To specify the behaviour for your own alphabet type, simply provide one of the three functions specified above. </p>

</div>
</div>
<a id="ga0a0a566ce2f93ff9cb4f556b0a3a2445"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga0a0a566ce2f93ff9cb4f556b0a3a2445">&#9670;&nbsp;</a></span>to_phred</h2>

<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">constexpr auto bio::to_phred = detail::adl_only::to_phred_fn{}</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">constexpr</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>The public getter function for the phred representation of a quality score. </p>
<dl class="tparams"><dt>Template Parameters</dt><dd>
  <table class="tparams">
    <tr><td class="paramname">your_type</td><td>The type of alphabet. Must model the <a class="el" href="interfacebio_1_1quality__alphabet.html" title="A concept that indicates whether an alphabet represents quality scores.">bio::quality_alphabet</a>. </td></tr>
  </table>
  </dd>
</dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">chr</td><td>The quality value to convert into the phred score. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the phred representation of a quality score.</dd></dl>
<p>This is a function object. Invoke it with the parameter(s) specified above.</p>
<p>It acts as a wrapper and looks for three possible implementations (in this order):</p>
<ol type="1">
<li>A static member function <code>to_phred(your_type const a)</code> of the class <code><a class="el" href="structbio_1_1custom_1_1alphabet.html" title="A type that can be specialised to provide customisation point implementations so that third party typ...">bio::custom::alphabet</a>&lt;your_type&gt;</code>.</li>
<li>A free function <code>to_phred(your_type const a)</code> in the namespace of your type (or as <code>friend</code>).</li>
<li>A member function called <code><a class="el" href="group__quality.html#ga0a0a566ce2f93ff9cb4f556b0a3a2445" title="The public getter function for the phred representation of a quality score.">to_phred()</a></code>.</li>
</ol>
<p>Functions are only considered for one of the above cases if they are marked <code>noexcept</code> (<code>constexpr</code> is not required, but recommended) and if the returned type is convertible to <code>size_t</code>.</p>
<p>Every quality alphabet type must provide one of the above.</p>
<h3>Customisation point</h3>
<p>This is a customisation point (see <a class="el" href="biocpp_customisation.html">Customisation</a>). To specify the behaviour for your own alphabet type, simply provide one of the three functions specified above. </p>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="footer">Generated on Sun Jun 12 2022 17:23:28 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
