<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: Class Hierarchy</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">0.5.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('hierarchy.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="headertitle">
<div class="title">Class Hierarchy</div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock">
<p><a href="inherits.html">Go to the graphical class hierarchy</a></p>
This inheritance list is sorted roughly, but not completely, alphabetically:</div><div class="directory">
<div class="levels">[detail level <span onclick="javascript:toggleLevel(1);">1</span><span onclick="javascript:toggleLevel(2);">2</span><span onclick="javascript:toggleLevel(3);">3</span><span onclick="javascript:toggleLevel(4);">4</span><span onclick="javascript:toggleLevel(5);">5</span><span onclick="javascript:toggleLevel(6);">6</span>]</div><table class="directory">
<tr id="row_0_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_0_" class="arrow" onclick="toggleFolder('0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>bio::ranges::detail::adaptor_base&lt; deep&lt; underlying_adaptor_t &gt;, underlying_adaptor_t &gt;</b></td><td class="desc"></td></tr>
<tr id="row_0_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1views_1_1deep.html" target="_self">bio::ranges::views::deep&lt; underlying_adaptor_t &gt;</a></td><td class="desc">A wrapper type around an existing view adaptor that enables "deep view" behaviour for that view </td></tr>
<tr id="row_1_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1aligned__allocator.html" target="_self">bio::ranges::aligned_allocator&lt; value_t, alignment_v &gt;</a></td><td class="desc">Allocates uninitialized storage whose memory-alignment is specified by <em>alignment</em> </td></tr>
<tr id="row_2_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_2_" class="arrow" onclick="toggleFolder('2_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; derived_type, size, char_t &gt;</a></td><td class="desc">A CRTP-base that makes defining a custom alphabet easier </td></tr>
<tr id="row_2_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_0_" class="arrow" onclick="toggleFolder('2_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa27, 27 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa27.html" target="_self">bio::alphabet::aa27</a></td><td class="desc">The twenty-seven letter amino acid alphabet </td></tr>
<tr id="row_2_1_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_1_" class="arrow" onclick="toggleFolder('2_1_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa20, 20 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_1_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa20.html" target="_self">bio::alphabet::aa20</a></td><td class="desc">The canonical amino acid alphabet </td></tr>
<tr id="row_2_2_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_2_" class="arrow" onclick="toggleFolder('2_2_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa10murphy, 10 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_2_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html" target="_self">bio::alphabet::aa10murphy</a></td><td class="desc">The reduced Murphy amino acid alphabet </td></tr>
<tr id="row_2_3_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_3_" class="arrow" onclick="toggleFolder('2_3_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa10li, 10 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_3_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10li.html" target="_self">bio::alphabet::aa10li</a></td><td class="desc">The reduced Li amino acid alphabet </td></tr>
<tr id="row_2_4_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_4_" class="arrow" onclick="toggleFolder('2_4_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; rna5, 5 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_4_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna5.html" target="_self">bio::alphabet::rna5</a></td><td class="desc">The five letter RNA alphabet of A,C,G,U and the unknown character N </td></tr>
<tr id="row_2_5_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_5_" class="arrow" onclick="toggleFolder('2_5_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; rna4, 4 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_5_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna4.html" target="_self">bio::alphabet::rna4</a></td><td class="desc">The four letter RNA alphabet of A,C,G,U </td></tr>
<tr id="row_2_6_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_6_" class="arrow" onclick="toggleFolder('2_6_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; dna16sam, 16 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_6_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" target="_self">bio::alphabet::dna16sam</a></td><td class="desc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=') </td></tr>
<tr id="row_2_7_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_7_" class="arrow" onclick="toggleFolder('2_7_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; dna15, 15 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_7_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna15.html" target="_self">bio::alphabet::dna15</a></td><td class="desc">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_2_8_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_8_" class="arrow" onclick="toggleFolder('2_8_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; rna15, 15 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_8_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna15.html" target="_self">bio::alphabet::rna15</a></td><td class="desc">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_2_9_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_9_" class="arrow" onclick="toggleFolder('2_9_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; dna5, 5 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_9_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna5.html" target="_self">bio::alphabet::dna5</a></td><td class="desc">The five letter DNA alphabet of A,C,G,T and the unknown character N </td></tr>
<tr id="row_2_10_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_10_" class="arrow" onclick="toggleFolder('2_10_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; dna4, 4 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_10_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna4.html" target="_self">bio::alphabet::dna4</a></td><td class="desc">The four letter DNA alphabet of A,C,G,T </td></tr>
<tr id="row_2_11_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_11_" class="arrow" onclick="toggleFolder('2_11_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1quality__base.html" target="_self">bio::alphabet::quality_base&lt; phred42, 42 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_11_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred42.html" target="_self">bio::alphabet::phred42</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (typical range) </td></tr>
<tr id="row_2_12_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_12_" class="arrow" onclick="toggleFolder('2_12_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1quality__base.html" target="_self">bio::alphabet::quality_base&lt; phred63, 63 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_12_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred63.html" target="_self">bio::alphabet::phred63</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (full range) </td></tr>
<tr id="row_2_13_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_13_" class="arrow" onclick="toggleFolder('2_13_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1quality__base.html" target="_self">bio::alphabet::quality_base&lt; phred68legacy, 68 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_2_13_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred68legacy.html" target="_self">bio::alphabet::phred68legacy</a></td><td class="desc">Quality type for Solexa and deprecated Illumina formats </td></tr>
<tr id="row_3_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_3_" class="arrow" onclick="toggleFolder('3_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; alphabet_variant&lt; alternative_types... &gt;,(static_cast&lt; size_t &gt;(size&lt; alternative_types &gt;)+...), char &gt;</a></td><td class="desc"></td></tr>
<tr id="row_3_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__variant.html" target="_self">bio::alphabet::alphabet_variant&lt; alternative_types &gt;</a></td><td class="desc">A combined alphabet that can hold values of either of its alternatives </td></tr>
<tr id="row_4_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_4_" class="arrow" onclick="toggleFolder('4_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; cigar,(1 *... *size&lt; component_types &gt;), void &gt;</a></td><td class="desc"></td></tr>
<tr id="row_4_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_4_0_" class="arrow" onclick="toggleFolder('4_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; cigar, uint32_t, cigar_op &gt;</a></td><td class="desc"></td></tr>
<tr id="row_4_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar.html" target="_self">bio::alphabet::cigar</a></td><td class="desc">The cigar semialphabet pairs a counter with a <a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" title="The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,=.">bio::alphabet::cigar_op</a> letter </td></tr>
<tr id="row_5_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_5_" class="arrow" onclick="toggleFolder('5_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; cigar_op, 9, char &gt;</a></td><td class="desc"></td></tr>
<tr id="row_5_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" target="_self">bio::alphabet::cigar_op</a></td><td class="desc">The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,= </td></tr>
<tr id="row_6_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base_3_01derived__type_00_011ul_00_01char__t_01_4.html" target="_self">bio::alphabet::alphabet_base&lt; derived_type, 1ul, char_t &gt;</a></td><td class="desc">Specialisation of <a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::alphabet_base</a> for alphabets of size 1 </td></tr>
<tr id="row_7_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_7_" class="arrow" onclick="toggleFolder('7_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; derived_type, size, char &gt;</a></td><td class="desc"></td></tr>
<tr id="row_7_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; derived_type, size &gt;</a></td><td class="desc">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::alphabet_base</a> and is used by the amino acids </td></tr>
<tr id="row_7_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html" target="_self">bio::alphabet::nucleotide_base&lt; derived_type, size &gt;</a></td><td class="desc">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::alphabet_base</a> and is used by the nucleotides </td></tr>
<tr id="row_7_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1quality__base.html" target="_self">bio::alphabet::quality_base&lt; derived_type, size &gt;</a></td><td class="desc">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::alphabet_base</a> and is used by the quality alphabets </td></tr>
<tr id="row_8_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_8_" class="arrow" onclick="toggleFolder('8_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; derived_type,(1 *... *size&lt; component_types &gt;), void &gt;</a></td><td class="desc"></td></tr>
<tr id="row_8_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; derived_type, component_types &gt;</a></td><td class="desc">The CRTP base for a combined alphabet that contains multiple values of different alphabets at the same time </td></tr>
<tr id="row_9_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_9_" class="arrow" onclick="toggleFolder('9_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; gap, 1ull, char &gt;</a></td><td class="desc"></td></tr>
<tr id="row_9_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1gap.html" target="_self">bio::alphabet::gap</a></td><td class="desc">The alphabet of a gap character '-' </td></tr>
<tr id="row_10_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_10_" class="arrow" onclick="toggleFolder('10_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; mask, 2, void &gt;</a></td><td class="desc"></td></tr>
<tr id="row_10_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1mask.html" target="_self">bio::alphabet::mask</a></td><td class="desc">Implementation of a masked alphabet to be used for tuple composites </td></tr>
<tr id="row_11_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_11_" class="arrow" onclick="toggleFolder('11_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; masked&lt; sequence_alphabet_t &gt;,(1 *... *size&lt; component_types &gt;), void &gt;</a></td><td class="desc"></td></tr>
<tr id="row_11_0_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_11_0_" class="arrow" onclick="toggleFolder('11_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; masked&lt; sequence_alphabet_t &gt;, sequence_alphabet_t, mask &gt;</a></td><td class="desc"></td></tr>
<tr id="row_11_0_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1masked.html" target="_self">bio::alphabet::masked&lt; sequence_alphabet_t &gt;</a></td><td class="desc">Implementation of a masked composite, which extends a given alphabet with a mask </td></tr>
<tr id="row_12_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_12_" class="arrow" onclick="toggleFolder('12_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;,(1 *... *size&lt; component_types &gt;), void &gt;</a></td><td class="desc"></td></tr>
<tr id="row_12_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_12_0_" class="arrow" onclick="toggleFolder('12_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;, sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc"></td></tr>
<tr id="row_12_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1qualified.html" target="_self">bio::alphabet::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc">Joins an arbitrary alphabet with a quality alphabet </td></tr>
<tr id="row_13_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_13_" class="arrow" onclick="toggleFolder('13_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" target="_self">bio::alphabet::alphabet_base&lt; semialphabet_any&lt; size_ &gt;, size_, void &gt;</a></td><td class="desc"></td></tr>
<tr id="row_13_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1semialphabet__any.html" target="_self">bio::alphabet::semialphabet_any&lt; size_ &gt;</a></td><td class="desc">A semi-alphabet that type erases all other semi-alphabets of the same size </td></tr>
<tr id="row_14_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__proxy.html" target="_self">bio::alphabet::alphabet_proxy&lt; derived_type, alphabet_type &gt;</a></td><td class="desc">A CRTP-base that eases the definition of proxy types returned in place of regular alphabets </td></tr>
<tr id="row_15_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__proxy.html" target="_self">bio::alphabet::alphabet_proxy&lt; component_proxy&lt; alphabet_type, index &gt;, alphabet_type &gt;</a></td><td class="desc"></td></tr>
<tr id="row_16_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__proxy.html" target="_self">bio::alphabet::alphabet_proxy&lt; reference_proxy_type, alphabet_type &gt;</a></td><td class="desc"></td></tr>
<tr id="row_17_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_17_" class="arrow" onclick="toggleFolder('17_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1aminoacid__empty__base.html" target="_self">bio::alphabet::aminoacid_empty_base</a></td><td class="desc">This is an empty base class that can be inherited by types that shall model <a class="el" href="interfacebio_1_1alphabet_1_1aminoacid__alphabet.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::alphabet::aminoacid_alphabet</a> </td></tr>
<tr id="row_17_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa27, 27 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_17_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa20, 20 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_17_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa10murphy, 10 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_17_3_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; aa10li, 10 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_17_4_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html" target="_self">bio::alphabet::aminoacid_base&lt; derived_type, size &gt;</a></td><td class="desc">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1alphabet__base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::alphabet_base</a> and is used by the amino acids </td></tr>
<tr id="row_18_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_18_" class="arrow" onclick="toggleFolder('18_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1arithmetic.html" target="_self">arithmetic</a></td><td class="desc">A type that satisfies std::is_arithmetic_v&lt;t&gt; </td></tr>
<tr id="row_18_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1floating__point.html" target="_self">floating_point</a></td><td class="desc">An arithmetic type that also satisfies std::is_floating_point_v&lt;t&gt; </td></tr>
<tr id="row_19_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1assign__char__to.html" target="_self">bio::alphabet::cpo::assign_char_to</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#ga0c2eb25dabf59ccf9992b2661083ea5b" title="Assign a char to an alphabet object.">bio::alphabet::assign_char_to</a> </td></tr>
<tr id="row_20_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1assign__phred__to.html" target="_self">bio::alphabet::cpo::assign_phred_to</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#ga0c2eb25dabf59ccf9992b2661083ea5b" title="Assign a char to an alphabet object.">bio::alphabet::assign_char_to</a> </td></tr>
<tr id="row_21_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1assign__rank__to.html" target="_self">bio::alphabet::cpo::assign_rank_to</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#gab718eab4e0f0e1b44f78b7600dee5e4d" title="Assign a rank to an alphabet object.">bio::alphabet::assign_rank_to</a>.# </td></tr>
<tr id="row_22_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_22_" class="arrow" onclick="toggleFolder('22_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>bio::cerealisable</b></td><td class="desc"></td></tr>
<tr id="row_22_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_22_0_" class="arrow" onclick="toggleFolder('22_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1small__vector.html" target="_self">bio::ranges::small_vector&lt; char, capacity_+1 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_22_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1small__string.html" target="_self">bio::ranges::small_string&lt; capacity_ &gt;</a></td><td class="desc">Implements a small string that can be used for compile time computations </td></tr>
<tr id="row_22_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1bitcompressed__vector.html" target="_self">bio::ranges::bitcompressed_vector&lt; alphabet_type &gt;</a></td><td class="desc">A space-optimised version of <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a> that compresses multiple letters into a single byte </td></tr>
<tr id="row_22_2_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1dynamic__bitset.html" target="_self">bio::ranges::dynamic_bitset&lt; bit_capacity &gt;</a></td><td class="desc">A constexpr bitset implementation with dynamic size at compile time </td></tr>
<tr id="row_22_3_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1small__vector.html" target="_self">bio::ranges::small_vector&lt; value_type_, capacity_ &gt;</a></td><td class="desc">A constexpr vector implementation with dynamic size at compile time </td></tr>
<tr id="row_23_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1char__is__valid__for.html" target="_self">bio::alphabet::cpo::char_is_valid_for</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#ga0c2eb25dabf59ccf9992b2661083ea5b" title="Assign a char to an alphabet object.">bio::alphabet::assign_char_to</a> </td></tr>
<tr id="row_24_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_24_" class="arrow" onclick="toggleFolder('24_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::ranges::common_range</b></td><td class="desc"></td></tr>
<tr id="row_24_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_24_0_" class="arrow" onclick="toggleFolder('24_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>container</b></td><td class="desc">The (most general) container concept as defined by the standard library </td></tr>
<tr id="row_24_0_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span id="arr_24_0_0_" class="arrow" onclick="toggleFolder('24_0_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>sequence_container</b></td><td class="desc">A more refined container concept than bio::ranges::detail::container </td></tr>
<tr id="row_24_0_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span id="arr_24_0_0_0_" class="arrow" onclick="toggleFolder('24_0_0_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>random_access_container</b></td><td class="desc">A more refined container concept than bio::ranges::detail::sequence_container </td></tr>
<tr id="row_24_0_0_0_0_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span id="arr_24_0_0_0_0_" class="arrow" onclick="toggleFolder('24_0_0_0_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>reservible_container</b></td><td class="desc">A more refined container concept than bio::ranges::detail::random_access_container </td></tr>
<tr id="row_24_0_0_0_0_0_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1small__vector.html" target="_self">bio::ranges::small_vector&lt; char, capacity_+1 &gt;</a></td><td class="desc"></td></tr>
<tr id="row_24_0_0_0_0_1_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1bitcompressed__vector.html" target="_self">bio::ranges::bitcompressed_vector&lt; alphabet_type &gt;</a></td><td class="desc">A space-optimised version of <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a> that compresses multiple letters into a single byte </td></tr>
<tr id="row_24_0_0_0_0_2_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1concatenated__sequences.html" target="_self">bio::ranges::concatenated_sequences&lt; underlying_container_type, data_delimiters_type &gt;</a></td><td class="desc">Container that stores sequences concatenated internally </td></tr>
<tr id="row_24_0_0_0_0_3_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1dynamic__bitset.html" target="_self">bio::ranges::dynamic_bitset&lt; bit_capacity &gt;</a></td><td class="desc">A constexpr bitset implementation with dynamic size at compile time </td></tr>
<tr id="row_24_0_0_0_0_4_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1small__string.html" target="_self">bio::ranges::small_string&lt; capacity_ &gt;</a></td><td class="desc">Implements a small string that can be used for compile time computations </td></tr>
<tr id="row_24_0_0_0_0_5_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1ranges_1_1small__vector.html" target="_self">bio::ranges::small_vector&lt; value_type_, capacity_ &gt;</a></td><td class="desc">A constexpr vector implementation with dynamic size at compile time </td></tr>
<tr id="row_25_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1complement.html" target="_self">bio::alphabet::cpo::complement</a></td><td class="desc">Customisation tag for <a class="el" href="group__nucleotide.html#ga174261f8f888bbce628ffe01097a9b48" title="Return the complement of a nucleotide object.">bio::alphabet::complement</a> </td></tr>
<tr id="row_26_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_26_" class="arrow" onclick="toggleFolder('26_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::copy_constructible</b></td><td class="desc"></td></tr>
<tr id="row_26_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_26_0_" class="arrow" onclick="toggleFolder('26_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1semialphabet.html" target="_self">semialphabet</a></td><td class="desc">The basis for <a class="el" href="interfacebio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a>, but requires only rank interface (not char) </td></tr>
<tr id="row_26_0_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span id="arr_26_0_0_" class="arrow" onclick="toggleFolder('26_0_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1alphabet.html" target="_self">alphabet</a></td><td class="desc">The generic alphabet concept that covers most data types used in ranges </td></tr>
<tr id="row_26_0_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span id="arr_26_0_0_0_" class="arrow" onclick="toggleFolder('26_0_0_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1aminoacid__alphabet.html" target="_self">aminoacid_alphabet</a></td><td class="desc">A concept that indicates whether an alphabet represents amino acids </td></tr>
<tr id="row_26_0_0_0_0_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10li.html" target="_self">bio::alphabet::aa10li</a></td><td class="desc">The reduced Li amino acid alphabet </td></tr>
<tr id="row_26_0_0_0_1_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html" target="_self">bio::alphabet::aa10murphy</a></td><td class="desc">The reduced Murphy amino acid alphabet </td></tr>
<tr id="row_26_0_0_0_2_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa20.html" target="_self">bio::alphabet::aa20</a></td><td class="desc">The canonical amino acid alphabet </td></tr>
<tr id="row_26_0_0_0_3_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa27.html" target="_self">bio::alphabet::aa27</a></td><td class="desc">The twenty-seven letter amino acid alphabet </td></tr>
<tr id="row_26_0_0_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span id="arr_26_0_0_1_" class="arrow" onclick="toggleFolder('26_0_0_1_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1nucleotide__alphabet.html" target="_self">nucleotide_alphabet</a></td><td class="desc">A concept that indicates whether an alphabet represents nucleotides </td></tr>
<tr id="row_26_0_0_1_0_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna15.html" target="_self">bio::alphabet::dna15</a></td><td class="desc">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_26_0_0_1_1_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" target="_self">bio::alphabet::dna16sam</a></td><td class="desc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=') </td></tr>
<tr id="row_26_0_0_1_2_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna4.html" target="_self">bio::alphabet::dna4</a></td><td class="desc">The four letter DNA alphabet of A,C,G,T </td></tr>
<tr id="row_26_0_0_1_3_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna5.html" target="_self">bio::alphabet::dna5</a></td><td class="desc">The five letter DNA alphabet of A,C,G,T and the unknown character N </td></tr>
<tr id="row_26_0_0_1_4_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna15.html" target="_self">bio::alphabet::rna15</a></td><td class="desc">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_26_0_0_1_5_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna4.html" target="_self">bio::alphabet::rna4</a></td><td class="desc">The four letter RNA alphabet of A,C,G,U </td></tr>
<tr id="row_26_0_0_1_6_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna5.html" target="_self">bio::alphabet::rna5</a></td><td class="desc">The five letter RNA alphabet of A,C,G,U and the unknown character N </td></tr>
<tr id="row_26_0_0_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span id="arr_26_0_0_2_" class="arrow" onclick="toggleFolder('26_0_0_2_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1quality__alphabet.html" target="_self">quality_alphabet</a></td><td class="desc">A concept that indicates whether an alphabet represents quality scores </td></tr>
<tr id="row_26_0_0_2_0_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span id="arr_26_0_0_2_0_" class="arrow" onclick="toggleFolder('26_0_0_2_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1writable__quality__alphabet.html" target="_self">writable_quality_alphabet</a></td><td class="desc">A concept that indicates whether a writable alphabet represents quality scores </td></tr>
<tr id="row_26_0_0_2_0_0_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred42.html" target="_self">bio::alphabet::phred42</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (typical range) </td></tr>
<tr id="row_26_0_0_2_0_1_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred63.html" target="_self">bio::alphabet::phred63</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (full range) </td></tr>
<tr id="row_26_0_0_2_0_2_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred68legacy.html" target="_self">bio::alphabet::phred68legacy</a></td><td class="desc">Quality type for Solexa and deprecated Illumina formats </td></tr>
<tr id="row_26_0_0_2_0_3_" style="display:none;"><td class="entry"><span style="width:96px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1qualified.html" target="_self">bio::alphabet::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc">Joins an arbitrary alphabet with a quality alphabet </td></tr>
<tr id="row_26_0_0_3_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span id="arr_26_0_0_3_" class="arrow" onclick="toggleFolder('26_0_0_3_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1writable__alphabet.html" target="_self">writable_alphabet</a></td><td class="desc">Refines <a class="el" href="interfacebio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a> and adds assignability </td></tr>
<tr id="row_26_0_0_3_0_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10li.html" target="_self">bio::alphabet::aa10li</a></td><td class="desc">The reduced Li amino acid alphabet </td></tr>
<tr id="row_26_0_0_3_1_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html" target="_self">bio::alphabet::aa10murphy</a></td><td class="desc">The reduced Murphy amino acid alphabet </td></tr>
<tr id="row_26_0_0_3_2_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa20.html" target="_self">bio::alphabet::aa20</a></td><td class="desc">The canonical amino acid alphabet </td></tr>
<tr id="row_26_0_0_3_3_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa27.html" target="_self">bio::alphabet::aa27</a></td><td class="desc">The twenty-seven letter amino acid alphabet </td></tr>
<tr id="row_26_0_0_3_4_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__variant.html" target="_self">bio::alphabet::alphabet_variant&lt; alternative_types &gt;</a></td><td class="desc">A combined alphabet that can hold values of either of its alternatives </td></tr>
<tr id="row_26_0_0_3_5_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" target="_self">bio::alphabet::cigar_op</a></td><td class="desc">The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,= </td></tr>
<tr id="row_26_0_0_3_6_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna15.html" target="_self">bio::alphabet::dna15</a></td><td class="desc">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_26_0_0_3_7_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" target="_self">bio::alphabet::dna16sam</a></td><td class="desc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=') </td></tr>
<tr id="row_26_0_0_3_8_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna4.html" target="_self">bio::alphabet::dna4</a></td><td class="desc">The four letter DNA alphabet of A,C,G,T </td></tr>
<tr id="row_26_0_0_3_9_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna5.html" target="_self">bio::alphabet::dna5</a></td><td class="desc">The five letter DNA alphabet of A,C,G,T and the unknown character N </td></tr>
<tr id="row_26_0_0_3_10_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1gap.html" target="_self">bio::alphabet::gap</a></td><td class="desc">The alphabet of a gap character '-' </td></tr>
<tr id="row_26_0_0_3_11_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1masked.html" target="_self">bio::alphabet::masked&lt; sequence_alphabet_t &gt;</a></td><td class="desc">Implementation of a masked composite, which extends a given alphabet with a mask </td></tr>
<tr id="row_26_0_0_3_12_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna15.html" target="_self">bio::alphabet::rna15</a></td><td class="desc">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_26_0_0_3_13_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna4.html" target="_self">bio::alphabet::rna4</a></td><td class="desc">The four letter RNA alphabet of A,C,G,U </td></tr>
<tr id="row_26_0_0_3_14_" style="display:none;"><td class="entry"><span style="width:80px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna5.html" target="_self">bio::alphabet::rna5</a></td><td class="desc">The five letter RNA alphabet of A,C,G,U and the unknown character N </td></tr>
<tr id="row_26_0_0_4_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1writable__quality__alphabet.html" target="_self">writable_quality_alphabet</a></td><td class="desc">A concept that indicates whether a writable alphabet represents quality scores </td></tr>
<tr id="row_26_0_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1semialphabet__any.html" target="_self">bio::alphabet::semialphabet_any&lt; size_ &gt;</a></td><td class="desc">A semi-alphabet that type erases all other semi-alphabets of the same size </td></tr>
<tr id="row_26_0_2_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span id="arr_26_0_2_" class="arrow" onclick="toggleFolder('26_0_2_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1writable__semialphabet.html" target="_self">writable_semialphabet</a></td><td class="desc">A refinement of <a class="el" href="interfacebio_1_1alphabet_1_1semialphabet.html" title="The basis for bio::alphabet::alphabet, but requires only rank interface (not char).">bio::alphabet::semialphabet</a> that adds assignability </td></tr>
<tr id="row_26_0_2_0_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; masked&lt; sequence_alphabet_t &gt;, sequence_alphabet_t, mask &gt;</a></td><td class="desc"></td></tr>
<tr id="row_26_0_2_1_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;, sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc"></td></tr>
<tr id="row_26_0_2_2_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; cigar, uint32_t, cigar_op &gt;</a></td><td class="desc"></td></tr>
<tr id="row_26_0_2_3_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__tuple__base.html" target="_self">bio::alphabet::alphabet_tuple_base&lt; derived_type, component_types &gt;</a></td><td class="desc">The CRTP base for a combined alphabet that contains multiple values of different alphabets at the same time </td></tr>
<tr id="row_26_0_2_4_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar.html" target="_self">bio::alphabet::cigar</a></td><td class="desc">The cigar semialphabet pairs a counter with a <a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" title="The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,=.">bio::alphabet::cigar_op</a> letter </td></tr>
<tr id="row_26_0_2_5_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1mask.html" target="_self">bio::alphabet::mask</a></td><td class="desc">Implementation of a masked alphabet to be used for tuple composites </td></tr>
<tr id="row_26_0_2_6_" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1writable__alphabet.html" target="_self">writable_alphabet</a></td><td class="desc">Refines <a class="el" href="interfacebio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a> and adds assignability </td></tr>
<tr id="row_27_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_27_" class="arrow" onclick="toggleFolder('27_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::copyable</b></td><td class="desc"></td></tr>
<tr id="row_27_0_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_27_0_" class="arrow" onclick="toggleFolder('27_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1trivially__copyable.html" target="_self">trivially_copyable</a></td><td class="desc">A type that satisfies std::is_trivially_copyable_v&lt;t&gt; </td></tr>
<tr id="row_27_0_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10li.html" target="_self">bio::alphabet::aa10li</a></td><td class="desc">The reduced Li amino acid alphabet </td></tr>
<tr id="row_27_0_1_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html" target="_self">bio::alphabet::aa10murphy</a></td><td class="desc">The reduced Murphy amino acid alphabet </td></tr>
<tr id="row_27_0_2_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa20.html" target="_self">bio::alphabet::aa20</a></td><td class="desc">The canonical amino acid alphabet </td></tr>
<tr id="row_27_0_3_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa27.html" target="_self">bio::alphabet::aa27</a></td><td class="desc">The twenty-seven letter amino acid alphabet </td></tr>
<tr id="row_27_0_4_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__variant.html" target="_self">bio::alphabet::alphabet_variant&lt; alternative_types &gt;</a></td><td class="desc">A combined alphabet that can hold values of either of its alternatives </td></tr>
<tr id="row_27_0_5_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar.html" target="_self">bio::alphabet::cigar</a></td><td class="desc">The cigar semialphabet pairs a counter with a <a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" title="The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,=.">bio::alphabet::cigar_op</a> letter </td></tr>
<tr id="row_27_0_6_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" target="_self">bio::alphabet::cigar_op</a></td><td class="desc">The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,= </td></tr>
<tr id="row_27_0_7_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna15.html" target="_self">bio::alphabet::dna15</a></td><td class="desc">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_27_0_8_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" target="_self">bio::alphabet::dna16sam</a></td><td class="desc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=') </td></tr>
<tr id="row_27_0_9_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna4.html" target="_self">bio::alphabet::dna4</a></td><td class="desc">The four letter DNA alphabet of A,C,G,T </td></tr>
<tr id="row_27_0_10_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna5.html" target="_self">bio::alphabet::dna5</a></td><td class="desc">The five letter DNA alphabet of A,C,G,T and the unknown character N </td></tr>
<tr id="row_27_0_11_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1gap.html" target="_self">bio::alphabet::gap</a></td><td class="desc">The alphabet of a gap character '-' </td></tr>
<tr id="row_27_0_12_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1mask.html" target="_self">bio::alphabet::mask</a></td><td class="desc">Implementation of a masked alphabet to be used for tuple composites </td></tr>
<tr id="row_27_0_13_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1masked.html" target="_self">bio::alphabet::masked&lt; sequence_alphabet_t &gt;</a></td><td class="desc">Implementation of a masked composite, which extends a given alphabet with a mask </td></tr>
<tr id="row_27_0_14_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred42.html" target="_self">bio::alphabet::phred42</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (typical range) </td></tr>
<tr id="row_27_0_15_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred63.html" target="_self">bio::alphabet::phred63</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (full range) </td></tr>
<tr id="row_27_0_16_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred68legacy.html" target="_self">bio::alphabet::phred68legacy</a></td><td class="desc">Quality type for Solexa and deprecated Illumina formats </td></tr>
<tr id="row_27_0_17_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1qualified.html" target="_self">bio::alphabet::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc">Joins an arbitrary alphabet with a quality alphabet </td></tr>
<tr id="row_27_0_18_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna15.html" target="_self">bio::alphabet::rna15</a></td><td class="desc">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_27_0_19_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna4.html" target="_self">bio::alphabet::rna4</a></td><td class="desc">The four letter RNA alphabet of A,C,G,U </td></tr>
<tr id="row_27_0_20_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna5.html" target="_self">bio::alphabet::rna5</a></td><td class="desc">The five letter RNA alphabet of A,C,G,U and the unknown character N </td></tr>
<tr id="row_27_0_21_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1semialphabet__any.html" target="_self">bio::alphabet::semialphabet_any&lt; size_ &gt;</a></td><td class="desc">A semi-alphabet that type erases all other semi-alphabets of the same size </td></tr>
<tr id="row_27_0_22_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1trivial.html" target="_self">trivial</a></td><td class="desc">A type that satisfies <a class="el" href="interfacebio_1_1meta_1_1trivially__copyable.html" title="A type that satisfies std::is_trivially_copyable_v&lt;t&gt;.">bio::meta::trivially_copyable</a> and <a class="el" href="interfacebio_1_1meta_1_1trivially__destructible.html" title="A type that satisfies std::is_trivially_destructible_v&lt;t&gt;.">bio::meta::trivially_destructible</a> </td></tr>
<tr id="row_28_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_28_" class="arrow" onclick="toggleFolder('28_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::destructible</b></td><td class="desc"></td></tr>
<tr id="row_28_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_28_0_" class="arrow" onclick="toggleFolder('28_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1trivially__destructible.html" target="_self">trivially_destructible</a></td><td class="desc">A type that satisfies std::is_trivially_destructible_v&lt;t&gt; </td></tr>
<tr id="row_28_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1trivial.html" target="_self">trivial</a></td><td class="desc">A type that satisfies <a class="el" href="interfacebio_1_1meta_1_1trivially__copyable.html" title="A type that satisfies std::is_trivially_copyable_v&lt;t&gt;.">bio::meta::trivially_copyable</a> and <a class="el" href="interfacebio_1_1meta_1_1trivially__destructible.html" title="A type that satisfies std::is_trivially_destructible_v&lt;t&gt;.">bio::meta::trivially_destructible</a> </td></tr>
<tr id="row_29_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1explicitly__convertible__to.html" target="_self">explicitly_convertible_to</a></td><td class="desc">Resolves to <code>std::ranges::explicitly_convertible_to&lt;type1, type2&gt;()</code> </td></tr>
<tr id="row_30_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_30_" class="arrow" onclick="toggleFolder('30_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/integral_constant.html">std::false_type</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_30_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__constexpr__default__constructible.html" target="_self">bio::meta::is_constexpr_default_constructible&lt; t &gt;</a></td><td class="desc">Whether a type <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/is_default_constructible.html">std::is_default_constructible</a> in <code>constexpr</code>-context </td></tr>
<tr id="row_30_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__type__specialisation__of.html" target="_self">bio::meta::is_type_specialisation_of&lt; source_t, target_template &gt;</a></td><td class="desc">Determines whether a source_type is a specialisation of another template </td></tr>
<tr id="row_31_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_31_" class="arrow" onclick="toggleFolder('31_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::forward_iterator</b></td><td class="desc"></td></tr>
<tr id="row_31_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1ranges_1_1pseudo__random__access__iterator.html" target="_self">pseudo_random_access_iterator</a></td><td class="desc">This concept checks if an iterator type models pseudo random access </td></tr>
<tr id="row_32_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_32_" class="arrow" onclick="toggleFolder('32_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::ranges::forward_range</b></td><td class="desc"></td></tr>
<tr id="row_32_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><b>container</b></td><td class="desc">The (most general) container concept as defined by the standard library </td></tr>
<tr id="row_32_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1ranges_1_1pseudo__random__access__range.html" target="_self">pseudo_random_access_range</a></td><td class="desc">This concept checks if a type models a pseudo random access range </td></tr>
<tr id="row_33_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1function__traits_3_01std_1_1function_3_01return__t_07args__t_8_8_8_08_4_01_4.html" target="_self">bio::meta::function_traits&lt; std::function&lt; return_t(args_t...)&gt; &gt;</a></td><td class="desc">A traits class to provide a uniform interface to the properties of a function type </td></tr>
<tr id="row_34_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1hash_3_01alphabet__t_01_4.html" target="_self">std::hash&lt; alphabet_t &gt;</a></td><td class="desc">Struct for hashing a character </td></tr>
<tr id="row_35_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1hash_3_01bio_1_1ranges_1_1dynamic__bitset_3_01cap_01_4_01_4.html" target="_self">std::hash&lt; bio::ranges::dynamic_bitset&lt; cap &gt; &gt;</a></td><td class="desc">Struct for hashing a <code><a class="el" href="classbio_1_1ranges_1_1dynamic__bitset.html" title="A constexpr bitset implementation with dynamic size at compile time.">bio::ranges::dynamic_bitset</a></code> </td></tr>
<tr id="row_36_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1hash_3_01urng__t_01_4.html" target="_self">std::hash&lt; urng_t &gt;</a></td><td class="desc">Struct for hashing a range of characters </td></tr>
<tr id="row_37_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1implicitly__convertible__to.html" target="_self">implicitly_convertible_to</a></td><td class="desc">Resolves to <code>std::ranges::implicitly_convertible_to&lt;type1, type2&gt;()</code> </td></tr>
<tr id="row_38_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_38_" class="arrow" onclick="toggleFolder('38_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::input_range</b></td><td class="desc"></td></tr>
<tr id="row_38_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_38_0_" class="arrow" onclick="toggleFolder('38_0_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1ranges_1_1const__iterable__range.html" target="_self">const_iterable_range</a></td><td class="desc">Specifies requirements of an input range type for which the <code>const</code> version of that type satisfies the same strength input range concept as the non-const version </td></tr>
<tr id="row_38_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><b>container</b></td><td class="desc">The (most general) container concept as defined by the standard library </td></tr>
<tr id="row_39_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_39_" class="arrow" onclick="toggleFolder('39_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::ranges::input_range</b></td><td class="desc"></td></tr>
<tr id="row_39_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1ranges_1_1sequence.html" target="_self">sequence</a></td><td class="desc">The generic concept for a sequence </td></tr>
<tr id="row_40_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_40_" class="arrow" onclick="toggleFolder('40_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::integral</b></td><td class="desc"></td></tr>
<tr id="row_40_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1builtin__character.html" target="_self">builtin_character</a></td><td class="desc">This concept encompasses exactly the types <code>char</code>, <code>signed char</code>, <code>unsigned char</code>, <code>wchar_t</code>, <code>char16_t</code> and <code>char32_t</code> </td></tr>
<tr id="row_40_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1nonint__character.html" target="_self">nonint_character</a></td><td class="desc">This concept encompasses exactly the types <code>char</code>, <code>wchar_t</code>, <code>char16_t</code> and <code>char32_t</code> </td></tr>
<tr id="row_41_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_41_" class="arrow" onclick="toggleFolder('41_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/integral_constant.html">std::integral_constant&lt; bool, BIOCPP_IS_CONSTEXPR(t{})&gt;</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_41_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__constexpr__default__constructible_3_01t_01_4.html" target="_self">bio::meta::is_constexpr_default_constructible&lt; t &gt;</a></td><td class="desc">Whether a type <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/is_default_constructible.html">std::is_default_constructible</a> in <code>constexpr</code>-context (<a class="el" href="interfacebio_1_1meta_1_1unary__type__trait.html" title="Concept for a unary traits type.">unary_type_trait</a> specialisation) </td></tr>
<tr id="row_42_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_42_" class="arrow" onclick="toggleFolder('42_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/integral_constant.html">std::integral_constant&lt; size_t, bio::meta::list_traits::size&lt; tuple_t::biocpp_required_types &gt; &gt;</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_42_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1tuple__size_3_01tuple__t_01_4.html" target="_self">std::tuple_size&lt; tuple_t &gt;</a></td><td class="desc">Provides access to the number of elements in a tuple as a compile-time constant expression </td></tr>
<tr id="row_43_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_43_" class="arrow" onclick="toggleFolder('43_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/integral_constant.html">std::integral_constant&lt; std::size_t, sizeof...(types)&gt;</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_43_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1tuple__size_3_01t_3_01types_8_8_8_01_4_01_4.html" target="_self">std::tuple_size&lt; t&lt; types... &gt; &gt;</a></td><td class="desc">Provides access to the number of elements in a tuple as a compile-time constant expression </td></tr>
<tr id="row_44_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_44_" class="arrow" onclick="toggleFolder('44_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/is_same.html">std::is_same&lt; source_t, transfer_template_args_onto_t&lt; source_t, target_template &gt; &gt;</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_44_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__type__specialisation__of_3_01source__t_00_01target__template_01_4.html" target="_self">bio::meta::is_type_specialisation_of&lt; source_t, target_template &gt;</a></td><td class="desc">This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts </td></tr>
<tr id="row_45_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_45_" class="arrow" onclick="toggleFolder('45_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/is_same.html">std::is_same&lt; source_t, transfer_template_vargs_onto_t&lt; source_t, target_template &gt; &gt;</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_45_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__value__specialisation__of_3_01source__t_00_01target__template_01_4.html" target="_self">bio::meta::is_value_specialisation_of&lt; source_t, target_template &gt;</a></td><td class="desc">Determines whether a source_type is a specialisation of another template </td></tr>
<tr id="row_46_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1pod__tuple.html" target="_self">bio::meta::pod_tuple&lt; types &gt;</a></td><td class="desc">Behaves like <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/tuple.html">std::tuple</a> but is an aggregate <a href="https://en.cppreference.com/w/cpp/concept/PODType">PODType</a> </td></tr>
<tr id="row_47_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1pod__tuple_3_01type0_01_4.html" target="_self">bio::meta::pod_tuple&lt; type0 &gt;</a></td><td class="desc">Recursion anchor for <a class="el" href="structbio_1_1meta_1_1pod__tuple.html" title="Behaves like std::tuple but is an aggregate PODType.">pod_tuple</a> </td></tr>
<tr id="row_48_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_48_" class="arrow" onclick="toggleFolder('48_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::ranges::random_access_range</b></td><td class="desc"></td></tr>
<tr id="row_48_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><b>random_access_container</b></td><td class="desc">A more refined container concept than bio::ranges::detail::sequence_container </td></tr>
<tr id="row_49_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1ranges_1_1range__compatible.html" target="_self">range_compatible</a></td><td class="desc">Two types are "compatible" if their <a class="el" href="group__range.html#ga6328137c2baad5fa78067433d4a21546" title="Returns the number of times you can call std::ranges::value_type_t recursively on t (type trait).">bio::ranges::range_dimension_v</a> and their <a class="el" href="group__range.html#ga7f6b48fe07fecaad1c6c747d2cc7383d" title="Shortcut for bio::ranges::range_innermost_value (transformation_trait shortcut).">bio::ranges::range_innermost_value_t</a> are the same </td></tr>
<tr id="row_50_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1ranges_1_1aligned__allocator_1_1rebind.html" target="_self">bio::ranges::aligned_allocator&lt; value_t, alignment_v &gt;::rebind&lt; new_value_type &gt;</a></td><td class="desc">The <a class="el" href="classbio_1_1ranges_1_1aligned__allocator.html" title="Allocates uninitialized storage whose memory-alignment is specified by alignment.">aligned_allocator</a> member template class <a class="el" href="structbio_1_1ranges_1_1aligned__allocator_1_1rebind.html" title="The aligned_allocator member template class aligned_allocator::rebind provides a way to obtain an all...">aligned_allocator::rebind</a> provides a way to obtain an allocator for a different type </td></tr>
<tr id="row_51_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_51_" class="arrow" onclick="toggleFolder('51_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::regular</b></td><td class="desc"></td></tr>
<tr id="row_51_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10li.html" target="_self">bio::alphabet::aa10li</a></td><td class="desc">The reduced Li amino acid alphabet </td></tr>
<tr id="row_51_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html" target="_self">bio::alphabet::aa10murphy</a></td><td class="desc">The reduced Murphy amino acid alphabet </td></tr>
<tr id="row_51_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa20.html" target="_self">bio::alphabet::aa20</a></td><td class="desc">The canonical amino acid alphabet </td></tr>
<tr id="row_51_3_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa27.html" target="_self">bio::alphabet::aa27</a></td><td class="desc">The twenty-seven letter amino acid alphabet </td></tr>
<tr id="row_51_4_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar.html" target="_self">bio::alphabet::cigar</a></td><td class="desc">The cigar semialphabet pairs a counter with a <a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" title="The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,=.">bio::alphabet::cigar_op</a> letter </td></tr>
<tr id="row_51_5_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" target="_self">bio::alphabet::cigar_op</a></td><td class="desc">The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,= </td></tr>
<tr id="row_51_6_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna15.html" target="_self">bio::alphabet::dna15</a></td><td class="desc">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_51_7_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna4.html" target="_self">bio::alphabet::dna4</a></td><td class="desc">The four letter DNA alphabet of A,C,G,T </td></tr>
<tr id="row_51_8_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna5.html" target="_self">bio::alphabet::dna5</a></td><td class="desc">The five letter DNA alphabet of A,C,G,T and the unknown character N </td></tr>
<tr id="row_51_9_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1gap.html" target="_self">bio::alphabet::gap</a></td><td class="desc">The alphabet of a gap character '-' </td></tr>
<tr id="row_51_10_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1mask.html" target="_self">bio::alphabet::mask</a></td><td class="desc">Implementation of a masked alphabet to be used for tuple composites </td></tr>
<tr id="row_51_11_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1masked.html" target="_self">bio::alphabet::masked&lt; sequence_alphabet_t &gt;</a></td><td class="desc">Implementation of a masked composite, which extends a given alphabet with a mask </td></tr>
<tr id="row_51_12_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred42.html" target="_self">bio::alphabet::phred42</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (typical range) </td></tr>
<tr id="row_51_13_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred63.html" target="_self">bio::alphabet::phred63</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (full range) </td></tr>
<tr id="row_51_14_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred68legacy.html" target="_self">bio::alphabet::phred68legacy</a></td><td class="desc">Quality type for Solexa and deprecated Illumina formats </td></tr>
<tr id="row_51_15_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1qualified.html" target="_self">bio::alphabet::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc">Joins an arbitrary alphabet with a quality alphabet </td></tr>
<tr id="row_51_16_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna15.html" target="_self">bio::alphabet::rna15</a></td><td class="desc">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_51_17_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna4.html" target="_self">bio::alphabet::rna4</a></td><td class="desc">The four letter RNA alphabet of A,C,G,U </td></tr>
<tr id="row_51_18_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna5.html" target="_self">bio::alphabet::rna5</a></td><td class="desc">The five letter RNA alphabet of A,C,G,U and the unknown character N </td></tr>
<tr id="row_52_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_52_" class="arrow" onclick="toggleFolder('52_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/error/runtime_error.html">std::runtime_error</a>&#160;[external]</td><td class="desc"></td></tr>
<tr id="row_52_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1invalid__char__assignment.html" target="_self">bio::alphabet::invalid_char_assignment</a></td><td class="desc">An exception typically thrown by bio::alphabet::assign_char_strict </td></tr>
<tr id="row_53_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1size.html" target="_self">bio::alphabet::cpo::size</a></td><td class="desc">CPO tag definition for <a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d" title="A type trait that holds the size of a (semi-)alphabet.">bio::alphabet::size</a> </td></tr>
<tr id="row_54_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_54_" class="arrow" onclick="toggleFolder('54_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::ranges::sized_range</b></td><td class="desc"></td></tr>
<tr id="row_54_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><b>container</b></td><td class="desc">The (most general) container concept as defined by the standard library </td></tr>
<tr id="row_55_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_55_" class="arrow" onclick="toggleFolder('55_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1standard__layout.html" target="_self">standard_layout</a></td><td class="desc">Resolves to std::is_standard_layout_v&lt;t&gt; </td></tr>
<tr id="row_55_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10li.html" target="_self">bio::alphabet::aa10li</a></td><td class="desc">The reduced Li amino acid alphabet </td></tr>
<tr id="row_55_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html" target="_self">bio::alphabet::aa10murphy</a></td><td class="desc">The reduced Murphy amino acid alphabet </td></tr>
<tr id="row_55_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa20.html" target="_self">bio::alphabet::aa20</a></td><td class="desc">The canonical amino acid alphabet </td></tr>
<tr id="row_55_3_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1aa27.html" target="_self">bio::alphabet::aa27</a></td><td class="desc">The twenty-seven letter amino acid alphabet </td></tr>
<tr id="row_55_4_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1alphabet__variant.html" target="_self">bio::alphabet::alphabet_variant&lt; alternative_types &gt;</a></td><td class="desc">A combined alphabet that can hold values of either of its alternatives </td></tr>
<tr id="row_55_5_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar.html" target="_self">bio::alphabet::cigar</a></td><td class="desc">The cigar semialphabet pairs a counter with a <a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" title="The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,=.">bio::alphabet::cigar_op</a> letter </td></tr>
<tr id="row_55_6_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" target="_self">bio::alphabet::cigar_op</a></td><td class="desc">The (extended) cigar operation alphabet of M,D,I,H,N,P,S,X,= </td></tr>
<tr id="row_55_7_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna15.html" target="_self">bio::alphabet::dna15</a></td><td class="desc">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_55_8_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html" target="_self">bio::alphabet::dna16sam</a></td><td class="desc">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=') </td></tr>
<tr id="row_55_9_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna4.html" target="_self">bio::alphabet::dna4</a></td><td class="desc">The four letter DNA alphabet of A,C,G,T </td></tr>
<tr id="row_55_10_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1dna5.html" target="_self">bio::alphabet::dna5</a></td><td class="desc">The five letter DNA alphabet of A,C,G,T and the unknown character N </td></tr>
<tr id="row_55_11_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1gap.html" target="_self">bio::alphabet::gap</a></td><td class="desc">The alphabet of a gap character '-' </td></tr>
<tr id="row_55_12_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1mask.html" target="_self">bio::alphabet::mask</a></td><td class="desc">Implementation of a masked alphabet to be used for tuple composites </td></tr>
<tr id="row_55_13_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1masked.html" target="_self">bio::alphabet::masked&lt; sequence_alphabet_t &gt;</a></td><td class="desc">Implementation of a masked composite, which extends a given alphabet with a mask </td></tr>
<tr id="row_55_14_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred42.html" target="_self">bio::alphabet::phred42</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (typical range) </td></tr>
<tr id="row_55_15_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred63.html" target="_self">bio::alphabet::phred63</a></td><td class="desc">Quality type for traditional Sanger and modern Illumina Phred scores (full range) </td></tr>
<tr id="row_55_16_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1phred68legacy.html" target="_self">bio::alphabet::phred68legacy</a></td><td class="desc">Quality type for Solexa and deprecated Illumina formats </td></tr>
<tr id="row_55_17_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1qualified.html" target="_self">bio::alphabet::qualified&lt; sequence_alphabet_t, quality_alphabet_t &gt;</a></td><td class="desc">Joins an arbitrary alphabet with a quality alphabet </td></tr>
<tr id="row_55_18_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna15.html" target="_self">bio::alphabet::rna15</a></td><td class="desc">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap </td></tr>
<tr id="row_55_19_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna4.html" target="_self">bio::alphabet::rna4</a></td><td class="desc">The four letter RNA alphabet of A,C,G,U </td></tr>
<tr id="row_55_20_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1rna5.html" target="_self">bio::alphabet::rna5</a></td><td class="desc">The five letter RNA alphabet of A,C,G,U and the unknown character N </td></tr>
<tr id="row_55_21_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="classbio_1_1alphabet_1_1semialphabet__any.html" target="_self">bio::alphabet::semialphabet_any&lt; size_ &gt;</a></td><td class="desc">A semi-alphabet that type erases all other semi-alphabets of the same size </td></tr>
<tr id="row_56_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1template__specialisation__of.html" target="_self">template_specialisation_of</a></td><td class="desc">Provides concept <code><a class="el" href="interfacebio_1_1meta_1_1template__specialisation__of.html" title="Provides concept bio::meta::template_specialisation_of&lt;mytype, [...]&gt; for checking the type specialis...">bio::meta::template_specialisation_of</a>&lt;mytype, [...]&gt;</code> for checking the type specialisation of some type with a given template, for example a specialized <code><a class="el" href="structbio_1_1meta_1_1type__list.html" title="Type that contains multiple types.">type_list</a>&lt;float&gt;</code> with the <code><a class="el" href="structbio_1_1meta_1_1type__list.html" title="Type that contains multiple types.">type_list</a></code> template </td></tr>
<tr id="row_57_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1to__char.html" target="_self">bio::alphabet::cpo::to_char</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#ga064e386f9360cca0fd3702b4957a27a0" title="Return the char representation of an alphabet object.">bio::alphabet::to_char</a> </td></tr>
<tr id="row_58_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1to__phred.html" target="_self">bio::alphabet::cpo::to_phred</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#ga0c2eb25dabf59ccf9992b2661083ea5b" title="Assign a char to an alphabet object.">bio::alphabet::assign_char_to</a> </td></tr>
<tr id="row_59_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1alphabet_1_1cpo_1_1to__rank.html" target="_self">bio::alphabet::cpo::to_rank</a></td><td class="desc">Customisation tag for <a class="el" href="group__alphabet.html#ga2a48a5c1883184e362f09e32ac33241e" title="Return the rank representation of a (semi-)alphabet object.">bio::alphabet::to_rank</a> </td></tr>
<tr id="row_60_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_60_" class="arrow" onclick="toggleFolder('60_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>std::totally_ordered</b></td><td class="desc"></td></tr>
<tr id="row_60_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1alphabet_1_1semialphabet.html" target="_self">semialphabet</a></td><td class="desc">The basis for <a class="el" href="interfacebio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a>, but requires only rank interface (not char) </td></tr>
<tr id="row_60_1_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_60_1_" class="arrow" onclick="toggleFolder('60_1_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1tuple__like.html" target="_self">tuple_like</a></td><td class="desc">Whether a type behaves like a tuple </td></tr>
<tr id="row_60_1_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1pair__like.html" target="_self">pair_like</a></td><td class="desc">Whether a type behaves like a tuple with exactly two elements </td></tr>
<tr id="row_60_1_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1pod__tuple_3_01type0_00_01types_8_8_8_01_4.html" target="_self">bio::meta::pod_tuple&lt; type0, types... &gt;</a></td><td class="desc">Behaves like <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/tuple.html">std::tuple</a> but is an aggregate <a href="https://en.cppreference.com/w/cpp/concept/PODType">PODType</a> </td></tr>
<tr id="row_61_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_61_" class="arrow" onclick="toggleFolder('61_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1transformation__trait.html" target="_self">transformation_trait</a></td><td class="desc">Concept for a transformation trait </td></tr>
<tr id="row_61_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1deferred__type.html" target="_self">bio::meta::deferred_type&lt; t, dependent_ts &gt;</a></td><td class="desc">Return the type identity; further arguments are ignored, but can make this type dependent if they are </td></tr>
<tr id="row_61_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1remove__rvalue__reference.html" target="_self">bio::meta::remove_rvalue_reference&lt; t &gt;</a></td><td class="desc">Return the input type with <code>&amp;&amp;</code> removed, but lvalue references preserved </td></tr>
<tr id="row_61_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1transfer__template__args__onto_3_01source__template_3_01source__arg__types_a0c928784694c351c065f80d4d89b15d.html" target="_self">bio::meta::transfer_template_args_onto&lt; source_template&lt; source_arg_types... &gt;, target_template &gt;</a></td><td class="desc">Extracts a type template's <b>type</b> arguments and specialises another template with them </td></tr>
<tr id="row_61_3_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1transfer__template__vargs__onto_3_01source__template_3_01source__varg__type4e348813352083c94c53b6206e44e292.html" target="_self">bio::meta::transfer_template_vargs_onto&lt; source_template&lt; source_varg_types... &gt;, target_template &gt;</a></td><td class="desc">Extracts a type template's <b>non-type</b> arguments and specialises another template with them </td></tr>
<tr id="row_61_4_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1valid__template__spec__or.html" target="_self">bio::meta::valid_template_spec_or&lt; fallback_t, templ_t, spec_t &gt;</a></td><td class="desc">Exposes <code>templ_t&lt;spec_t...&gt;</code> if that is valid, otherwise <code>fallback_t</code> </td></tr>
<tr id="row_61_5_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1ranges_1_1range__innermost__value.html" target="_self">bio::ranges::range_innermost_value&lt; t &gt;</a></td><td class="desc">Recursively determines the <code>value_type</code> on containers and/or iterators </td></tr>
<tr id="row_61_6_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1tuple__element_3_01i_00_01t_3_01types_8_8_8_01_4_01_4.html" target="_self">std::tuple_element&lt; i, t&lt; types... &gt; &gt;</a></td><td class="desc">Obtains the type of the specified element </td></tr>
<tr id="row_61_7_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1tuple__element_3_01i_00_01tuple__t_01_4.html" target="_self">std::tuple_element&lt; i, tuple_t &gt;</a></td><td class="desc">Obtains the type of the specified element </td></tr>
<tr id="row_62_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1type__list.html" target="_self">bio::meta::type_list&lt; types &gt;</a></td><td class="desc">Type that contains multiple types </td></tr>
<tr id="row_63_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_63_" class="arrow" onclick="toggleFolder('63_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1unary__type__trait.html" target="_self">unary_type_trait</a></td><td class="desc">Concept for a unary traits type </td></tr>
<tr id="row_63_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__constexpr__default__constructible.html" target="_self">bio::meta::is_constexpr_default_constructible&lt; t &gt;</a></td><td class="desc">Whether a type <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/types/is_default_constructible.html">std::is_default_constructible</a> in <code>constexpr</code>-context </td></tr>
<tr id="row_63_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__type__specialisation__of.html" target="_self">bio::meta::is_type_specialisation_of&lt; source_t, target_template &gt;</a></td><td class="desc">Determines whether a source_type is a specialisation of another template </td></tr>
<tr id="row_63_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1is__value__specialisation__of_3_01source__t_00_01target__template_01_4.html" target="_self">bio::meta::is_value_specialisation_of&lt; source_t, target_template &gt;</a></td><td class="desc">Determines whether a source_type is a specialisation of another template </td></tr>
<tr id="row_63_3_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1tuple__size_3_01t_3_01types_8_8_8_01_4_01_4.html" target="_self">std::tuple_size&lt; t&lt; types... &gt; &gt;</a></td><td class="desc">Provides access to the number of elements in a tuple as a compile-time constant expression </td></tr>
<tr id="row_63_4_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structstd_1_1tuple__size_3_01tuple__t_01_4.html" target="_self">std::tuple_size&lt; tuple_t &gt;</a></td><td class="desc">Provides access to the number of elements in a tuple as a compile-time constant expression </td></tr>
<tr id="row_64_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1valid__template__spec__or_3_01fallback__t_00_01templ__t_00_01spec__t_8_8_8_01_4.html" target="_self">bio::meta::valid_template_spec_or&lt; fallback_t, templ_t, spec_t... &gt;</a></td><td class="desc">This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts </td></tr>
<tr id="row_65_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1weakly__assignable__from.html" target="_self">weakly_assignable_from</a></td><td class="desc">Resolves to std::is_assignable_v&lt;t&gt; </td></tr>
<tr id="row_66_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1weakly__equality__comparable__with.html" target="_self">weakly_equality_comparable_with</a></td><td class="desc">Requires the two operands to be comparable with <code>==</code> and <code>!=</code> in both directions </td></tr>
<tr id="row_67_"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="interfacebio_1_1meta_1_1weakly__ordered__with.html" target="_self">weakly_ordered_with</a></td><td class="desc">Requires the two operands to be comparable with <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code> and <code>&gt;=</code> in both directions </td></tr>
<tr id="row_68_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_68_" class="arrow" onclick="toggleFolder('68_')">&#9658;</span><span class="icona"><span class="icon">C</span></span><b>bio::meta::invocable_ts</b></td><td class="desc"></td></tr>
<tr id="row_68_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span class="icona"><span class="icon">C</span></span><a class="el" href="structbio_1_1meta_1_1overloaded.html" target="_self">bio::meta::overloaded&lt; invocable_ts &gt;</a></td><td class="desc">A type that can conveniently inherit multiple invocables and acts as a union over them </td></tr>
</table>
</div><!-- directory -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="footer">Generated on Fri Jul 22 2022 15:04:38 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
