<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=11"/>
<meta name="generator" content="Doxygen 1.9.6"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: bio::alphabet Namespace Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr id="projectrow">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign">
   <div id="projectname">BioC++<span id="projectnumber">&#160;core-0.7.0</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
    <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <span id="MSearchSelect"                onmouseover="return searchBox.OnSearchSelectShow()"                onmouseout="return searchBox.OnSearchSelectHide()">&#160;</span>
          <input type="text" id="MSearchField" value="" placeholder="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.6 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
var searchBox = new SearchBox("searchBox", "search/",'.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:d3d9a9a6595521f9666a5e94cc830dab83b65699&amp;dn=expat.txt MIT */
$(document).ready(function(){initNavTree('namespacebio_1_1alphabet.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<div id="MSearchResults">
<div class="SRPage">
<div id="SRIndex">
<div id="SRResults"></div>
<div class="SRStatus" id="Loading">Loading...</div>
<div class="SRStatus" id="Searching">Searching...</div>
<div class="SRStatus" id="NoMatches">No Matches</div>
</div>
</div>
</div>
</div>

<div class="header">
  <div class="summary">
<a href="#namespaces">Namespaces</a> &#124;
<a href="#nested-classes">Classes</a> &#124;
<a href="#concepts">Concepts</a> &#124;
<a href="#typedef-members">Typedefs</a> &#124;
<a href="#enum-members">Enumerations</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle"><div class="title">bio::alphabet Namespace Reference<div class="ingroups"><a class="el" href="group__alphabet.html">Alphabet</a></div></div></div>
</div><!--header-->
<div class="contents">

<p>The alphabet module's namespace.  
<a href="namespacebio_1_1alphabet.html#details">More...</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="namespaces" name="namespaces"></a>
Namespaces</h2></td></tr>
<tr class="memitem:namespacebio_1_1alphabet_1_1custom"><td class="memItemLeft" align="right" valign="top">namespace &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebio_1_1alphabet_1_1custom.html">custom</a></td></tr>
<tr class="memdesc:namespacebio_1_1alphabet_1_1custom"><td class="mdescLeft">&#160;</td><td class="mdescRight">A namespace for third party and standard library specialisations of BioC++ customisation points. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:namespacebio_1_1alphabet_1_1literals"><td class="memItemLeft" align="right" valign="top">namespace &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacebio_1_1alphabet_1_1literals.html">literals</a></td></tr>
<tr class="memdesc:namespacebio_1_1alphabet_1_1literals"><td class="mdescLeft">&#160;</td><td class="mdescRight">An inline namespace for alphabet literals. It exists to safely allow <code>using namespace</code>. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="nested-classes" name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1aa10li.html">aa10li</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The reduced Li amino acid alphabet..  <a href="classbio_1_1alphabet_1_1aa10li.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1aa10murphy.html">aa10murphy</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The reduced Murphy amino acid alphabet..  <a href="classbio_1_1alphabet_1_1aa10murphy.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1aa20.html">aa20</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The canonical amino acid alphabet..  <a href="classbio_1_1alphabet_1_1aa20.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1aa27.html">aa27</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The twenty-seven letter amino acid alphabet..  <a href="classbio_1_1alphabet_1_1aa27.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1aminoacid__base.html">aminoacid_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::base</a> and is used by the amino acids.  <a href="classbio_1_1alphabet_1_1aminoacid__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structbio_1_1alphabet_1_1aminoacid__empty__base.html">aminoacid_empty_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">This is an empty base class that can be inherited by types that shall model <a class="el" href="conceptbio_1_1alphabet_1_1aminoacid.html" title="A concept that indicates whether an alphabet represents amino acids.">bio::alphabet::aminoacid</a>.  <a href="structbio_1_1alphabet_1_1aminoacid__empty__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1base.html">base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that makes defining a custom alphabet easier.  <a href="classbio_1_1alphabet_1_1base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1base_3_01derived__type_00_011ul_00_01char__t_01_4.html">base&lt; derived_type, 1ul, char_t &gt;</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Specialisation of <a class="el" href="classbio_1_1alphabet_1_1base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::base</a> for alphabets of size 1.  <a href="classbio_1_1alphabet_1_1base_3_01derived__type_00_011ul_00_01char__t_01_4.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1cigar.html">cigar</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The cigar semialphabet pairs a counter with a <a class="el" href="classbio_1_1alphabet_1_1cigar__op.html" title="The cigar operation alphabet..">bio::alphabet::cigar_op</a> letter.  <a href="classbio_1_1alphabet_1_1cigar.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1cigar__op.html">cigar_op</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The cigar operation alphabet..  <a href="classbio_1_1alphabet_1_1cigar__op.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna15.html">dna15</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..  <a href="classbio_1_1alphabet_1_1dna15.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna16sam.html">dna16sam</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=')..  <a href="classbio_1_1alphabet_1_1dna16sam.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna4.html">dna4</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The four letter DNA alphabet of A,C,G,T..  <a href="classbio_1_1alphabet_1_1dna4.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1dna5.html">dna5</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The five letter DNA alphabet of A,C,G,T and the unknown character N..  <a href="classbio_1_1alphabet_1_1dna5.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1gap.html">gap</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The alphabet of a gap character '-'.  <a href="classbio_1_1alphabet_1_1gap.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structbio_1_1alphabet_1_1invalid__char__assignment.html">invalid_char_assignment</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">An exception typically thrown by bio::alphabet::assign_char_strict.  <a href="structbio_1_1alphabet_1_1invalid__char__assignment.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1mask.html">mask</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Implementation of a masked semialphabet to be used for tuple composites..  <a href="classbio_1_1alphabet_1_1mask.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1masked.html">masked</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A template for composite alphabets that differentiate between upper and lower case characters..  <a href="classbio_1_1alphabet_1_1masked.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1nucleotide__base.html">nucleotide_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::base</a> and is used by the nucleotides.  <a href="classbio_1_1alphabet_1_1nucleotide__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quality type for traditional Sanger and modern Illumina Phred scores (typical range)..  <a href="classbio_1_1alphabet_1_1phred42.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1phred63.html">phred63</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quality type for traditional Sanger and modern Illumina Phred scores (full range)..  <a href="classbio_1_1alphabet_1_1phred63.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1phred68legacy.html">phred68legacy</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quality type for Solexa and deprecated Illumina formats..  <a href="classbio_1_1alphabet_1_1phred68legacy.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1proxy__base.html">proxy_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that eases the definition of proxy types returned in place of regular alphabets.  <a href="classbio_1_1alphabet_1_1proxy__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Joins an arbitrary alphabet with a quality alphabet.  <a href="classbio_1_1alphabet_1_1qualified.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1quality__base.html">quality_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A CRTP-base that refines <a class="el" href="classbio_1_1alphabet_1_1base.html" title="A CRTP-base that makes defining a custom alphabet easier.">bio::alphabet::base</a> and is used by the quality alphabets.  <a href="classbio_1_1alphabet_1_1quality__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1rna15.html">rna15</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap..  <a href="classbio_1_1alphabet_1_1rna15.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1rna4.html">rna4</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The four letter RNA alphabet of A,C,G,U..  <a href="classbio_1_1alphabet_1_1rna4.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1rna5.html">rna5</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The five letter RNA alphabet of A,C,G,U and the unknown character N..  <a href="classbio_1_1alphabet_1_1rna5.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1semialphabet__any.html">semialphabet_any</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A semi-alphabet that type erases all other semi-alphabets of the same size.  <a href="classbio_1_1alphabet_1_1semialphabet__any.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1tuple__base.html">tuple_base</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The CRTP base for a combined alphabet that contains multiple values of different alphabets at the same time..  <a href="classbio_1_1alphabet_1_1tuple__base.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classbio_1_1alphabet_1_1variant.html">variant</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A combined alphabet that can hold values of either of its alternatives..  <a href="classbio_1_1alphabet_1_1variant.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="concepts" name="concepts"></a>
Concepts</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1aminoacid.html">aminoacid</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether an alphabet represents amino acids. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1semialphabet.html">semialphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The basis for <a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a>, but requires only rank interface (not char). <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1writable__semialphabet.html">writable_semialphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A refinement of <a class="el" href="conceptbio_1_1alphabet_1_1semialphabet.html" title="The basis for bio::alphabet::alphabet, but requires only rank interface (not char).">bio::alphabet::semialphabet</a> that adds assignability. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html">alphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">The generic alphabet concept that covers most data types used in ranges. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1writable__alphabet.html">writable_alphabet</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">Refines <a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a> and adds assignability. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1nucleotide.html">nucleotide</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether an alphabet represents nucleotides. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1quality.html">quality</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether an alphabet represents quality scores. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">concept &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="conceptbio_1_1alphabet_1_1writable__quality.html">writable_quality</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">A concept that indicates whether a writable alphabet represents quality scores. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="typedef-members" name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:gaa16ddb3f7d394f68af7f530aeb24e5cf"><td class="memTemplParams" colspan="2">
template&lt;typename alphabet_type &gt; </td></tr>
<tr class="memitem:gaa16ddb3f7d394f68af7f530aeb24e5cf"><td class="memTemplItemLeft" align="right" valign="top">using&#160;</td><td class="memTemplItemRight" valign="bottom"><b>char_t</b> = decltype(<a class="el" href="group__alphabet.html#ga064e386f9360cca0fd3702b4957a27a0">bio::alphabet::to_char</a>(<a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/declval.html">std::declval</a>&lt; alphabet_type const  &gt;()))</td></tr>
<tr class="memdesc:gaa16ddb3f7d394f68af7f530aeb24e5cf"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <code>char_type</code> of the alphabet; defined as the return type of <a class="el" href="group__alphabet.html#ga064e386f9360cca0fd3702b4957a27a0" title="Return the char representation of an alphabet object.">bio::alphabet::to_char</a>. <br /></td></tr>
<tr class="separator:gaa16ddb3f7d394f68af7f530aeb24e5cf"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5e656ee1491695573c945efa713a95d"><td class="memItemLeft" align="right" valign="top"><a id="ad5e656ee1491695573c945efa713a95d" name="ad5e656ee1491695573c945efa713a95d"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>dna15q</b> = <a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna15.html">dna15</a>, <a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a> &gt;</td></tr>
<tr class="memdesc:ad5e656ee1491695573c945efa713a95d"><td class="mdescLeft">&#160;</td><td class="mdescRight">An alphabet that stores a <a class="el" href="classbio_1_1alphabet_1_1dna15.html" title="The 15 letter DNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::dna15</a> letter and an <a class="el" href="classbio_1_1alphabet_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::alphabet::qualified</a> letter at each position. <br /></td></tr>
<tr class="separator:ad5e656ee1491695573c945efa713a95d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aee3aa5dc0944a8c475a5be966eb4b389"><td class="memItemLeft" align="right" valign="top"><a id="aee3aa5dc0944a8c475a5be966eb4b389" name="aee3aa5dc0944a8c475a5be966eb4b389"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>dna4q</b> = <a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna4.html">dna4</a>, <a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a> &gt;</td></tr>
<tr class="memdesc:aee3aa5dc0944a8c475a5be966eb4b389"><td class="mdescLeft">&#160;</td><td class="mdescRight">An alphabet that stores a <a class="el" href="classbio_1_1alphabet_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T..">bio::alphabet::dna4</a> letter and an <a class="el" href="classbio_1_1alphabet_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range)....">bio::alphabet::phred42</a> letter at each position. <br /></td></tr>
<tr class="separator:aee3aa5dc0944a8c475a5be966eb4b389"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a75fb9b7d161113029421a7e0d559b6f6"><td class="memItemLeft" align="right" valign="top"><a id="a75fb9b7d161113029421a7e0d559b6f6" name="a75fb9b7d161113029421a7e0d559b6f6"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>dna5q</b> = <a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1dna5.html">dna5</a>, <a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a> &gt;</td></tr>
<tr class="memdesc:a75fb9b7d161113029421a7e0d559b6f6"><td class="mdescLeft">&#160;</td><td class="mdescRight">An alphabet that stores a <a class="el" href="classbio_1_1alphabet_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N..">bio::alphabet::dna5</a> letter and an <a class="el" href="classbio_1_1alphabet_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range)....">bio::alphabet::phred42</a> letter at each position. <br /></td></tr>
<tr class="separator:a75fb9b7d161113029421a7e0d559b6f6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabf64db166268e1c52fc6fc5d00603003"><td class="memTemplParams" colspan="2">template&lt;typename alphabet_t &gt; </td></tr>
<tr class="memitem:gabf64db166268e1c52fc6fc5d00603003"><td class="memTemplItemLeft" align="right" valign="top">using&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__gap.html#gabf64db166268e1c52fc6fc5d00603003">gapped</a> = <a class="el" href="classbio_1_1alphabet_1_1variant.html">variant</a>&lt; alphabet_t, <a class="el" href="classbio_1_1alphabet_1_1gap.html">gap</a> &gt;</td></tr>
<tr class="memdesc:gabf64db166268e1c52fc6fc5d00603003"><td class="mdescLeft">&#160;</td><td class="mdescRight">Extends a given alphabet with a gap character.  <br /></td></tr>
<tr class="separator:gabf64db166268e1c52fc6fc5d00603003"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabd0d2f8bc6672c42a829132de848745d"><td class="memTemplParams" colspan="2">
template&lt;typename alphabet_type &gt; </td></tr>
<tr class="memitem:gabd0d2f8bc6672c42a829132de848745d"><td class="memTemplItemLeft" align="right" valign="top">using&#160;</td><td class="memTemplItemRight" valign="bottom"><b>phred_t</b> = decltype(<a class="el" href="group__quality.html#gac7faff4dee7175ef96d5f94479453df2">bio::alphabet::to_phred</a>(<a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/declval.html">std::declval</a>&lt; alphabet_type &gt;()))</td></tr>
<tr class="memdesc:gabd0d2f8bc6672c42a829132de848745d"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <code>phred_type</code> of the alphabet; defined as the return type of <a class="el" href="group__quality.html#gac7faff4dee7175ef96d5f94479453df2" title="The public getter function for the phred representation of a quality score.">bio::alphabet::to_phred</a>. <br /></td></tr>
<tr class="separator:gabd0d2f8bc6672c42a829132de848745d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa96971d545b8ab7260d1023b83acb63a"><td class="memTemplParams" colspan="2">
template&lt;typename semi_alphabet_type &gt; </td></tr>
<tr class="memitem:gaa96971d545b8ab7260d1023b83acb63a"><td class="memTemplItemLeft" align="right" valign="top">using&#160;</td><td class="memTemplItemRight" valign="bottom"><b>rank_t</b> = decltype(<a class="el" href="group__alphabet.html#ga2a48a5c1883184e362f09e32ac33241e">bio::alphabet::to_rank</a>(<a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/declval.html">std::declval</a>&lt; semi_alphabet_type &gt;()))</td></tr>
<tr class="memdesc:gaa96971d545b8ab7260d1023b83acb63a"><td class="mdescLeft">&#160;</td><td class="mdescRight">The <code>rank_type</code> of the semi-alphabet; defined as the return type of <a class="el" href="group__alphabet.html#ga2a48a5c1883184e362f09e32ac33241e" title="Return the rank representation of a (semi-)alphabet object.">bio::alphabet::to_rank</a>. <br /></td></tr>
<tr class="separator:gaa96971d545b8ab7260d1023b83acb63a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3025f9f9c14e04ca2532df256e0a2d3a"><td class="memItemLeft" align="right" valign="top"><a id="a3025f9f9c14e04ca2532df256e0a2d3a" name="a3025f9f9c14e04ca2532df256e0a2d3a"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>rna15q</b> = <a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1rna15.html">rna15</a>, <a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a> &gt;</td></tr>
<tr class="memdesc:a3025f9f9c14e04ca2532df256e0a2d3a"><td class="mdescLeft">&#160;</td><td class="mdescRight">An alphabet that stores a <a class="el" href="classbio_1_1alphabet_1_1rna15.html" title="The 15 letter RNA alphabet, containing all IUPAC smybols minus the gap..">bio::alphabet::rna15</a> letter and an <a class="el" href="classbio_1_1alphabet_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::alphabet::qualified</a> letter at each position. <br /></td></tr>
<tr class="separator:a3025f9f9c14e04ca2532df256e0a2d3a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae752183f71604cbe30e27c9fdaf7a2a5"><td class="memItemLeft" align="right" valign="top"><a id="ae752183f71604cbe30e27c9fdaf7a2a5" name="ae752183f71604cbe30e27c9fdaf7a2a5"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>rna4q</b> = <a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1rna4.html">rna4</a>, <a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a> &gt;</td></tr>
<tr class="memdesc:ae752183f71604cbe30e27c9fdaf7a2a5"><td class="mdescLeft">&#160;</td><td class="mdescRight">An alphabet that stores a <a class="el" href="classbio_1_1alphabet_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U..">bio::alphabet::rna4</a> letter and an <a class="el" href="classbio_1_1alphabet_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range)....">bio::alphabet::phred42</a> letter at each position. <br /></td></tr>
<tr class="separator:ae752183f71604cbe30e27c9fdaf7a2a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a310d5422eb8f67cb0b1ef07b7786e9d6"><td class="memItemLeft" align="right" valign="top"><a id="a310d5422eb8f67cb0b1ef07b7786e9d6" name="a310d5422eb8f67cb0b1ef07b7786e9d6"></a>
using&#160;</td><td class="memItemRight" valign="bottom"><b>rna5q</b> = <a class="el" href="classbio_1_1alphabet_1_1qualified.html">qualified</a>&lt; <a class="el" href="classbio_1_1alphabet_1_1rna5.html">rna5</a>, <a class="el" href="classbio_1_1alphabet_1_1phred42.html">phred42</a> &gt;</td></tr>
<tr class="memdesc:a310d5422eb8f67cb0b1ef07b7786e9d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">An alphabet that stores a <a class="el" href="classbio_1_1alphabet_1_1rna5.html" title="The five letter RNA alphabet of A,C,G,U and the unknown character N..">bio::alphabet::rna5</a> letter and an <a class="el" href="classbio_1_1alphabet_1_1phred42.html" title="Quality type for traditional Sanger and modern Illumina Phred scores (typical range)....">bio::alphabet::phred42</a> letter at each position. <br /></td></tr>
<tr class="separator:a310d5422eb8f67cb0b1ef07b7786e9d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="enum-members" name="enum-members"></a>
Enumerations</h2></td></tr>
<tr class="memitem:gaa0d4217bd42a17a19c079ede29a22172"><td class="memItemLeft" align="right" valign="top">enum struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#gaa0d4217bd42a17a19c079ede29a22172">genetic_code</a> : uint8_t { <b>CANONICAL</b> = 1
 }</td></tr>
<tr class="memdesc:gaa0d4217bd42a17a19c079ede29a22172"><td class="mdescLeft">&#160;</td><td class="mdescRight">Genetic codes used for translation of nucleotides into amino acids.  <a href="group__aminoacid.html#gaa0d4217bd42a17a19c079ede29a22172">More...</a><br /></td></tr>
<tr class="separator:gaa0d4217bd42a17a19c079ede29a22172"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6f577a16bf27585908616e19371b2e8a"><td class="memItemLeft" align="right" valign="top">enum class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> : uint8_t { <br />
&#160;&#160;<a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aae2a0a67be49369d2de4d5a5508768614">translation_frames::FWD_FRAME_0</a> = 1
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aabe4e69cb05e6b1ceb3ac154abfae144a">translation_frames::FWD_FRAME_1</a> = 1 &lt;&lt; 1
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aa98bcbb2fee2fb9460d27c485cab912d9">translation_frames::FWD_FRAME_2</a> = 1 &lt;&lt; 2
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aa549af46ce0321e25697124b4636506bb">translation_frames::REV_FRAME_0</a> = 1 &lt;&lt; 3
, <br />
&#160;&#160;<a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aa7366ff691b198a53e51723d2e46ba861">translation_frames::REV_FRAME_1</a> = 1 &lt;&lt; 4
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aacfa6260521090681eed88c1bf1ddbf03">translation_frames::REV_FRAME_2</a> = 1 &lt;&lt; 5
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aa088caa4dc3d1f6cfa30f0b021d98c5ad">translation_frames::FWD_REV_0</a> = FWD_FRAME_0 | REV_FRAME_0
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aae8b73dbff37c98f043083f9d5a973123">translation_frames::FWD_REV_1</a> = FWD_FRAME_1 | REV_FRAME_1
, <br />
&#160;&#160;<a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aa7955521726f0f8eb020623ce73a3c514">translation_frames::FWD_REV_2</a> = FWD_FRAME_2 | REV_FRAME_2
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aac9261effebe177a8679909dd7ec2552e">translation_frames::FWD</a> = FWD_FRAME_0 | FWD_FRAME_1 | FWD_FRAME_2
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aad51e4632904a9c370000c19748a91d66">translation_frames::REV</a> = REV_FRAME_0 | REV_FRAME_1 | REV_FRAME_2
, <a class="el" href="group__aminoacid.html#gga6f577a16bf27585908616e19371b2e8aabd54c9328aea3fb30efc906118a5d25f">translation_frames::SIX_FRAME</a> = FWD | REV
<br />
 }</td></tr>
<tr class="memdesc:ga6f577a16bf27585908616e19371b2e8a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Specialisation values for single and multiple translation frames.  <a href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">More...</a><br /></td></tr>
<tr class="separator:ga6f577a16bf27585908616e19371b2e8a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="func-members" name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga32a00ef9570995f13d47d3cac9e0f5a5"><td class="memTemplParams" colspan="2">template&lt;<a class="el" href="group__aminoacid.html#gaa0d4217bd42a17a19c079ede29a22172">genetic_code</a> gc = genetic_code::CANONICAL, nucleotide nucl_type&gt; </td></tr>
<tr class="memitem:ga32a00ef9570995f13d47d3cac9e0f5a5"><td class="memTemplItemLeft" align="right" valign="top">constexpr <a class="el" href="classbio_1_1alphabet_1_1aa27.html">aa27</a>&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__aminoacid.html#ga32a00ef9570995f13d47d3cac9e0f5a5">translate_triplet</a> (nucl_type const &amp;n1, nucl_type const &amp;n2, nucl_type const &amp;n3) noexcept</td></tr>
<tr class="memdesc:ga32a00ef9570995f13d47d3cac9e0f5a5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Translate one nucleotide triplet into single amino acid (single nucleotide interface).  <br /></td></tr>
<tr class="separator:ga32a00ef9570995f13d47d3cac9e0f5a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Binary operators for bio::alphabet::translation_frames.</div></td></tr>
<tr><td colspan="2"><div class="groupText"><p>Perform binary operations like on ints or weak enums. </p>
</div></td></tr>
<tr class="memitem:a2745a775999e24744174741d7e0ebd76"><td class="memItemLeft" align="right" valign="top"><a id="a2745a775999e24744174741d7e0ebd76" name="a2745a775999e24744174741d7e0ebd76"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a>&#160;</td><td class="memItemRight" valign="bottom"><b>operator&amp;</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> lhs, <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> rhs) noexcept</td></tr>
<tr class="memdesc:a2745a775999e24744174741d7e0ebd76"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:a2745a775999e24744174741d7e0ebd76"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aeb4e6ec76811b98d61d0eeddb113b004"><td class="memItemLeft" align="right" valign="top"><a id="aeb4e6ec76811b98d61d0eeddb113b004" name="aeb4e6ec76811b98d61d0eeddb113b004"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a>&#160;</td><td class="memItemRight" valign="bottom"><b>operator|</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> lhs, <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> rhs) noexcept</td></tr>
<tr class="memdesc:aeb4e6ec76811b98d61d0eeddb113b004"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:aeb4e6ec76811b98d61d0eeddb113b004"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7734ea719a38023eb7a90b5a77406c7a"><td class="memItemLeft" align="right" valign="top"><a id="a7734ea719a38023eb7a90b5a77406c7a" name="a7734ea719a38023eb7a90b5a77406c7a"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a>&#160;</td><td class="memItemRight" valign="bottom"><b>operator^</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> lhs, <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> rhs) noexcept</td></tr>
<tr class="memdesc:a7734ea719a38023eb7a90b5a77406c7a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:a7734ea719a38023eb7a90b5a77406c7a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae98abba3166e07ba30241d4643fc714"><td class="memItemLeft" align="right" valign="top"><a id="aae98abba3166e07ba30241d4643fc714" name="aae98abba3166e07ba30241d4643fc714"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a>&#160;</td><td class="memItemRight" valign="bottom"><b>operator~</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> lhs) noexcept</td></tr>
<tr class="memdesc:aae98abba3166e07ba30241d4643fc714"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:aae98abba3166e07ba30241d4643fc714"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7c523bbc4c7a79ce5aa5b1d09b0830fe"><td class="memItemLeft" align="right" valign="top"><a id="a7c523bbc4c7a79ce5aa5b1d09b0830fe" name="a7c523bbc4c7a79ce5aa5b1d09b0830fe"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator&amp;=</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> &amp;lhs, <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> rhs) noexcept</td></tr>
<tr class="memdesc:a7c523bbc4c7a79ce5aa5b1d09b0830fe"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:a7c523bbc4c7a79ce5aa5b1d09b0830fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac4cf344565c6118330edd6e4cf86ca0c"><td class="memItemLeft" align="right" valign="top"><a id="ac4cf344565c6118330edd6e4cf86ca0c" name="ac4cf344565c6118330edd6e4cf86ca0c"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator|=</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> &amp;lhs, <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> rhs) noexcept</td></tr>
<tr class="memdesc:ac4cf344565c6118330edd6e4cf86ca0c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:ac4cf344565c6118330edd6e4cf86ca0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2802f8c473f0b830482296375b56842c"><td class="memItemLeft" align="right" valign="top"><a id="a2802f8c473f0b830482296375b56842c" name="a2802f8c473f0b830482296375b56842c"></a>
constexpr <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>operator^=</b> (<a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> &amp;lhs, <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a">translation_frames</a> rhs) noexcept</td></tr>
<tr class="memdesc:a2802f8c473f0b830482296375b56842c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Binary operators for <a class="el" href="group__aminoacid.html#ga6f577a16bf27585908616e19371b2e8a" title="Specialisation values for single and multiple translation frames.">bio::alphabet::translation_frames</a>. <br /></td></tr>
<tr class="separator:a2802f8c473f0b830482296375b56842c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a id="var-members" name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:ga299ddd5bf7e8221cb9d283f8a8d24f8d"><td class="memTemplParams" colspan="2">template&lt;typename alph_t , typename wrap_t  = meta::default_initialisable_wrap_t&lt;alph_t&gt;&gt; </td></tr>
<tr class="memitem:ga299ddd5bf7e8221cb9d283f8a8d24f8d"><td class="memTemplItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__alphabet.html#ga299ddd5bf7e8221cb9d283f8a8d24f8d">size</a></td></tr>
<tr class="memdesc:ga299ddd5bf7e8221cb9d283f8a8d24f8d"><td class="mdescLeft">&#160;</td><td class="mdescRight">A type trait that holds the size of a (semi-)alphabet.  <br /></td></tr>
<tr class="separator:ga299ddd5bf7e8221cb9d283f8a8d24f8d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Function objects</div></td></tr>
<tr class="memitem:ga2a48a5c1883184e362f09e32ac33241e"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__alphabet.html#ga2a48a5c1883184e362f09e32ac33241e">to_rank</a></td></tr>
<tr class="memdesc:ga2a48a5c1883184e362f09e32ac33241e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the rank representation of a (semi-)alphabet object.  <br /></td></tr>
<tr class="separator:ga2a48a5c1883184e362f09e32ac33241e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab718eab4e0f0e1b44f78b7600dee5e4d"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__alphabet.html#gab718eab4e0f0e1b44f78b7600dee5e4d">assign_rank_to</a></td></tr>
<tr class="memdesc:gab718eab4e0f0e1b44f78b7600dee5e4d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign a rank to an alphabet object.  <br /></td></tr>
<tr class="separator:gab718eab4e0f0e1b44f78b7600dee5e4d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga064e386f9360cca0fd3702b4957a27a0"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__alphabet.html#ga064e386f9360cca0fd3702b4957a27a0">to_char</a></td></tr>
<tr class="memdesc:ga064e386f9360cca0fd3702b4957a27a0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the char representation of an alphabet object.  <br /></td></tr>
<tr class="separator:ga064e386f9360cca0fd3702b4957a27a0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0c2eb25dabf59ccf9992b2661083ea5b"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__alphabet.html#ga0c2eb25dabf59ccf9992b2661083ea5b">assign_char_to</a></td></tr>
<tr class="memdesc:ga0c2eb25dabf59ccf9992b2661083ea5b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign a char to an alphabet object.  <br /></td></tr>
<tr class="separator:ga0c2eb25dabf59ccf9992b2661083ea5b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga64c00e15570d3e7394f425c810e617ce"><td class="memTemplParams" colspan="2">template&lt;typename alph_t &gt; </td></tr>
<tr class="memitem:ga64c00e15570d3e7394f425c810e617ce"><td class="memTemplItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="group__alphabet.html#ga64c00e15570d3e7394f425c810e617ce">char_is_valid_for</a></td></tr>
<tr class="memdesc:ga64c00e15570d3e7394f425c810e617ce"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns whether a character is in the valid set of a <a class="el" href="conceptbio_1_1alphabet_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet::alphabet</a> (usually implies a bijective mapping to an alphabet value).  <br /></td></tr>
<tr class="separator:ga64c00e15570d3e7394f425c810e617ce"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gadec8dc2856700f9763590213b6cae625"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__alphabet.html#gadec8dc2856700f9763590213b6cae625">assign_char_strictly_to</a></td></tr>
<tr class="memdesc:gadec8dc2856700f9763590213b6cae625"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign a character to an alphabet object, throw if the character is not valid.  <br /></td></tr>
<tr class="separator:gadec8dc2856700f9763590213b6cae625"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Function objects (Nucleotide)</div></td></tr>
<tr class="memitem:ga174261f8f888bbce628ffe01097a9b48"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__nucleotide.html#ga174261f8f888bbce628ffe01097a9b48">complement</a></td></tr>
<tr class="memdesc:ga174261f8f888bbce628ffe01097a9b48"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return the complement of a nucleotide object.  <br /></td></tr>
<tr class="separator:ga174261f8f888bbce628ffe01097a9b48"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr><td colspan="2"><div class="groupHeader">Function objects (Quality)</div></td></tr>
<tr class="memitem:gac7faff4dee7175ef96d5f94479453df2"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__quality.html#gac7faff4dee7175ef96d5f94479453df2">to_phred</a></td></tr>
<tr class="memdesc:gac7faff4dee7175ef96d5f94479453df2"><td class="mdescLeft">&#160;</td><td class="mdescRight">The public getter function for the phred representation of a quality score.  <br /></td></tr>
<tr class="separator:gac7faff4dee7175ef96d5f94479453df2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8ea6aae1e3eb068822e6d7dd5f73f78d"><td class="memItemLeft" align="right" valign="top">constexpr auto&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__quality.html#ga8ea6aae1e3eb068822e6d7dd5f73f78d">assign_phred_to</a></td></tr>
<tr class="memdesc:ga8ea6aae1e3eb068822e6d7dd5f73f78d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Assign a phred score to a quality alphabet object.  <br /></td></tr>
<tr class="separator:ga8ea6aae1e3eb068822e6d7dd5f73f78d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>The alphabet module's namespace. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="biocpp_customisation.html">Customisation</a> </dd></dl>
</div></div><!-- contents -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="navelem"><a class="el" href="namespacebio.html">bio</a></li><li class="navelem"><a class="el" href="namespacebio_1_1alphabet.html">alphabet</a></li>
    <li class="footer">Generated on Thu Mar 30 2023 15:25:52 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.6 </li>
  </ul>
</div>
</body>
</html>
