<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>BioC++: Alphabets in SeqAn</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() { init_search(); });
/* @license-end */
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="biocpp.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="biocpp_logo.svg"/></td>
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">BioC++
   &#160;<span id="projectnumber">3.0.2</span>
   </div>
   <div id="projectbrief">The Modern C++ libraries for Bioinformatics.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.svg"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.svg" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('tutorial_alphabets.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="PageDoc"><div class="header">
  <div class="headertitle">
<div class="title">Alphabets in SeqAn </div>  </div>
</div><!--header-->
<div class="contents">
<div class="toc"><h3>Table of Contents</h3>
<ul><li class="level1"><a href="#autotoc_md14">Introduction</a></li>
<li class="level1"><a href="#autotoc_md15">The nucleotide alphabets</a><ul><li class="level2"><a href="#autotoc_md16">Construction and assignment of alphabet symbols</a></li>
<li class="level2"><a href="#autotoc_md17">The rank of an alphabet symbol</a></li>
<li class="level2"><a href="#autotoc_md18">The char representation of an alphabet symbol</a></li>
<li class="level2"><a href="#autotoc_md19">Obtaining the alphabet size</a></li>
<li class="level2"><a href="#autotoc_md20">containers over alphabets</a></li>
<li class="level2"><a href="#autotoc_md21">Comparability</a></li>
<li class="level2"><a href="#autotoc_md22">Example</a></li>
</ul>
</li>
<li class="level1"><a href="#autotoc_md23">Other alphabets</a><ul><li class="level2"><a href="#autotoc_md24">The amino acid alphabet</a></li>
<li class="level2"><a href="#autotoc_md25">Structure and quality alphabets</a></li>
<li class="level2"><a href="#autotoc_md26">Gap alphabet</a></li>
</ul>
</li>
</ul>
</div>
<div class="textblock"><p><a class="anchor" id="md__home_hannes_devel_bio_core_doc_biocpp_core_tutorial_alphabet_index"></a> <em><b>Learning Objective:</b></em></p>
<p>In this tutorial we look at alphabets and you will learn how to work with nucleotides and amino acids in SeqAn. We guide you through the most important properties of SeqAn's alphabets and show you the different implemented types. After completion, you will be able to use the alphabets inside of STL containers and to compare alphabet values.</p>
<table class="doxtable">
<tr>
<th>Difficulty</th><td>Easy</td></tr>
<tr>
<th>Duration</th><td>45 min</td></tr>
<tr>
<th>Prerequisite tutorials</th><td><a class="el" href="setup.html">Quick Setup (using CMake)</a></td></tr>
<tr>
<th>Recommended reading</th><td>None</td></tr>
</table>
<p>The links on this page mostly point straight into the API documentation which you should use as a reference. The code examples and assignments are designed to provide some practical experience with our interface as well as a code basis for your own program development.</p>
<h1><a class="anchor" id="autotoc_md14"></a>
Introduction</h1>
<p>An alphabet is the set of symbols of which a biological sequence – or in general a text – is composed. SeqAn implements specific and optimised alphabets not only for sequences of RNA, DNA and protein components, but also for quality, secondary structure and gap annotation as well as combinations of the aforementioned.</p>
<p> <div class="assignemnt"><h4>Task</h4><div class="assignment_content"> Read the section <em>Detailed Description</em> of the API reference page for <a class="el" href="group__alphabet.html">Alphabet</a>. This is a detailed introduction to the alphabet module and demonstrates its main advantages.  </div></div></p>
<h1><a class="anchor" id="autotoc_md15"></a>
The nucleotide alphabets</h1>
<p>Nucleotides are the components of (Deoxy)Ribonucleic acid (DNA/RNA) and contain one of the nucleobases Adenine (A), Cytosine (C), Guanine (G), Thymine (T, only DNA) and Uracil (U, only RNA). In SeqAn the alphabets <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">bio::dna4</a> and <a class="el" href="classbio_1_1rna4.html" title="The four letter RNA alphabet of A,C,G,U.">bio::rna4</a> contain exactly the four respective nucleotides. The trailed number in the alphabets' name represents the number of entities the alphabet holds – we denote this number as <em>alphabet size</em>. For instance, the alphabet <a class="el" href="classbio_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N.">bio::dna5</a> represents five entities as it contains the additional symbol 'N' to refer to an unknown nucleotide.</p>
<h2><a class="anchor" id="autotoc_md16"></a>
Construction and assignment of alphabet symbols</h2>
<p>Let's look at some example code which demonstrates how objects of the <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">bio::dna4</a> alphabet are assigned from characters.</p>
<div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="code" href="alphabet_2all_8hpp.html">bio/alphabet/all.hpp</a>&gt;</span> <span class="comment">// for working with alphabets directly</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna4;</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main ()</div>
<div class="line">{</div>
<div class="line">    <span class="comment">// Two objects of bio::dna4 alphabet constructed with a char literal.</span></div>
<div class="line">    <a class="code" href="classbio_1_1dna4.html">bio::dna4</a> ade = <span class="charliteral">&#39;A&#39;</span>_dna4;</div>
<div class="line">    <a class="code" href="classbio_1_1dna4.html">bio::dna4</a> gua = <span class="charliteral">&#39;G&#39;</span>_dna4;</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Two additional objects assigned explicitly from char or rank.</span></div>
<div class="line">    <a class="code" href="classbio_1_1dna4.html">bio::dna4</a> cyt, thy;</div>
<div class="line">    cyt.<a class="code" href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">assign_char</a>(<span class="charliteral">&#39;C&#39;</span>);</div>
<div class="line">    thy.<a class="code" href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">assign_rank</a>(3);</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Further code here...</span></div>
<div class="ttc" id="aalphabet_2all_8hpp_html"><div class="ttname"><a href="alphabet_2all_8hpp.html">all.hpp</a></div><div class="ttdoc">Meta-header for the alphabet module.</div></div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_a822d399ab808842870a4a4feb5490849"><div class="ttname"><a href="classbio_1_1alphabet__base.html#a822d399ab808842870a4a4feb5490849">bio::alphabet_base::assign_char</a></div><div class="ttdeci">constexpr derived_type &amp; assign_char(char_type const c) noexcept</div><div class="ttdoc">Assign from a character, implicitly converts invalid characters.</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:143</div></div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_acdd042ae2f04d986f65a7b60edd6e83e"><div class="ttname"><a href="classbio_1_1alphabet__base.html#acdd042ae2f04d986f65a7b60edd6e83e">bio::alphabet_base::assign_rank</a></div><div class="ttdeci">constexpr derived_type &amp; assign_rank(rank_type const c) noexcept</div><div class="ttdoc">Assign from a numeric value.</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:168</div></div>
<div class="ttc" id="aclassbio_1_1dna4_html"><div class="ttname"><a href="classbio_1_1dna4.html">bio::dna4</a></div><div class="ttdoc">The four letter DNA alphabet of A,C,G,T.</div><div class="ttdef"><b>Definition:</b> dna4.hpp:49</div></div>
</div><!-- fragment --><div class="fragment"><div class="line">    <span class="keywordflow">return</span> 0;</div>
<div class="line">}</div>
</div><!-- fragment --><p> We have shown three solutions for assigning variables of alphabet type.</p><ol type="1">
<li>Construction by character literal, i.e. appending the operator <code>_dna4</code> to the respective char symbol. <br  />
 This is the handiest way as it can be also used as a temporary object.</li>
<li>Assignment by <code>char</code> via the global function bio::assign_char. <br  />
 This is useful if the assignment target already exists, e.g. in a sequence vector.</li>
<li>Assignment by rank via the global function bio::assign_rank. <br  />
 May be used when the <em>rank</em> is known.</li>
</ol>
<h2><a class="anchor" id="autotoc_md17"></a>
The rank of an alphabet symbol</h2>
<p>The rank of a symbol is a number in range [0..alphabet_size) where each number is paired with an alphabet symbol by a bijective function. In SeqAn the rank is always determined by the lexicographical order of the underlying characters. For instance, in <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">bio::dna4</a> the bijection is <br  />
 <code>'A'_dna4 ⟼ 0</code> <br  />
 <code>'C'_dna4 ⟼ 1</code> <br  />
 <code>'G'_dna4 ⟼ 2</code> <br  />
 <code>'T'_dna4 ⟼ 3</code>.</p>
<p>SeqAn provides the function <a class="el" href="group__alphabet.html#ga9189b35f800d7e5abad51e021ea811a0" title="Return the rank representation of a (semi-)alphabet object.">bio::to_rank</a> for converting a symbol to its rank value as demonstrated in the following code example. Note that the data type of the rank is usually the smallest possible unsigned type that is required for storing the values of the alphabet.</p>
<div class="fragment"><div class="line">    <span class="comment">// Get the rank type of the alphabet (here uint8_t).</span></div>
<div class="line">    <span class="keyword">using</span> rank_type = <a class="code" href="group__alphabet.html#ga80d7465944e58981cc10298677b82fec">bio::alphabet_rank_t&lt;bio::dna4&gt;</a>;</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Retrieve the numerical representation (rank) of the objects.</span></div>
<div class="line">    rank_type rank_a = ade.<a class="code" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">to_rank</a>();   <span class="comment">// =&gt; 0</span></div>
<div class="line">    rank_type rank_g = gua.<a class="code" href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">to_rank</a>();   <span class="comment">// =&gt; 2</span></div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_a443c2f7716e384ea2acec1d2bced7c8b"><div class="ttname"><a href="classbio_1_1alphabet__base.html#a443c2f7716e384ea2acec1d2bced7c8b">bio::alphabet_base::to_rank</a></div><div class="ttdeci">constexpr rank_type to_rank() const noexcept</div><div class="ttdoc">Return the letter's numeric value (rank in the alphabet).</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:119</div></div>
<div class="ttc" id="agroup__alphabet_html_ga80d7465944e58981cc10298677b82fec"><div class="ttname"><a href="group__alphabet.html#ga80d7465944e58981cc10298677b82fec">bio::alphabet_rank_t</a></div><div class="ttdeci">decltype(bio::to_rank(std::declval&lt; semi_alphabet_type &gt;())) alphabet_rank_t</div><div class="ttdoc">The rank_type of the semi-alphabet; defined as the return type of bio::to_rank.</div><div class="ttdef"><b>Definition:</b> concept.hpp:153</div></div>
</div><!-- fragment --> <h2><a class="anchor" id="autotoc_md18"></a>
The char representation of an alphabet symbol</h2>
<p>Our alphabets also have a character representation because it is more intuitive to work with them than using the rank. Each alphabet symbol is represented by its respective character whenever possible (<code>A ⟼ 'A'</code>). Analogously to the rank, SeqAn provides the function <a class="el" href="group__alphabet.html#gab295f04aea840000aafc640b5e6528c8" title="Return the char representation of an alphabet object.">bio::to_char</a> for converting a symbol to its character representation.</p>
<div class="fragment"><div class="line">    <span class="comment">// Get the character type of the alphabet (here char).</span></div>
<div class="line">    <span class="keyword">using</span> char_type = <a class="code" href="group__alphabet.html#ga474c9ea1cc2c81b65e22ee09c746fc17">bio::alphabet_char_t&lt;bio::dna4&gt;</a>;</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Retrieve the character representation.</span></div>
<div class="line">    char_type char_a = ade.<a class="code" href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">to_char</a>();   <span class="comment">// =&gt; &#39;A&#39;</span></div>
<div class="line">    char_type char_g = gua.<a class="code" href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">to_char</a>();   <span class="comment">// =&gt; &#39;G&#39;</span></div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_a6616ba2cb25069f5192589ad53651a88"><div class="ttname"><a href="classbio_1_1alphabet__base.html#a6616ba2cb25069f5192589ad53651a88">bio::alphabet_base::to_char</a></div><div class="ttdeci">constexpr char_type to_char() const noexcept</div><div class="ttdoc">Return the letter as a character of char_type.</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:97</div></div>
<div class="ttc" id="agroup__alphabet_html_ga474c9ea1cc2c81b65e22ee09c746fc17"><div class="ttname"><a href="group__alphabet.html#ga474c9ea1cc2c81b65e22ee09c746fc17">bio::alphabet_char_t</a></div><div class="ttdeci">decltype(bio::to_char(std::declval&lt; alphabet_type const  &gt;())) alphabet_char_t</div><div class="ttdoc">The char_type of the alphabet; defined as the return type of bio::to_char.</div><div class="ttdef"><b>Definition:</b> concept.hpp:331</div></div>
</div><!-- fragment --><p> Above you have seen that you can assign an alphabet symbol from a character with bio::from_char. In contrast to the rank interface, this assignment is not a bijection because the whole spectrum of available chars is mapped to values inside the alphabet. For instance, assigning to <a class="el" href="classbio_1_1dna4.html" title="The four letter DNA alphabet of A,C,G,T.">bio::dna4</a> from any character other than <code>C</code>, <code>G</code> or <code>T</code> results in the value <code>'A'_dna4</code> and assigning from any character except <code>A</code>, <code>C</code>, <code>G</code> or <code>T</code> to <a class="el" href="classbio_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N.">bio::dna5</a> results in the value <code>'N'_dna5</code>. You can avoid the implicit conversion by using bio::assign_char_strict which throws <a class="el" href="structbio_1_1invalid__char__assignment.html" title="An exception typically thrown by bio::alphabet::assign_char_strict.">bio::invalid_char_assignment</a> on invalid characters.</p>
<div class="fragment"><div class="line">    <span class="comment">// Assign from character with value check.</span></div>
<div class="line">    <a class="code" href="group__alphabet.html#gab055c8c702ef4b8b48b7d301c12b18b2">bio::assign_char_strictly_to</a>(<span class="charliteral">&#39;C&#39;</span>, cyt);</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// bio::assign_char_strictly_to(&#39;X&#39;, thy); // would throw bio::invalid_char_assignment</span></div>
<div class="ttc" id="agroup__alphabet_html_gab055c8c702ef4b8b48b7d301c12b18b2"><div class="ttname"><a href="group__alphabet.html#gab055c8c702ef4b8b48b7d301c12b18b2">bio::assign_char_strictly_to</a></div><div class="ttdeci">constexpr auto assign_char_strictly_to</div><div class="ttdoc">Assign a character to an alphabet object, throw if the character is not valid.</div><div class="ttdef"><b>Definition:</b> concept.hpp:599</div></div>
</div><!-- fragment --><p> You can test the validity of a character by calling <a class="el" href="group__alphabet.html#ga510b0657b8a3ae7d304999afd8fdcd52" title="Returns whether a character is in the valid set of a bio::alphabet (usually implies a bijective mappi...">bio::char_is_valid_for</a>. It returns true if the character is valid and false otherwise.</p>
<h2><a class="anchor" id="autotoc_md19"></a>
Obtaining the alphabet size</h2>
<p>You can retrieve the alphabet size by accessing the class member variable <code>alphabet_size</code> which is implemented in most <a class="el" href="interfacebio_1_1alphabet.html" title="The generic alphabet concept that covers most data types used in ranges.">bio::alphabet</a> instances.</p>
<div class="fragment"><div class="line">    <span class="comment">// Get the alphabet size as class member of the alphabet.</span></div>
<div class="line">    uint8_t <span class="keyword">const</span> size1 = <a class="code" href="classbio_1_1alphabet__base.html#a989cf18d33d15b0d858c2dd540480468">bio::dna4::alphabet_size</a>;        <span class="comment">// =&gt; 4</span></div>
<div class="ttc" id="aclassbio_1_1alphabet__base_html_a989cf18d33d15b0d858c2dd540480468"><div class="ttname"><a href="classbio_1_1alphabet__base.html#a989cf18d33d15b0d858c2dd540480468">bio::alphabet_base::alphabet_size</a></div><div class="ttdeci">static constexpr detail::min_viable_uint_t&lt; size &gt; alphabet_size</div><div class="ttdoc">The size of the alphabet, i.e. the number of different values it can take.</div><div class="ttdef"><b>Definition:</b> alphabet_base.hpp:177</div></div>
</div><!-- fragment --> <h2><a class="anchor" id="autotoc_md20"></a>
containers over alphabets</h2>
<p>In SeqAn you can use the STL containers to model e.g. sequences, sets or mappings with our alphabets. The following example shows some exemplary contexts for their use. For <b>sequences</b> we recommend the <a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a> with one of SeqAn's alphabet types. Please note how easily a sequence can be created via the string literal.</p>
<div class="fragment"><div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna5;</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Examples of different container types with SeqAn&#39;s alphabets.</span></div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector&lt;bio::dna5&gt;</a> dna_sequence{<span class="stringliteral">&quot;GATTANAG&quot;</span>_dna5};</div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/utility/pair.html">std::pair&lt;bio::gapped&lt;bio::dna4&gt;</a>, <a class="code" href="classbio_1_1alphabet__variant.html">bio::gapped&lt;bio::dna4&gt;</a>&gt; alignment_column{<a class="code" href="classbio_1_1gap.html">bio::gap</a>{}, thy};</div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/set.html">std::set&lt;bio::dna4&gt;</a> pyrimidines{<span class="charliteral">&#39;C&#39;</span>_dna4, <span class="charliteral">&#39;T&#39;</span>_dna4};</div>
<div class="ttc" id="aclassbio_1_1alphabet__variant_html"><div class="ttname"><a href="classbio_1_1alphabet__variant.html">bio::alphabet_variant</a></div><div class="ttdoc">A combined alphabet that can hold values of either of its alternatives.</div><div class="ttdef"><b>Definition:</b> alphabet_variant.hpp:120</div></div>
<div class="ttc" id="aclassbio_1_1gap_html"><div class="ttname"><a href="classbio_1_1gap.html">bio::gap</a></div><div class="ttdoc">The alphabet of a gap character '-'.</div><div class="ttdef"><b>Definition:</b> gap.hpp:38</div></div>
<div class="ttc" id="apair_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/utility/pair.html">std::pair</a></div></div>
<div class="ttc" id="aset_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/container/set.html">std::set</a></div></div>
<div class="ttc" id="avector_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector</a></div></div>
</div><!-- fragment --> <h2><a class="anchor" id="autotoc_md21"></a>
Comparability</h2>
<p>All alphabets in SeqAn are regular and comparable. This means that you can use the <code>&lt;</code>, <code>&lt;=</code>, <code>&gt;</code> and <code>&gt;=</code> operators to compare the values based on the rank. For instance, this is useful for sorting a text over the alphabet. Regular implies that the equality and inequality of two values can be tested with <code>==</code> and <code>!=</code>.</p>
<div class="fragment"><div class="line">    <span class="comment">// Equality and comparison of seqan::dna4 symbols.</span></div>
<div class="line">    <span class="keywordtype">bool</span> eq  = (cyt == <span class="charliteral">&#39;C&#39;</span>_dna4); <span class="comment">// true</span></div>
<div class="line">    <span class="keywordtype">bool</span> neq = (thy != <span class="charliteral">&#39;C&#39;</span>_dna4); <span class="comment">// true</span></div>
<div class="line">    <span class="keywordtype">bool</span> geq = (cyt &gt;= <span class="charliteral">&#39;C&#39;</span>_dna4); <span class="comment">// true</span></div>
<div class="line">    <span class="keywordtype">bool</span> gt  = (thy &gt;  <span class="charliteral">&#39;C&#39;</span>_dna4); <span class="comment">// true</span></div>
<div class="line">    <span class="keywordtype">bool</span> seq = (cyt &lt;= <span class="charliteral">&#39;C&#39;</span>_dna4); <span class="comment">// true</span></div>
<div class="line">    <span class="keywordtype">bool</span> st  = (ade &lt;  <span class="charliteral">&#39;C&#39;</span>_dna4); <span class="comment">// true</span></div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Sort a vector of symbols.</span></div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector&lt;bio::dna4&gt;</a> some_nucl{<span class="stringliteral">&quot;GTA&quot;</span>_dna4};</div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/algorithm/sort.html">std::sort</a>(some_nucl.begin(), some_nucl.end()); <span class="comment">// some_nucl: &quot;AGT&quot;</span></div>
<div class="ttc" id="asort_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/algorithm/sort.html">std::sort</a></div><div class="ttdeci">T sort(T... args)</div></div>
</div><!-- fragment --> <h2><a class="anchor" id="autotoc_md22"></a>
Example</h2>
<p>To wrap up this section on the nucleotide alphabet, the following assignment lets you practise the use of a SeqAn alphabet and its related functions. It will also show you a handy advantage of using a vector over an alphabet instead of using <code><a class="elRef" target="_blank" href="https://en.cppreference.com/w/cpp/string/basic_string.html">std::string</a></code>: The rank representation can be used straight as an array index (opposed to using a map with logarithmic access times, for example).</p>
<p> <div class="assignemnt"><h4>Assignment 1: GC content of a sequence</h4><div class="assignment_content"> An important property of DNA and RNA sequences is the <em>GC content</em>, which is the percentage of nucleobases that are either Guanine or Cytosine. Given the nucleotide counts <img class="formulaInl" alt="$n_A$" src="form_11.png" width="23" height="12"/>, <img class="formulaInl" alt="$n_T$" src="form_12.png" width="23" height="12"/>, <img class="formulaInl" alt="$n_G$" src="form_13.png" width="23" height="12"/>, <img class="formulaInl" alt="$n_C$" src="form_14.png" width="23" height="12"/> the GC content <img class="formulaInl" alt="$c$" src="form_15.png" width="7" height="9"/> is calculated as </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ c = \frac{n_G + n_C}{n_A + n_T + n_G + n_C} \]" src="form_16.png" width="186" height="39"/>
</p>
<p> Write a program that</p><ol type="1">
<li>reads a sequence as command line argument into a vector of <a class="el" href="classbio_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N.">bio::dna5</a>,</li>
<li>counts the number of occurrences for each nucleotide in an array of size <code>alphabet size</code> and</li>
<li>calculates the GC content.</li>
</ol>
<p>The <a class="el" href="classbio_1_1dna5.html" title="The five letter DNA alphabet of A,C,G,T and the unknown character N.">bio::dna5</a> type ensures that invalid characters in the input sequence are converted to <code>'N'</code>. Note that these characters should not influence the GC content.</p>
<p>Pass the sequences <code>CATTACAG</code> (3/8 = 37.5%) and <code>ANNAGAT</code> (1/5 = 20%) to your program and check if your results are correct.</p>
<p> </div></div>  <div class="solution"> <details><summary><b>Solution</b></summary>  </p><div class="fragment"><div class="line"><span class="preprocessor">#include &lt;<a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/header/array.html">array</a>&gt;</span>     <span class="comment">// std::array</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/header/string.html">string</a>&gt;</span>    <span class="comment">// std::string</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/header/vector.html">vector</a>&gt;</span>    <span class="comment">// std::vector</span></div>
<div class="line"> </div>
<div class="line"><span class="preprocessor">#include &lt;sharg/all.hpp&gt;</span></div>
<div class="line"> </div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="alphabet_2all_8hpp.html">bio/alphabet/all.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="fmt_8hpp.html">bio/alphabet/fmt.hpp</a>&gt;</span></div>
<div class="line"><span class="preprocessor">#include &lt;<a class="code" href="ranges_2views_2all_8hpp.html">bio/ranges/views/all.hpp</a>&gt;</span>       <span class="comment">// optional: use views to convert the input string to a dna5 sequence</span></div>
<div class="line"> </div>
<div class="line"><span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_dna5;</div>
<div class="line"> </div>
<div class="line"><span class="keywordtype">int</span> main (<span class="keywordtype">int</span> argc, <span class="keywordtype">char</span> * argv[])</div>
<div class="line">{</div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/string/basic_string.html">std::string</a> input{};</div>
<div class="line">    sharg::argument_parser parser(<span class="stringliteral">&quot;GC-Content&quot;</span>, argc, argv);</div>
<div class="line">    parser.add_positional_option(input, <span class="stringliteral">&quot;Specify an input sequence.&quot;</span>);</div>
<div class="line">    <span class="keywordflow">try</span></div>
<div class="line">    {</div>
<div class="line">        parser.parse();</div>
<div class="line">    }</div>
<div class="line">    <span class="keywordflow">catch</span> (sharg::argument_parser_error <span class="keyword">const</span> &amp; ext) <span class="comment">// the input is invalid</span></div>
<div class="line">    {</div>
<div class="line">        fmt::print(stderr, <span class="stringliteral">&quot;[PARSER ERROR] {}\n&quot;</span>, ext.what());</div>
<div class="line">        <span class="keywordflow">return</span> 0;</div>
<div class="line">    }</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Option1: Convert the input to a dna5 sequence by copy-conversion</span></div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/vector.html">std::vector&lt;bio::dna5&gt;</a> <a class="code" href="interfacebio_1_1sequence.html">sequence</a>;</div>
<div class="line">    <a class="code" href="interfacebio_1_1sequence.html">sequence</a>.resize(input.size());</div>
<div class="line">    <span class="keywordflow">for</span> (<span class="keywordtype">size_t</span> i = 0; i &lt; input.size(); ++i)</div>
<div class="line">        <a class="code" href="group__alphabet.html#ga05da42eb0b9a48f4d1a554c7d3a9044d">bio::assign_char_to</a>(input[i], <a class="code" href="interfacebio_1_1sequence.html">sequence</a>[i]);</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Option1: use views for the conversion. Views will be introduced in the next chapter.</span></div>
<div class="line">    <span class="comment">//auto sequence = input | bio::views::char_to&lt;bio::dna5&gt;;</span></div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Initialise an array with count values for dna5 symbols.</span></div>
<div class="line">    <a class="codeRef" target="_blank" href="https://en.cppreference.com/w/cpp/container/array.html">std::array&lt;size_t, bio::dna5::alphabet_size&gt;</a> <a class="code" href="group__core__type__pack.html#ga99b61a4ab3c4fcc3d78cdb8b8ac2a392">count</a>{}; <span class="comment">// default initialised with zeroes</span></div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Increase the symbol count according to the sequence.</span></div>
<div class="line">    <span class="keywordflow">for</span> (<a class="code" href="classbio_1_1dna5.html">bio::dna5</a> symbol : <a class="code" href="interfacebio_1_1sequence.html">sequence</a>)</div>
<div class="line">        ++<a class="code" href="group__core__type__pack.html#ga99b61a4ab3c4fcc3d78cdb8b8ac2a392">count</a>[symbol.to_rank()];</div>
<div class="line"> </div>
<div class="line">    <span class="comment">// Calculate the GC content: (#G + #C) / (#A + #T + #G + #C).</span></div>
<div class="line">    <span class="keywordtype">size_t</span> gc = <a class="code" href="group__core__type__pack.html#ga99b61a4ab3c4fcc3d78cdb8b8ac2a392">count</a>[<span class="charliteral">&#39;C&#39;</span>_dna5.to_rank()] + <a class="code" href="group__core__type__pack.html#ga99b61a4ab3c4fcc3d78cdb8b8ac2a392">count</a>[<span class="charliteral">&#39;G&#39;</span>_dna5.to_rank()];</div>
<div class="line">    <span class="keywordtype">size_t</span> atgc = input.size() - <a class="code" href="group__core__type__pack.html#ga99b61a4ab3c4fcc3d78cdb8b8ac2a392">count</a>[<span class="charliteral">&#39;N&#39;</span>_dna5.to_rank()];</div>
<div class="line">    <span class="keywordtype">float</span> gc_content = 1.0f * gc / atgc;</div>
<div class="line"> </div>
<div class="line">    fmt::print(<span class="stringliteral">&quot;The GC content of {} is {}%\n&quot;</span>, <a class="code" href="interfacebio_1_1sequence.html">sequence</a>, 100 * gc_content);</div>
<div class="line"> </div>
<div class="line">    <span class="keywordflow">return</span> 0;</div>
<div class="line">}</div>
<div class="ttc" id="aarray_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/header/array.html">array</a></div></div>
<div class="ttc" id="abasic_string_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/string/basic_string.html">std::string</a></div></div>
<div class="ttc" id="aclassbio_1_1dna5_html"><div class="ttname"><a href="classbio_1_1dna5.html">bio::dna5</a></div><div class="ttdoc">The five letter DNA alphabet of A,C,G,T and the unknown character N.</div><div class="ttdef"><b>Definition:</b> dna5.hpp:50</div></div>
<div class="ttc" id="afmt_8hpp_html"><div class="ttname"><a href="fmt_8hpp.html">fmt.hpp</a></div><div class="ttdoc">Core alphabet concept and free function/type trait wrappers.</div></div>
<div class="ttc" id="agroup__alphabet_html_ga05da42eb0b9a48f4d1a554c7d3a9044d"><div class="ttname"><a href="group__alphabet.html#ga05da42eb0b9a48f4d1a554c7d3a9044d">bio::assign_char_to</a></div><div class="ttdeci">constexpr auto assign_char_to</div><div class="ttdoc">Assign a character to an alphabet object.</div><div class="ttdef"><b>Definition:</b> concept.hpp:418</div></div>
<div class="ttc" id="agroup__core__type__pack_html_ga99b61a4ab3c4fcc3d78cdb8b8ac2a392"><div class="ttname"><a href="group__core__type__pack.html#ga99b61a4ab3c4fcc3d78cdb8b8ac2a392">bio::pack_traits::count</a></div><div class="ttdeci">constexpr ptrdiff_t count</div><div class="ttdoc">Count the occurrences of a type in a pack.</div><div class="ttdef"><b>Definition:</b> traits.hpp:170</div></div>
<div class="ttc" id="ainterfacebio_1_1sequence_html"><div class="ttname"><a href="interfacebio_1_1sequence.html">sequence</a></div><div class="ttdoc">The generic concept for a sequence.</div></div>
<div class="ttc" id="aranges_2views_2all_8hpp_html"><div class="ttname"><a href="ranges_2views_2all_8hpp.html">all.hpp</a></div><div class="ttdoc">Meta-header for the views submodule .</div></div>
<div class="ttc" id="astring_html"><div class="ttname"><a href="https://en.cppreference.com/w/cpp/header/string.html">string</a></div></div>
</div><!-- fragment --><p> </details> </div> </p>
<h1><a class="anchor" id="autotoc_md23"></a>
Other alphabets</h1>
<p>Until now, we have focused on alphabets for nucleotides to introduce the properties of SeqAn's alphabet on a specific example. SeqAn implements, however, many more alphabets. In this section, we want to give you an overview of the other existing alphabets.</p>
<h2><a class="anchor" id="autotoc_md24"></a>
The amino acid alphabet</h2>
<p>Proteins consist of one or more long chains of amino acids. The so-called primary structure of a protein is expressed as sequences over an amino acid alphabet. The <a class="el" href="classbio_1_1aa27.html" title="The twenty-seven letter amino acid alphabet.">bio::aa27</a> alphabet contains the standard one-letter code of the 20 canonical amino acids as well as the two proteinogenic amino acids, a termination symbol and some wildcard characters. For details read the <a class="el" href="group__aminoacid.html">Aminoacid</a> page.</p>
<h2><a class="anchor" id="autotoc_md25"></a>
Structure and quality alphabets</h2>
<p>The alphabets for structure and quality are sequence <em>annotations</em> since they describe additional properties of the respective sequence. We distinguish between three types:</p><ol type="1">
<li><b>Quality alphabet for nucleotides</b>. The values are produced by sequencing machines and represent the probability that a nucleobase was recorded incorrectly. The characters are most commonly found in FASTQ files. See <a class="el" href="group__quality.html">Quality</a> for details.</li>
<li><b>RNA structure alphabets</b>. They describe RNA nucleobases as unpaired or up-/downstream paired and can be found in annotated RNA sequence and alignment files (e.g. Stockholm format). Currently we provide the <a class="el" href="classbio_1_1dot__bracket3.html">Dot Bracket</a> and <a class="el" href="classbio_1_1wuss.html">WUSS</a> formats.</li>
<li><b>Protein structure alphabet</b>. The <a class="el" href="classbio_1_1dssp9.html">DSSP</a> format represents secondary structure elements like alpha helices and turns.</li>
</ol>
<p>You can build an <a class="el" href="classbio_1_1alphabet__tuple__base.html">Alphabet Tuple Composite</a> with a nucleotide and quality alphabet, or nucleotide / amino acid and structure alphabet that stores both information together. For the use cases just described we offer pre-defined composites (<a class="el" href="classbio_1_1qualified.html" title="Joins an arbitrary alphabet with a quality alphabet.">bio::qualified</a>, <a class="el" href="classbio_1_1structured__rna.html" title="A bio::alphabet_tuple_base that joins a nucleotide alphabet with an RNA structure alphabet.">bio::structured_rna</a>, <a class="el" href="classbio_1_1structured__aa.html" title="A bio::alphabet_tuple_base that joins an aminoacid alphabet with a protein structure alphabet.">bio::structured_aa</a>). See our API documentation for a detailed description of each.</p>
<h2><a class="anchor" id="autotoc_md26"></a>
Gap alphabet</h2>
<p>The <a class="el" href="classbio_1_1gap.html" title="The alphabet of a gap character &#39;-&#39;.">bio::gap</a> alphabet is the smallest alphabet in SeqAn, consisting only of the gap character. It is most often used in an <a class="el" href="classbio_1_1alphabet__variant.html">Alphabet Variant</a> with a nucleotide or amino acid alphabet to represent gapped sequences, e.g. in alignments. To create a gapped alphabet simply use <a class="el" href="group__gap.html#gac94d46e4d788b59d556962cf86633a4d" title="Extends a given alphabet with a gap character.">bio::gapped&lt;&gt;</a> with the alphabet type you want to refine.</p>
<div class="fragment"><div class="line">    <span class="comment">// Assign a gap symbol to a gapped RNA alphabet.</span></div>
<div class="line">    <a class="code" href="classbio_1_1alphabet__variant.html">bio::gapped&lt;bio::rna5&gt;</a> sym = <a class="code" href="classbio_1_1gap.html">bio::gap</a>{};                         <span class="comment">// =&gt; -</span></div>
<div class="line"> </div>
<div class="line">    <span class="keyword">using</span> bio::operator<span class="stringliteral">&quot;&quot;</span>_rna5;</div>
<div class="line">    <span class="comment">// Each bio::rna5 symbol is still valid.</span></div>
<div class="line">    sym = <span class="charliteral">&#39;U&#39;</span>_rna5;                                   <span class="comment">// =&gt; U</span></div>
<div class="line"> </div>
<div class="line">    <span class="comment">// The alphabet size is six (AUGCN-).</span></div>
<div class="line">    uint8_t <span class="keyword">const</span> size2 = <a class="code" href="classbio_1_1alphabet__variant.html">bio::gapped&lt;bio::rna5&gt;::alphabet_size</a>;   <span class="comment">// =&gt; 6</span></div>
</div><!-- fragment --></div></div><!-- contents -->
</div><!-- PageDoc -->
</div><!-- doc-content -->
<!-- HTML footer for doxygen 1.8.16-->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <div id="list_bottom_right" style="text-align:right; float:left; color:black; padding: 0px 0px 0px 10px"></div>
    <script type="text/javascript" charset="utf-8" src="version.js"></script>
    <li class="footer">Generated on Thu Jun 9 2022 19:02:37 for BioC++ by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
